---
title: "Supplementary note"
draft: true
output: 
  rjtools::rjournal_web_article:
    self_contained: yes
    toc: no
author:  
  - name: Bastien Chassagnol
    affiliation: LPSM
    email:  bastien_chassagnol@laposte.net
  - name: Antoine Bichat
    affiliation: Les Laboratoires Servier
  - name: Gregory Nuel
    affiliation: LPSM, UMR CNRS 8001
    email:  Gregory.Nuel@math.cnrs.fr
  - name: Etienne Becht
    affiliation: Les Laboratoires Servier
    email: etienne.becht@servier.com
header-includes:
  - \usepackage[most]{tcolorbox}
  - \usepackage{booktabs}
  - \usepackage{bm, amsmath}
  - \usepackage{array}
  - \usepackage{multirow}
  - \usepackage{wrapfig}
  - \usepackage{float}
  - \usepackage{colortbl}
  - \usepackage{hhline}
  - \usepackage{xcolor}
  - \usepackage{longtable}
  - \usepackage{tabu}
  - \usepackage{adjustbox}
  - \usepackage{threeparttable}
  - \usepackage{makecell}
  - \def\greentick{\includegraphics[scale=0.05]{figs/green_tick.png}}
  - \def\redcross{\includegraphics[scale=0.05]{figs/red_cross.png}}
  - \newtcolorbox{blackbox}[1]{colback=white, colframe=black, coltext=black, boxsep=1.5pt, arc=4pt, before=\centering, title=#1}
  - \newenvironment{cols}[1][]{}{} \newenvironment{col}[1]{\begin{minipage}{#1}\ignorespaces}{\end{minipage} \ifhmode\unskip\fi
  - \aftergroup\useignorespacesandallpars} \def\useignorespacesandallpars#1\ignorespaces\fi{#1\fi\ignorespacesandallpars} \makeatletter  
  - \def\ignorespacesandallpars{\@ifnextchar\par {\expandafter\ignorespacesandallpars\@gobble} {} } \makeatother
bibliography: chassagnol-becht-nuel-benchmark-of-Gaussian-mixtures.bib
zotero: false
date: "`r Sys.Date()`"
editor_options: 
  markdown: 
    wrap: 72
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = FALSE,
	warning = FALSE, 
	message = FALSE,
	fig.align = "center",
	fig.fullwidth=TRUE,
	message = FALSE,
	warning = FALSE,
	out.width = "100%",
	cache=FALSE, lazy.cache=FALSE)

library(ggplot2)
library(dplyr)
library(flextable)
library(kableExtra)
library(downlit)
library(RGMMBench)

options(dplyr.summarise.inform = FALSE) # remove the override .groups argument

# rjtools::initial_check_article(".", dic = "en_GB", ignore=c("gmms", "transcriptomic", "gaussian", "heteroscedascity", "equi", "clusterwise", "underbrace", "homoscedascity", "rebmix", "cran", "bioconductor", "mclust", "kd", "st", "th", "homoscedastic", "hyperparameters"))
```


```{css, echo=FALSE}
p.caption {
  font-size: 0.6em;
}

.blackbox {
  padding: 1em;
  background: #DBDBDB;
  color: black;
  border: 4px solid black;
  border-radius: 10px;
}
.center {
  text-align: center;
}

.cols {display: flex; }
```


```{r prepare-data}
#################################################################
##         prepare univariate scores and distributions         ##
#################################################################
univariate_distribution_parameters <- readRDS("./tables/univariate/univariate_distributions.rds") %>%
  dplyr::group_by(ID, package, initialisation_method) %>%
  dplyr::mutate(N.bootstrap=dplyr::row_number()) %>% dplyr::ungroup()
univariate_time_computation <- readRDS("./tables/univariate/univariate_time_computation.rds")
univariate_configuration <- readRDS("./tables/univariate/univariate_configuration_scenario.rds")

formatted_univariate_distribution_parameters <- univariate_distribution_parameters %>%
  rename_with(~gsub("^sd", "sigma", .x)) %>%
  mutate(color=dplyr::if_else(package %in% c("RGMMBench", "Rmixmod", "mixtools"), "#E7298A", "#1B9E77"),
         package=dplyr::if_else(package %in% c("flexmix", "mclust", "mixtools"),
                                glue::glue("<b style='color:{color}; font-size:22pt'>{package}</b>"),
                                glue::glue("<p style='color:{color}'>{package}</p>")))

univariate_parameters_local_scores <- readRDS("./tables/univariate/univariate_local_scores.rds") %>%
  filter(package %in% c("mixtools", "Rmixmod", "mclust", "EMCluster", "flexmix")) %>%
  mutate(package=forcats::fct_recode(package, `Rmixmod / RGMMBench`="Rmixmod", `mclust / bgmm`="mclust", `EMCluster / GMKMcharlie`="EMCluster"))

univariate_parameters_local_scores_summary <- univariate_parameters_local_scores %>% dplyr::filter(scores=="mse") %>%
  rowwise() %>% mutate(global_mse_p = sum(c(p1, p2, p3, p4)),
                       global_mse_mu = sum(c(mu1, mu2, mu3, mu4)),
                       global_mse_sigma=sum(c(sigma1, sigma2, sigma3, sigma4))) %>%
  dplyr::group_by(ID, initialisation_method, package) %>%
  dplyr::summarise(global_mse_p=mean(global_mse_p), global_mse_mu=mean(global_mse_mu), global_mse_sigma=mean(global_mse_sigma)) %>%
  dplyr::inner_join(univariate_parameters_local_scores %>% dplyr::filter(scores=="bias") %>%
                      rowwise() %>% mutate(global_bias_p = sum(abs(c(p1, p2, p3, p4))),
                                           global_bias_mu = sum(abs(c(mu1, mu2, mu3, mu4))),
                                           global_bias_sigma=sum(abs(c(sigma1, sigma2, sigma3, sigma4)))) %>%
                      dplyr::group_by(ID, initialisation_method, package) %>%
                      dplyr::summarise(global_bias_p=mean(global_bias_p), global_bias_mu=mean(global_bias_mu), global_bias_sigma=mean(global_bias_sigma))) %>%
  dplyr::arrange(ID, package, initialisation_method) %>%
  dplyr::relocate(initialisation_method, .after = package) %>%
  dplyr::rename(Package=package, `Initialisation Method`=initialisation_method) %>%
  mutate(across(where(is.numeric), ~format(.x, digits=2, nsmall=0, ) %>% as.numeric)) %>% dplyr::ungroup()


#################################################################
##        prepare bivariate scores and distributions        ##
#################################################################

bivariate_parameters <- readRDS("./tables/bivariate/bivariate_distributions.rds")%>%
  dplyr::group_by(ID, package, initialisation_method) %>%
  dplyr::mutate(N.bootstrap=dplyr::row_number()) %>% dplyr::ungroup()
bivariate_time_computation <- readRDS("./tables/bivariate/bivariate_time_computation.rds")
bivariate_configuration <- readRDS("./tables/bivariate/bivariate_configuration_scenario.rds")

formatted_bivariate_parameters <- bivariate_parameters %>%
  mutate(color=dplyr::if_else(package %in% c("RGMMBench", "Rmixmod", "mixtools"), "#E7298A", "#1B9E77"),
         package=dplyr::if_else(package %in% c("flexmix", "mclust", "mixtools"),
                                glue::glue("<b style='color:{color}; font-size:22pt'>{package}</b>"),
                                glue::glue("<p style='color:{color}'>{package}</p>")))


bivariate_parameters_local_scores <- readRDS("./tables/bivariate/bivariate_local_scores.rds") %>%
  filter(package %in% c("mixtools", "Rmixmod", "mclust", "EMCluster", "flexmix")) %>%
  mutate(package=forcats::fct_recode(package, `Rmixmod / RGMMBench`="Rmixmod", `mclust / bgmm`="mclust", `EMCluster / GMKMcharlie`="EMCluster"))

bivariate_parameters_local_scores_summary <- bivariate_parameters_local_scores %>% dplyr::filter(scores=="mse") %>%
  rowwise() %>% mutate(global_mse_p = sum(dplyr::c_across(dplyr::matches("^p[[:digit:]]+$"))),
                       global_mse_mu = sum(dplyr::c_across(dplyr::matches("mu"))),
                       global_mse_sigma=sum(dplyr::c_across(dplyr::matches("sd")))) %>%
  dplyr::group_by(ID, initialisation_method, package) %>%
  dplyr::summarise(global_mse_p=mean(global_mse_p), global_mse_mu=mean(global_mse_mu), global_mse_sigma=mean(global_mse_sigma)) %>%
  dplyr::inner_join(bivariate_parameters_local_scores %>% dplyr::filter(scores=="bias") %>%
                      rowwise() %>% mutate(global_bias_p = sum(abs(dplyr::c_across(dplyr::matches("^p[[:digit:]]+$")))),
                                           global_bias_mu = sum(abs(dplyr::c_across(dplyr::matches("mu")))),
                                           global_bias_sigma=sum(abs(dplyr::c_across(dplyr::matches("sd"))))) %>%
                      dplyr::group_by(ID, initialisation_method, package) %>%
                      dplyr::summarise(global_bias_p=mean(global_bias_p), global_bias_mu=mean(global_bias_mu), global_bias_sigma=mean(global_bias_sigma))) %>%
  dplyr::arrange(ID, package, initialisation_method) %>%
  dplyr::relocate(initialisation_method, .after = package) %>%
  dplyr::rename(Package=package, `Initialisation Method`=initialisation_method) %>%
  mutate(across(where(is.numeric), ~format(.x, digits=2, nsmall=0, ) %>% as.numeric)) %>% ungroup()


#################################################################
##        prepare HD scores and distributions        ##
#################################################################

HD_distributions <- readRDS("./tables/HD/HD_distributions.rds")
HD_configuration <- readRDS("./tables/HD/HD_configuration_scenario.rds")
HD_time_computation <- readRDS("./tables/HD/HD_time_computation.rds")
HD_parameters_local_scores <- readRDS("./tables/HD/HD_local_scores.rds") 


formatted_HD_distributions <- HD_distributions %>% 
  mutate(package=forcats::fct_relevel(package, "RGMMBench", "Rmixmod", "mixtools",
                                      "GMKMcharlie", "mclust", "flexmix",
                                      "bgmm", "EMCluster", "HDclassif", "EMMIXmfa")) %>% 
  arrange(ID, initialisation_method, N.bootstrap, package) %>% 
  mutate(color=dplyr::case_match(package, 
                                 c("RGMMBench", "Rmixmod", "mixtools") ~ "#E7298A", 
                                 c("HDclassif", "EMMIXmfa") ~"#7570B3", .default = "#1B9E77"), 
         package=dplyr::case_match(package, 
                                   c("flexmix", "mclust", "mixtools") ~ glue::glue("<b style='color:{color}; font-size:22pt'>{package}</b>"), 
                                   c("HDclassif", "EMMIXmfa") ~ glue::glue("<i style='color:{color}'>{package}</i>"),
                                   .default = glue::glue("<p style='color:{color}'>{package}</p>"))) 


num_failed <- HD_distributions %>%
  filter(package %in% c("Rmixmod", "mclust", "bgmm", "EMCluster", "HDclassif", "EMMIXmfa")) %>%
  mutate(package=forcats::fct_recode(package, `mixtools / Rmixmod / RGMMBench`="Rmixmod",
                                     `mclust / flexmix / GMKMcharlie`="mclust")) %>% 
  group_by(ID, initialisation_method, package) %>% 
  summarise(Success=n())

HD_parameters_local_scores <- readRDS("./tables/HD/HD_local_scores.rds") %>% 
  filter(package %in% c("Rmixmod", "mclust", "bgmm", "EMCluster", "HDclassif", "EMMIXmfa")) %>%
  mutate(package=forcats::fct_relevel(package, "Rmixmod", "mclust", 
                                      "bgmm", "EMCluster", "HDclassif", "EMMIXmfa")) %>% 
  arrange(ID, initialisation_method, package) %>% 
  mutate(package=forcats::fct_recode(package, `mixtools / Rmixmod / RGMMBench`="Rmixmod",
                                     `mclust / flexmix / GMKMcharlie`="mclust"))

HD_parameters_local_scores_summary <- HD_parameters_local_scores %>% dplyr::filter(scores=="mse") %>%
  rowwise() %>% summarise(ID, initialisation_method, package, 
                          global_mse_p = sum(dplyr::c_across(dplyr::matches("p[[:digit:]]+"))),
                          global_mse_mu = sum(dplyr::c_across(dplyr::matches("mu"))),
                          global_mse_sigma=sum(dplyr::c_across(dplyr::matches("sd")))) %>% 
  dplyr::inner_join(HD_parameters_local_scores %>% dplyr::filter(scores=="bias") %>%
                      rowwise() %>% summarise(ID, initialisation_method, package, 
                                              global_bias_p = sum(abs(dplyr::c_across(dplyr::matches("^p[[:digit:]]+$")))),
                                              global_bias_mu = sum(abs(dplyr::c_across(dplyr::matches("mu")))),
                                              global_bias_sigma=sum(abs(dplyr::c_across(dplyr::matches("sd"))))),
                    by = join_by(ID, initialisation_method, package)) %>%
  dplyr::inner_join(num_failed, join_by(ID, initialisation_method, package)) %>% 
  dplyr::filter(if_all(everything(), ~ !is.na(.))) %>% # remove columns with NULL values  
  dplyr::arrange(ID, package, initialisation_method) %>%
  dplyr::relocate(initialisation_method, .after = package) %>%
  dplyr::rename(Package=package, `Initialisation Method`=initialisation_method) %>%
  mutate(across(where(is.numeric), ~format(.x, digits=2, nsmall=0, ) %>% as.numeric)) 


metric_colnames <- c("global_mse_p", "global_mse_mu", "global_mse_sigma", "global_bias_p",
                        "global_bias_mu", "global_bias_sigma")
```

For ease of reading, we show again in Table `r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(tab:general-parameter-description-html)', '\\@ref(tab:general-parameter-description-pdf)'))` the general configuration used to run all the benchmarks tested. 

```{r general-parameter-description-html, eval=knitr::is_html_output()}
read.table("./tables/general_simulation_parameters.csv", sep = ";", header = T, check.names = F) %>%
  flextable() %>%
  compose(j = "Criterion threshold", value = as_paragraph(as_equation(`Criterion threshold`, width = 2, height = .5))) %>%
  set_caption(caption = "Global options shared by all the benchmarked packages.", html_escape = F) %>%
  theme_vanilla() %>% align(align = "center", part = "header") %>%   flextable::autofit()
```

```{r general-parameter-description-pdf, eval=knitr::is_latex_output()}
formatted_data <- readr::read_delim("./tables/general_simulation_parameters.csv", delim=";") %>%
  dplyr::mutate(`Criterion threshold`=paste0("$", `Criterion threshold`, "$"))


formatted_data %>% kbl(booktabs=T, caption = "Global options shared by all the benchmarked packages.", escape=F, midrule="midrule", align = "c") %>%
  kable_styling(latex_options=c("hold_position", "scale_down")) %>%
  row_spec(0,bold=T) %>%
  row_spec(1, hline_after = T) %>%
  column_spec(1:ncol(formatted_data), width = "4.1cm") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"))
```

\appendix

# Appendix A: In-depth statistical elements about parameters estimation in GMMs {.appendix}

## Application of the EM algorithm to GMMs {.appendix}

While solving Equation (10) to retrieve the MLE estimates in the M-step of the EM algorithm, we have to enforce the non-negativity and sum-to-one constraint of the mixture models (Equation (2)). This is enabled by the \emph{Lagrange multipliers} tip, which consists in practice to add the equality constraint over the parameters to estimate, here $-\lambda (\sum_{j=1}^k p_j -1)$, to the function to be optimised [@walsh75].

The evaluation of the roots of the derivative of the auxiliary function
(see Equation (10)) at the parameter $p_j$ with the additional unit simplex constraint (2) allows to readily compute a MLE estimate of the ratios, valid for any finite mixture model (Equation \@ref(eq:ratios-estimate)):

\begin{equation}
\hat{p_j}= \frac{\sum_{i=1}^n \eta_{i}(j)}{n}
(\#eq:ratios-estimate)
\end{equation}

Additionally, we restrained in both the univariate and
multivariate settings to the fully *unconstrained parametrisation*, in which each component follows its own parametric distribution. The general
derivative of the auxiliary function with respect to each
component parametric distribution $\zeta_j$, is given by Equation
\@ref(eq:parameter-component-derivative)^[It is equivalent to compute the MLE of a sample following distribution $f_{\zeta_j}$ weighted by the vector of posterior probabilities.]:

```{=tex}
\begin{equation}
\frac{\partial Q(\theta|\hat{\theta}_{q-1})}{\partial \zeta_j}=\sum_{i=1}^n  \eta_{i}(j) \frac{\partial \log (f_{\zeta_j}(X_i|S_i=j))}{\partial \zeta_j}
(\#eq:parameter-component-derivative)
\end{equation}
```

Accordingly, if a closed form for the computation of the MLE in
supervised cases is known (and fortunately this is the case for both the univariate and multivariate Gaussian distributions), the computation of the maximum of the auxiliary function can be readily calculated.

Plug-in the corresponding parametric distribution in the auxiliary function  (10) yields the following formula for the univariate GMM (Equation
\@ref(eq:univariate-auxiliary)):

```{=tex}
\begin{equation}
Q(\theta| \hat{\theta}_{q-1}) = \sum_{i=1}^n \sum_{j=1}^k \eta_i(j) \left( \log (p_j) - \log (\sigma_j) - \frac{(X_i-\mu_j)^2}{2\sigma_j^2} \right) + K
(\#eq:univariate-auxiliary)
\end{equation}
```

and Equation \@ref(eq:multivariate-auxiliary) for the multivariate GMM:

```{=tex}
\begin{equation}
Q(\theta| \hat{\theta}_{q-1}) = \sum_{i=1}^n \sum_{j=1}^k \eta_i(j) \left[ \log (p_j) - \frac{1}{2} \left( \log(\operatorname{det}(\boldsymbol{\Sigma}_j)) + (x_i - \boldsymbol{\mu}_j)^\top \boldsymbol{\Sigma}_j^{-1}(x_i - \boldsymbol{\mu}_j)\right) \right] + K
(\#eq:multivariate-auxiliary)
\end{equation}
```

$K$ is a constant with respective values of $\frac{-nD\log(2\pi)}{2}$ and $\frac{-n\log(2\pi)}{2}$ in the univariate and multivariate setting.


In the univariate setting, the individual MLE mean $\mu_j$, and variance, $\sigma_j$, estimates are readily available (Equations \@ref(eq:mean-univariate) - \@ref(eq:sigma-univariate)):

```{=tex}
\begin{equation}
 \frac{\partial Q(\theta|\hat{\theta}_{q-1})}{\partial \mu_j} = 0
 \Leftrightarrow
  \mu_j = \frac{\sum_{i=1}^n \eta_i(j) X_i}{\sum_{i=1}^n \eta_i(j)}
(\#eq:mean-univariate)
\end{equation}
```

```{=tex}
\begin{equation}
 \frac{\partial Q(\theta|\hat{\theta}_{q-1})}{\partial \sigma_j} = 0
 \Leftrightarrow
\sigma_j^2  =  \frac{\sum_{i=1}^n \eta_i(j) (x_i - \mu_j)^2 }{\sum_{i=1}^n \eta_i(j)}
(\#eq:sigma-univariate)
\end{equation}
```


Before finding the optimum of the auxiliary function in the multivariate setting, we remind the interested reader of some relevant calculus formulas below:

::: {#properties-matrix .blackbox data-latex="{\\textbf{Transpose matrix properties}}"}
::: {.cols data-latex=""}
::: {.col data-latex="{0.3\\textwidth}"}
a.  $\operatorname{det}(p\boldsymbol{A})=p^G \operatorname{det} (\boldsymbol{A})$
:::

::: {.col data-latex="{0.3\\textwidth}"}
b.  $\operatorname{det}(A^{-1})=\frac{1}{\operatorname{det}(A)}$
:::

::: {.col data-latex="{0.3\\textwidth}"}
c.  $\left(\boldsymbol{A}^{-1} \right)^\top=\boldsymbol{A}^{-1}$[^6]
:::
:::
:::

[^6]: when $\boldsymbol{A}$ is itself symmetric, as by definition,
    $\boldsymbol{A}^\top=\boldsymbol{A}$

::: {#properties-calculus .blackbox data-latex="{\\textbf{Matrix calculus}}"}
Given a symmetric matrix $\boldsymbol{A}$ of full rank $D$ and two vectors
$\boldsymbol{x}$ and $\boldsymbol{\mu}$ of size $D$, the following derivative
properties hold:

::: {.cols data-latex=""}
::: {.col data-latex="{0.3\\textwidth}"}
a.  $\frac{\partial \boldsymbol{x}^\top \boldsymbol{A} \boldsymbol{x}}{\partial \boldsymbol{A}} = \boldsymbol{x} \boldsymbol{x}^\top$
:::

::: {.col data-latex="{0.3\\textwidth}"}
b.  $\frac{\partial (\boldsymbol{x}-\boldsymbol{\mu})^\top \boldsymbol{A} (\boldsymbol{x}-\boldsymbol{\mu})}{\partial \boldsymbol{\mu}} = -2 \boldsymbol{A}(\boldsymbol{x}-\boldsymbol{\mu})$
:::

::: {.col data-latex="{0.3\\textwidth}"}
c.  $\frac{\partial \log\left(\operatorname{det}(\boldsymbol{A})\right)}{\partial \boldsymbol{A}^{-1}} = -\boldsymbol{A}$
^[Other matrix calculus formulas and notations are available on [Matrix
calculus](https://en.wikipedia.org/wiki/Matrix_calculus#Scalar-by-vector)
and demonstration details from *The Matrix Cookbook* [@Petersen2008].]
:::
:::
:::

Using the calculus formulas derived in the previous boxes, a closed form for the MLE estimate of the mean, $\boldsymbol{\mu}_j$, and covariance, $\boldsymbol{\Sigma}_j$, is readily computed (see Equations \@ref(eq:mean-bivariate) - \@ref(eq:sigma-bivariate)):

```{=tex}
\begin{equation}
\frac{\partial Q(\theta|\hat{\theta}_{q-1})}{\partial \boldsymbol{\mu}_j} =
\sum_{i=1}^n \eta_i(j) \boldsymbol{\Sigma}_j^{-1}(x_i - \boldsymbol{\mu}_j) =0
\Leftrightarrow
  \mu_j = \frac{\sum_{i=1}^n \eta_i(j) \boldsymbol{x}_i}{\sum_{i=1}^n \eta_i(j)}
(\#eq:mean-bivariate)
\end{equation}
```
```{=tex}
\begin{equation}
 \frac{\partial Q(\theta|\hat{\theta}_{q-1})}{\partial \boldsymbol{\Sigma}_j^{-1}}=\frac{1}{2} \sum_{i=1}^n \eta_i(j) \left[ \boldsymbol{\Sigma}_j  - (x_i - \boldsymbol{\mu}_j)(x_i - \boldsymbol{\mu}_j)^\top \right]= 0
\Leftrightarrow
  \boldsymbol{\Sigma}_j = \frac{\sum_{i=1}^n \eta_i(j) (x_i - \boldsymbol{\mu}_j)(x_i - \boldsymbol{\mu}_j)^\top }{\sum_{i=1}^n \eta_i(j)}
(\#eq:sigma-bivariate)
\end{equation}
```

Explicitly optimising the equations (\@ref(eq:univariate-auxiliary)-\@ref(eq:multivariate-auxiliary)) yield the following MLE parameters in both the univariate and multivariate settings (Table `r knitr::asis_output(ifelse(knitr::is_html_output(), ' \\@ref(tab:summary-em-html)', '\\@ref(tab:summary-em-pdf)'))`), as detailed in [@leytham84; @redner_walker84]:


```{r summary-em-html, layout="l-body-outset", eval = knitr::is_html_output()}
data <- readr::read_delim("./tables/summary_em_estimation.csv", delim=";", show_col_types = F) %>%
  tibble::column_to_rownames("Column1")


data %>% flextable() %>%
  flextable::mk_par(j="Univariate GMM",
                    value = as_paragraph(as_equation(`Univariate GMM`, width = 10, height = 2, unit = "cm"))) %>%
  flextable::mk_par(j="Multivariate GMM",
                    value = as_paragraph(as_equation(`Multivariate GMM`, width = 10, height = 2, unit = "cm"))) %>%
  merge_at(i = 2, j = 1:2) %>%
  set_caption(caption = "An overview of the practical implementation of the EM algorithm in GMMs.",html_escape = F) %>%
  theme_vanilla() %>%
  align(align = "center", part = "header") %>%  # centre align header
  flextable::autofit()

```


```{r summary-em-pdf,  eval = knitr::is_latex_output()}
readr::read_delim("./tables/summary_em_estimation.csv", delim=";", show_col_types = F) %>%
  tibble::column_to_rownames("Column1") %>%
  mutate(dplyr::across(.cols = c("Univariate GMM", "Multivariate GMM"), ~ paste0("$", .x, "$"))) %>%
  kbl(booktabs=T, caption = "An overview of the practical implementation of the EM algorithm in GMMs.", escape=F) %>%
  kable_styling(latex_options=c("hold_position", "scale_down")) %>%
  row_spec(0,bold=T, align = "c") %>%
  row_spec(1:3, hline_after = T, align = "c") %>%
  column_spec(column = 1:3, width = rep("6.8cm", 3), latex_valign = "m") %>%
  column_spec(column = 1:2,  border_right = T)
```

In both cases, obtaining the parameters of each component's parametric distribution
turn to be equivalent to the computation of the mean and variance of a weighted sample, which can be computed in R with `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("stats::weighted.mean"),   '\\code{stats::weighted.mean}'))` and
`r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("stats::cov.wt"),   '\\code{stats::cov.wt}'))` functions^[We assign "ML" to the argument *method* to get the biased but true MLE estimate of the covariance]. Importantly, the value of the mapping function only depends on the set of the observations $X$, but does not depend on the parameter to estimate $\theta$. Indeed, the statistic computed by the EM algorithm is sufficient, which is one of its main advantages.


The complete code associated to our R implementation is implemented respectively
with `enmix_univariate` and `enmix_bivariate` for the univariate and multivariate setting,
available on GitHub at [RGMMBench](https://github.com/bastienchassagnol-servier/RGMMBench), as well as the
programs used to generate the several plots and tables of the article.
We additionally made two choices not clearly set in the literature:

-   The algorithm stops when when the absolute difference between
    consecutive log-likelihoods falls below a user-defined threshold
    *epsilon*, with a maximal number of *itmax* iterations allowed to reach this convergence.
-   In order to avoid numerical underflows resulting in inconsistent ratios, of type $0/0$, we rely on the fact that Gaussian distributions belong to the exponential family to log-rescale our observations and compute efficiently the posterior probabilities in the E-step of the EM algorithm. First, to avoid null values for highly unlikely observations, those far from the centroids, we use the log attribute of `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("stats::dnorm"),   '\\code{stats::dnorm}'))` and
`r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("mvtnorm::dmvnorm"),   '\\code{mvtnorm::dmvnorm}'))` functions, see Equation \@ref(eq:sample-likelihood):

```{=tex}
\begin{equation}
\begin{split}
        \ell (\theta|x) &= \log (\sum_{j=1}^k p_j f{\zeta_j} (x)) \\
        & = \log \left( \exp \left[ \log(p_j) + \log(f{\zeta_j} (x)) \right]\right) \quad
\end{split}
(\#eq:sample-likelihood)
\end{equation}
```

Second, we rewrite our sum of exponentials, the one enclosed into the log, to use the Taylor' series of $\log(1+x)$, with $|x|\ll 1$, see Equation \@ref(eq:logsumexp):

```{=tex}
\begin{equation}
\begin{split}
\log \left( \sum_{j=1}^k e^{a_j}\right)&=\log \left( \exp(a_j') \times \left[ 1+ \sum_{j\neq j'}\exp\left(\frac{a_j}{a_{j'}}\right) \right] \right) \\
& = a_{j'} + \operatorname{log1p} \left(\sum_{j\neq j'}\exp\left(\frac{a_j}{a_{j'}}\right)\right), \, \text{ with } j'=\arg \max_{\forall j \in \{1, \ldots, k\}} (e^{a_j})
\end{split}
(\#eq:logsumexp)
\end{equation}
```

with `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("base::log1p()"),   '\\code{log1p}'))` the R function dedicated for this Taylor's development. The posterior probabilities are then given by Equation
\@ref(eq:sample-posterior-probability):

```{=tex}
\begin{equation}
\log\left(\mathbb{P}_\theta(S=j | X=x)\right)=\log(p_j) + \log(f_{\zeta_j}) - \ell (\theta |x)
(\#eq:sample-posterior-probability)
\end{equation}
```
-   We stop the algorithm early when the estimates are trapped in the boundaries of the parameter space, typically when the ratio of a component or its associated variance tends to zero. This case rarely occurs in our simulations: once in univariate and never in multivariate.



## Parsimonious parametrisation of multivariate GMMs {.appendix}
\label{subsec:parsimonious-parametrisation}

Parsimonious parametrisation of GMMs models are provided by the following *eigenvalue* factorisation of the covariance matrix (Equation \@ref(eq:covariance-decomposition)):

```{=tex}
\begin{equation}
    \boldsymbol{\Sigma}_j=\lambda_j \boldsymbol{Q}_j \boldsymbol{D}_j \boldsymbol{Q}_j^\top
    (\#eq:covariance-decomposition)
\end{equation}
```


with $\lambda_j=\operatorname{det}\left(\boldsymbol{\Sigma}_j\right)^{\frac{1}{D}}$ a scalar proportional to the total volume of the ellipsoid (or area in bidimensional setting), $\boldsymbol{D}_j$ a diagonal matrix storing the eigenvalues normalised such that $|\boldsymbol{D}_j|=1$^[@R-Rmixmod enforces an additional but, in our opinion, superfluous constraint that the eigen values are sorted by decreasing order] and $\boldsymbol{Q}_j$ a $\mathcal{M}_D(\mathbb{R})$ *orthogonal matrix* whose
columns are $D$ linearly independent *eigenvectors* generating an orthonormal basis in $\mathbb{R}^D$ while $\boldsymbol{Q}_j^\top$ is its corresponding transpose matrix. The existence of the decomposition is guaranteed by the positive definiteness constraint over the covariance matrix while the orthogonality of $\boldsymbol{Q}_j$ results from its symmetry. When the matrix to factorise is positive-definite and symmetric, we also refer to it as *spectral decomposition*, a special case of *eigendecomposition*.


Each of these matrices can be constrained to be equal or variable across clusters, hence this decomposition reveals 14 possible models with different geometric characteristics, namely:

- two models with the *spherical family*, for which only $\lambda_j$ is used to control the *isotropic* (same radius in any dimension) volume of each component of the corresponding distribution structure

- four models with the *diagonal family*, using $\lambda_j$ with possibly distinct diagonal elements and $\boldsymbol{D}_j$ to specify the shape of the density contours. In that context, $\boldsymbol{Q}_j$ is henceforth a permutation matrix, whose inputs are only zeros and an unique one per row.

- eight models with the *general family*, using additionally $\boldsymbol{Q}_j$ to determine the orientation of the main axes of the ellipsoids. Indeed, in the last two families described, this matrix was equal to the identity, hence the axis of the ellipsoids were aligned with the standard $\mathbb{R}^D$ basis.

We detail the main characteristics of the 14 parametrisations (28 if we add for each model the equiproportional hypothesis) in Table (\@ref(tab:param-multivariate-gaussian)):


```{r param-multivariate-gaussian}
data <- readr::read_delim("./tables/parametrisation_multivariate_gaussian.csv", delim=",", show_col_types = F) %>%
  mutate(Representation="",
         dplyr::across(.cols = c("Notation", "Number of parameters"), ~ paste0("$", .x, "$")))
Representation <- file.path("./tables", "gaussian_param",paste0(data$Model, ".png"))


data %>% kbl(booktabs=T, caption = "The 14 canonical parametrisations of the within-group covariance matrix $\\boldsymbol{\\Sigma_j}$ with the corresponding geometric representations.", escape=F) %>%
  kable_styling(latex_options=c("scale_down"), font_size = 7) %>%
  row_spec(0,bold=T, align = "c") %>%
  row_spec(1:14, align = "c", hline_after = T) %>%
  column_spec(6, image=spec_image(Representation, 400, 235, res=600)) %>%
  column_spec(column = 1:6, width = c("2cm", "3cm", "3cm", "2cm", "4.5cm", "5cm"), latex_valign = "m")
```


* The first column describes in general and understandable terms each parametrisation, with I meaning invariant (alternatively, not used in the parametrisation), E means equal and V variable while the second column matches the corresponding matrix decomposition of the covariance matrix. These 14 models are all included in one of the three super-families: spherical, diagonal and ellipsoidal listed before. As an example, the model VEI has variable volumes $\lambda_j$ in relation with the cluster, however shares same general shape (as we can note on the Representations, all isodensities are distributed along the $x$-axis) and invariant directions (in other words, the transition matrix is the identity matrix, entailing that all scatter plots are aligned with the Cartesian coordinate axes).


* Varying the volume $\lambda_j$, given a fixed $\boldsymbol{Q}$ and $\boldsymbol{D}$, amounts to an *enlargement* (when all dimensions of a figure are changed in the same scale, also referred to as *isotropic* transformation),  varying the eigenvectors $\boldsymbol{Q}_j$, given a fixed volume $\lambda_j$ and $\boldsymbol{D}$ is equivalent to a rotation and finally varying the diagonal matrix $\boldsymbol{D}_j$, given the other parameters of Equation \@ref(eq:covariance-decomposition) are fixed, results in a *distortion* of the representation.


* CF means that the M-step is in closed form while IP entails that the M-step is iterative.

* The number of parameters enumerates the *degrees of freedom*, namely the number of parameters to truly estimate once the sum-to-one constraint is enforced (Equation (2)). In detail, $k$ is the number of components of the GMM model, $D$ its dimension, $\alpha=kD + k - 1$ is the number of parameters required to identify the mean vector of each component ($kD$) and the ratios $k-1$ and $\beta=\frac{D(D+1)}{2}$ the number of covariance terms to estimate for a given component ($D$ variance diagonal terms, the remaining terms being the pairwise symmetric covariance terms between the features). Note that the complexity of the covariance matrix in the fully unconstrained model (Model VVV) grows linearly with the number of components while exploding in the order $\mathcal{O}(D)$ with the number of dimensions. Meantime, the complexity of the parametrisation with the homoscedastic spherical family (Model EII) is constant.



* Last column displays the 14 most common GMMs parametrisations, by plotting the ellipses and centroids of a three components bivariate GMM parametrised by the mean vector and covariance of each component. For any additional detail, we refer the interested reader to \pkg{mclust} [@scrucca_etal16] and \pkg{Rmixmod} [@R-Rmixmod] vignettes for a general introduction to GMMs and to [@banfield_raftery93; @celeux_govaert92; @browne_mcnicholas14] for the closed formulas of the models.

## Parameters estimation in a high-dimensional context {.appendix}
\label{subsec:high-dimensional}

However, while parsimonious representations can largely reduce the computational burden, none of them in the general family is able to handle degenerate cases where the number of features, $D$, exceeds the number of observations $n$. Likewise situations, when the number of features is consequent, are referred to as high-dimensional, raising the well-known issue of the "curse of the dimensionality". Two distinct approaches have been developed in the literature to handle these degenerate cases:

- The most naive approach aims to eliminate the least informative variables by applying a strong Lasso-type penalty on the parameters to be estimated. We only came across such an approach twice among the reviewed R packages, in the specific context of regressions of mixtures (see\pkg{RobMixReg} and \pkg{fmerPack} packages).

- The second category includes a larger diversity of methods, all inspired from the *factor analysis* approach whose paradigm is to consider that all the $D$ features used to describe the observations can be spanned in a smaller subspace without lose of information. Precisely, the factor analysis theory describes the variability among observed and correlated variables by a substantial lower number of unobserved variables called *factors* or *latent variables*.
In practice, for a given component $j$, the diagonal matrix storing the eigenvalues is decomposed into two-blocks. The first upper-right diagonal block, assumed generally of dimension $d_j \ll D$, stores the largest $d_j$ eigenvalues and model the variance of the actual data of component $j$ while the lower-left diagonal block, of dimension $D-d_j$, stores an unique parameter that can be interpreted as the variance of the residual error terms, constrained to be strictly inferior to the lowest variability of the informative variables. The dimension $d_j$ can be considered as the intrinsic dimension of the latent subspace of cluster $j$ spanned by the first $d_j$ eigenvectors of $\boldsymbol{Q}_j$^[Starting from eigen-decomposition described in (Equation \@ref(eq:covariance-decomposition)), this approach is equivalent to consider only the $d_j$ largest eigenvalues resulting from the decomposition and sets the others to null.].

When the sub dimension $d_j$ is known, a closed version is generally available for the M-step of the EM algorithm, however $d_j$ is itself an hyperparameter to estimate. Though, [@bouveyron_etal11] has shown that a classical Cattell’s scree-test could be used to asymptotically estimate the intrinsic dimension of each cluster. Compared to the previous approach, this method has a strong theoretical background and strong impact on the running times performance.

Taking a concrete use case from the help documentation of the package \pkg{HDclassif}, it enabled to cluster a dataset of 10 classes with 130 observations overall and described in a 1024-dimensional space (consider the famous machine-learning digit recognition problem). Variants of these approaches have been developed in the following packages: \pkg{HDclassif}, \pkg{fabMix}, \pkg{EMMIXmfa} and \pkg{pgmm}. We refer the interested reader to the educational vignette of \pkg{HDclassif}: `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::href_topic("HDclassif", "HDclassif"),  paste0('\\href{', downlit::href_topic('HDclassif', 'HDclassif'),'}{HDclassif}')))` and papers [@mcnicholas_murphy08; @mcnicholas_etal10;@mcnicholas_murphy10].


Historically, the first mention of a probabilistic framework with an application to dimension reduction in the context of finite mixture models goes back to @tipping_bishop99, based on principal component analysis. @mclachlan_etal03 and @mclachlan_peel00a extend this original model by postulating that the distribution of the data within any latent class could be described using the tools of the factor analysis field^[Although principal component analysis and factor analysis are closely related, we can differentiate both approaches by their differing objective: while PCA seeks to capture the overall variability of the dataset, factor analysis focuses on describing the intra-variability between covariates. In practice, the differences between the two approaches are minor, we can notably show that the output of PCA is one of the solutions suggested by standard factor analysis.] Finally, building on the parsimonious parametrisations already theorised for GMMs (see previous section) , @mcnicholas_murphy08, @mcnicholas_etal10 and @bouveyron_etal07 proposed a variety of constraints, but this time directly defined on the projected subspace. Since all methods based on factor analysis provide a transition matrix, using the two or three most informative eigen values and their associated eigen vectors in order to project the dataset on a smaller subspace provides a simple visualisation tool for representing high dimensional datasets.
However, this method may is not suitable for unravelling the clustering structure. Instead, *the GMMDR method*, first proposed by @scrucca10 and implemented in the  `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("mclust::MclustDR()"),   '\\code{MclustDR}'))` function, from \pkg{mclust} package, aims at recovering the subspace that best captures the underlying latent clustering structure (we notably expect invariance of the global overlap in the sampling space and the corresponding projected subspace). More precisely, the main objective of the GMMDR technique is to infer the global *change-of-basis matrix* $\boldsymbol{Q}$ that minimises the differences in the a posteriori probabilities of assigning each observation $i$ to a given cluster $s_i$, knowing the value of the vector of observed covariates $\boldsymbol{x}_i$. Namely, we are looking for the orientation matrix $\boldsymbol{Q}$ that maximally ensures the following objective (Eq. \@ref(eq:GMMDR)):
```{=tex}
\begin{equation}
 \hat{\boldsymbol{Q}} = \arg \max_{\boldsymbol{Q}} \,  \left(\mathbb{P}_{\theta} (S_i=j | \boldsymbol{X} =\boldsymbol{x}_i) = \mathbb{P}_{\theta} (S_i=j | \boldsymbol{X} \boldsymbol{Q}) \right)
\text{ such that } S \perp \boldsymbol{X} | \boldsymbol{X} \boldsymbol{Q}
    (\#eq:GMMDR)
\end{equation}
```
This procedure itself derives from the *sliced inverse regression* algorithm [@li91], but instead of conditioning on the known response variable, GMMDR conditions on the estimated MAP cluster assignments. Since the solution returned by the following optimization problem is not unique, we generally constrain the projection matrix to be orthonormal (any of the vectors forming the basis are pairwise orthogonal, and individually of norm 1).



## Model selection {.appendix}

When comparing several models with several number of components or
parametrisations, the likelihood is uninformative as it can be
arbitrarily minimised by increasing the complexity of the model or
adding components. it is then necessary to penalise for complexity when
comparing them. The general form of the penalty metric, *GIC* (for
generalised information criteria), is given by Equation
\@ref(eq:GIC-score):

```{=tex}
\begin{equation}
    \operatorname{GIC} (\theta) =  \underbrace{p(\theta)}_{\text{penalty term}} - \underbrace{2 \ell (X | \theta)}_{\text{log-likelihood of the model}}
(\#eq:GIC-score)
\end{equation}
```
Among them, we set apart scores focused on selecting selecting the right
number of parameters and components, namely the *degrees of freedom*
(d.o.f.) of the model ($3k-1$ parameters for the univariate
unconstrained GMM), and those focusing on retrieving readable clusters.

In the first category, the *AIC* (Akaike information criterion)
[@schwarz78] is a *minimax-rate optimal* (score that minimises the risk
in the worst case) but inconsistent metric [@yang05] , proned to
overestimate the true number of components. *BIC* (Bayesian Information
Criterion), and CAIC (consistent AIC), accounting for both the number of
parameters and the sample size, are consistent metrics. Finally, the
*MDL*(Minimum Description Length) criterion accounts for the number of
parameters, sample size and number of components. Its core objective
differs from the others as it aims at reducing the amount of code to
encode both parameters and observations but is practically close to the
BIC metric. A thorough description of these scores, with their formulas
and theoretical properties, can be found in @fonseca08, @celeux_etal18.

In the second category, the most commonly implemented is the *ICL
(integrated complete-data likelihood)*, a BIC criterion with an
additional entropy penalty [@mclachlan_peel00]. As opposed to *BIC*,
the entropy term reduces the number of components to a well-separated
and readable clustering. Hence, it tends to underestimate their true
number when components are overlapping. Alternative similar metrics are
the *CLC* (Classification Likelihood Criterion), *AWE* (Approximate
Weight of Evidence) and *NEC* (Normalised Entropy Criterion) metrics
[@bacci_etal12]. The several metrics implemented by the reviewed
packages are listed in Table 2.

The *Likelihood-ratio test* (LRTS) can also be used to compare *nested
models*, with additional advantage to possibly derive a *p*-value
yielding the probability that a complex model (with more components)
should preferentially be used over a simpler one. Traditionally, common
process is to add one component after the other, until hypothesis $H_0$
can not be rejected anymore. Under standard regularity conditions of
Cramer's theorem, Wilk's theorem states that the Likelihood Ratio
distribution follows asymptotically a $\chi^2$ distribution, but
unfortunately these conditions are not met in mixture models
[@mclachlan_peel00]. To counterbalance it, bootstrap inference
[@mclachlan_peel00] is often used to derive an empirical distribution of
the Likelihood Ratio.

## Derivation of confidence intervals in GMMs {.appendix}

*Punctual estimation*, with a single estimate $\hat{\theta}$ for a given
*n*-sample, is not enough to evaluate the performance of a specific
method, as drawing another *n*-sample using the same parameters is
likely to lead to a different distribution and estimation of
$\hat{\theta}$. Instead, it can be interesting to retrieve the
distribution or at least the variability of the estimated parameters,
which can reveal useful to derive confidence intervals. However,
obtaining the distribution or even an asymptotic approximation of the
distribution of the parameters is not feasible in practice with mixture
models [@mclachlan_peel00]. Hence, most authors recommend to use
bootstrap methods for the generation of confidence intervals, as
suggested in [@efron_tibshirani93; @basford_etal97].

Bootstrap distributions of the parameters are generally retrieved via
*empirical* or *parametric* bootstrap, both available in the
\pkg{mclust} package. In the *empirical* or *non-parametric* bootstrap
@jaki_etal18, we draw iteratively $N$ samples of size $n$ with
replacement from the original observed variable $x_{1:n}$. In the
*parametric* bootstrap, $N$ simulations are built from the parameter
estimated with the available observations of $X$, via the EM algorithm
or any method used for parameter estimation. In both cases, we obtain an
empiric distribution of the parameter estimate:
$\hat{\theta}_{1:N}= (\hat{\theta}^1, \ldots, \hat{\theta}^N)$. Sample
mean and standard deviation (SD) of this empirical distribution can be
used to retrieve an asymptotic estimate of the variability of the
parameter estimate $\hat{\theta}$, the bias or the MSE of the parameter
estimates. To get unbiased estimates of the true standard deviation and
mean of the estimates, it is of common practice to compute the empirical
covariance matrix of the sample
$\operatorname{cov} [\hat{\theta}]= \frac{\sum_{j=1}^N (\hat{\theta_j} - \mathbb{E} [\hat{\theta}]) (\hat{\theta_j} - \mathbb{E} [\hat{\theta}])^\mathrm{T}}{N - 1}$,
the square roots of its diagonal terms corresponding to the empiric SDs.
Symmetric $1-\alpha$ asymptotic confidence intervals using the Central
Limit Theorem (CLT) can then be simply derived Equation
\@ref(eq:confidenceinterval):

```{=tex}
\begin{equation}
  \mathbb{E} [\hat{\theta}_t] \pm \frac{1}{\sqrt{n}}z_{1 - \frac{\alpha}{2}} \sqrt{\operatorname{var}(\hat{\theta}_t}, \quad \forall t \in \{1, \ldots, 3k\}
(\#eq:confidenceinterval)
\end{equation}
```

with $z_{1 - \frac{\alpha}{2}}$ the $1 - \frac{\alpha}{2}$ quantile of
the standard Gaussian distribution.

If computing the covariance matrix is not possible analytically, it can
be approximated by the expected Fisher Information Matrix
$\mathcal{I}_{\exp}(\theta)$ (FIM), given by Equation
\@ref(eq:expected-fim):

```{=tex}
\begin{equation}
     \left[\mathcal{I}_{\exp}(\theta)\right]_{1\le i \le 3k,1 \le j \le 3k} = - \mathbb{E} \left[\frac{\partial^2}{\partial \theta_i \partial \theta_j} \ell(\theta|X)\right]
(\#eq:expected-fim)
\end{equation}
```

Indeed, the Cramér-Rao theorem states that the diagonal elements of the
inverse of the FIM are upper bounded by the variability of the
parameters:
$\operatorname{var}(\hat{\theta}) \ge \frac{1}{\mathcal{I}(\theta)}$. This
implies that the ratio between inverse of the FIM and the variance
$e(\hat{\theta})=\frac{\mathcal{I}(\hat{\theta})^{-1}}{\operatorname{var}(\hat{\theta})}$
converges to 1, using the asymptotic efficiency of the MLE estimate of
GMMs.

Unfortunately, the computation of the expected FIM is still a hard task.
Hence it is generally replaced by the observed FIM, the negative of the
Hessian matrix of the incomplete log-likelihood function:
$\mathcal{I}_\text{obs}(\theta)= -\frac{\partial^2}{\partial \theta_i \partial \theta_j} \ell(\theta|X)$.
Exact general formulas are provided for the univariate case in @louis82
and for the multivariate case in @oakes99. Yet, it has to be noted that
the expected FIM generally outperforms the observed FIM in estimating
the covariance matrix of the MLE [@cao_spall12].

However all these methods require to compute second derivatives of the
log-likelihood leading to some disadvantages from a computational point
of view. More recently, @meng16 and @delattre_kuhn19 proposed an
accelerated algorithm requiring only computation of first order
derivatives. A similar alternative is implemented in the
\CRANpkg{mixsmsn} package [@R-mixsmsn]:
`r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("mixsmsn::im.smsn()"),   '\\code{mixsmsn::im.smsn}'))`, in which the Hessian matrix is approximated by the cross-product of the
gradient of the log-likelihood Equation
\@ref(eq:approximation-fim-gradient):

```{=tex}
\begin{equation}
        \mathcal{I}_{\text{obs}}(\theta) \approx - \frac{\partial \log (\ell (\theta|X))}{\partial \theta} \frac{\partial \log (\ell (\theta|X))}{\partial \theta}^T
    (\#eq:approximation-fim-gradient)
\end{equation}
```
according to an idea developed in paper @basford_etal97. For a more general introduction to Gaussian mixtures, including other models and parametrisations in the multivariate case, we refer the reader to the reference book *Gaussian parsimonious clustering models*
@celeux_govaert92.

## An analytic formula of the overlap for univariate Gaussian mixtures {.appendix}

From an analytic point of view, the overlap between $k$ components of
variable $X$ is given by Equation \@ref(eq:general-overlap):

```{=tex}
\begin{equation}
    \operatorname{OVL} (X)= 1 - \int_{\mathbb{R}} \max_j (p_j \varphi_{\zeta_j} (x)) dx
(\#eq:general-overlap)
\end{equation}
```
The 1 in Equation \@ref(eq:general-overlap) corresponds to the
integration of probability $f_{\theta}(X)$ distribution over its domain.
The second part is the area under the curve of the component density
function maximised on $\mathbb{R}$, with $j$ the index of the component
maximised at that point. It should be noted that the definition used here for the overlap is closely related to the definition of the  *false clustering rate* (FCR) [@ariane2022].

Equation \@ref(eq:general-overlap) simplifies for a two component
mixture distribution to Equation \@ref(eq:two-component-overlap):

```{=tex}
\begin{equation}
    \operatorname{OVL} (X)= \int_{\mathbb{R}} \min \left(p_1 \varphi_{\zeta_1} (x), p_2 \varphi_{\zeta_2} dx (x)\right)
(\#eq:two-component-overlap)
\end{equation}
```
From a probabilistic point of view, we can rewrite Equation
\@ref(eq:two-component-overlap) as the overall probability of assigning
a wrong label to a given observation. With two components, this simply
decomposes as the sum of the probability of mistakenly assigning an
observation from component 2 to component 1 and the probability of
assigning an observation from component 1 to component 2 Equation
\@ref(eq:misclassification-prob):

```{=tex}
\begin{equation}
\begin{split}
\operatorname{OVL} (1, 2) & = \operatorname{OVL} (1 | 2) + \operatorname{OVL} (2 | 1) \\
& = \mathbb{P}\left(p_1 \varphi (X, \mu_1, \sigma_1) \le p_2 \varphi (X, \mu_2, \sigma_2) \right) + \mathbb{P}\left(p_2 \varphi (X, \mu_2, \sigma_2) \le p_1 \varphi (X, \mu_1, \sigma_1)\right) \\
& = \int_{\mathbb{R}} p_1 \varphi_{\zeta_1} (x) \mathrm{1}_{p_1 \varphi_{\zeta_1}  \le p_2 \varphi_{\zeta_2}} dx + \int_{\mathbb{R}} p_2 \varphi_{\zeta_2} (x) \mathrm{1}_{p_2 \varphi_{\zeta_2} \le p_1 \varphi_{\zeta_1}} dx
\end{split}
(\#eq:misclassification-prob)
\end{equation}
```
We illustrate the computation of the overlap in some hard-hitting cases
below, showing relation between the level of entropy and the individual
standard deviations with the overlap measured in Figure
`r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(fig:OVL-represention-html)', '\\@ref(fig:OVL-represention-pdf)'))`.
Means of component 1 and 2 are 5.28 and 8.45. Panels A and C correspond
to balanced classes, while in panel B and D, class 1 is more abundant
with a frequency of 0.9. Finally, in panels A and B, the variance of
component 1 is smaller than the variance of component 2 with respective
SDs of 1 and 3 and reciprocally for panels B and D. Interestingly, in
panel D, using the MAP as defined in Equation \@ref(eq:posteriori), all
observations issued from class 2 are wrongly assigned to class 1. 

```{=tex}
\begin{equation}
        \eta_{i} (j) := \mathbb{P}_{\theta} (S_i=j |X_i=x_i)
    (\#eq:posteriori)
\end{equation}
```

The red area corresponds to the probability of misclassifying component 1 as
component 2, while the green area corresponds to the probability of
misclassifying component 2 as component 1. Total overlap is since the
sum of red and green area, in Figure
`r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(fig:OVL-represention-html)', '\\@ref(fig:OVL-represention-pdf)'))`.

```{r generate-overlap-plot-univariate, layout = "l-body-outset"}
temp_simu <- tibble::tibble(x=seq(0, 15, length.out = 1000))
true_theta <- list(p=c(0.5, 0.5), mu=c(5.28, 8.45), sigma=c(1, 3),skew=c(0,0))
for (i in 1:2) {
  temp_simu <- temp_simu %>% tibble::add_column("component {i}" := true_theta$p[i] *
                                                  sn::dsn(temp_simu$x, xi=true_theta$mu[i], omega = true_theta$sigma[i], alpha = true_theta$skew[i], tau=0))
}
temp_simu <- temp_simu %>% dplyr::mutate(total=rowSums(dplyr::across(dplyr::starts_with("component ")))) %>%
  tidyr::pivot_longer(cols =dplyr::starts_with(c("component ", "total")), names_to = "components", values_to = "y")

roots <- compute_overlap_roots(0.5, 0.5, 5.28, 8.45, 1,3)
overlap_plot_pairwise_balanced_1 <- ggplot(temp_simu, aes(x = x, color = gsub("component ", "", components), y=y,linetype=gsub("component ", "", components))) +
  geom_line(size=1) +
  theme_bw() +
  theme(plot.title = element_blank(), legend.position = "bottom", axis.title=element_blank(),
        legend.key.width = unit(1.5, 'cm'), legend.text = element_text(size=14),
        plot.subtitle = element_blank(), legend.title = element_blank(), plot.tag = element_text(size=16, face = "bold")) +
  scale_linetype_manual(values = c(2, 3, 1)) +
  geom_vline(xintercept=roots, color="red", linetype="dashed", size=1.15) +
  geom_area(data = dplyr::filter(temp_simu, x>max(roots) & components=="component 1"), fill = '#FDC086',show.legend = F) +
  geom_area(data = dplyr::filter(temp_simu, x <min(roots) & components=="component 1"), fill = '#FDC086',show.legend = F) +
  geom_area(data = dplyr::filter(temp_simu, x<max(roots) & x >min(roots) & components=="component 2"), fill = '#7FC97F',show.legend = F) +
  labs(tag="A")



temp_simu <- tibble::tibble(x=seq(0, 15, length.out = 1000))
true_theta <- list(p=c(0.9, 0.1), mu=c(5.28, 8.45), sigma=c(1, 3),skew=c(0,0))
for (i in 1:2) {
  temp_simu <- temp_simu %>% tibble::add_column("component {i}" := true_theta$p[i] *
                                                  sn::dsn(temp_simu$x, xi=true_theta$mu[i], omega = true_theta$sigma[i], alpha = true_theta$skew[i], tau=0))
}
temp_simu <- temp_simu %>% dplyr::mutate(total=rowSums(dplyr::across(dplyr::starts_with("component ")))) %>%
  tidyr::pivot_longer(cols =dplyr::starts_with(c("component ", "total")), names_to = "components", values_to = "y")
roots <- compute_overlap_roots(0.9, 0.1, 5.28, 8.45, 1,3)
overlap_plot_pairwise_balanced_2 <- ggplot(temp_simu, aes(x = x, color = gsub("component ", "", components), y=y,linetype=gsub("component ", "", components))) +
  geom_line(size=1) +
  theme_bw() +
  theme(plot.title = element_blank(), legend.position = "bottom", axis.title=element_blank(),
        legend.key.width = unit(1.5, 'cm'), legend.text = element_text(size=14),
        plot.subtitle = element_blank(), legend.title = element_blank(), plot.tag = element_text(size=16, face = "bold")) +
  scale_linetype_manual(values = c(2, 3, 1)) +
  geom_vline(xintercept=roots, color="red", linetype="dashed", size=1.15) +
  geom_area(data = dplyr::filter(temp_simu, x>max(roots) & components=="component 1"), fill = '#FDC086',show.legend = F) +
  geom_area(data = dplyr::filter(temp_simu, x <min(roots) & components=="component 1"), fill = '#FDC086',show.legend = F) +
  geom_area(data = dplyr::filter(temp_simu, x<max(roots) & x >min(roots) & components=="component 2"), fill = '#7FC97F',show.legend = F) +
  labs(tag="B")



temp_simu <- tibble::tibble(x=seq(0, 15, length.out = 1000))
true_theta <- list(p=c(0.5, 0.5), mu=c(5.28, 8.45), sigma=c(3, 1),skew=c(0,0))
for (i in 1:2) {
  temp_simu <- temp_simu %>% tibble::add_column("component {i}" := true_theta$p[i] *
                                                  sn::dsn(temp_simu$x, xi=true_theta$mu[i], omega = true_theta$sigma[i], alpha = true_theta$skew[i], tau=0))
}
temp_simu <- temp_simu %>% dplyr::mutate(total=rowSums(dplyr::across(dplyr::starts_with("component ")))) %>%
  tidyr::pivot_longer(cols =dplyr::starts_with(c("component ", "total")), names_to = "components", values_to = "y")
roots <- compute_overlap_roots(0.5, 0.5, 5.28, 8.45, 3, 1)
overlap_plot_pairwise_balanced_3 <- ggplot(temp_simu, aes(x = x, color = gsub("component ", "", components), y=y,linetype=gsub("component ", "", components))) +
  geom_line(size=1) +
  theme_bw() +
  theme(plot.title = element_blank(), legend.position = "bottom", axis.title=element_blank(),
        legend.key.width = unit(1.5, 'cm'), legend.text = element_text(size=14),
        plot.subtitle = element_blank(), legend.title = element_blank(), plot.tag = element_text(size=16, face = "bold")) +
  scale_linetype_manual(values = c(2, 3, 1)) +
  geom_vline(xintercept=roots, color="red", linetype="dashed", size=1.15) +
  geom_area(data = dplyr::filter(temp_simu, x>max(roots) & components=="component 1"), fill = '#FDC086',show.legend = F) +
  geom_area(data = dplyr::filter(temp_simu, x <min(roots) & components=="component 1"), fill = '#FDC086',show.legend = F) +
  geom_area(data = dplyr::filter(temp_simu, x<max(roots) & x >min(roots) & components=="component 2"), fill = '#7FC97F',show.legend = F) +
  labs(tag="C")




temp_simu <- tibble::tibble(x=seq(-5, 15, length.out = 1000))
true_theta <- list(p=c(0.9, 0.1), mu=c(5.28, 8.45), sigma=c(3, 1),skew=c(0,0))
for (i in 1:2) {
  temp_simu <- temp_simu %>% tibble::add_column("component {i}" := true_theta$p[i] *
                                                  sn::dsn(temp_simu$x, xi=true_theta$mu[i], omega = true_theta$sigma[i], alpha = true_theta$skew[i], tau=0))
}
temp_simu <- temp_simu %>% dplyr::mutate(total=rowSums(dplyr::across(dplyr::starts_with("component ")))) %>%
  tidyr::pivot_longer(cols =dplyr::starts_with(c("component ", "total")), names_to = "components", values_to = "y")
# roots <- compute_overlap_roots(0.9, 0.1, 5.28, 8.45, 3, 1)
overlap_plot_pairwise_balanced_4 <- ggplot(temp_simu, aes(x = x, color = gsub("component ", "", components), y=y,linetype=gsub("component ", "", components))) +
  geom_line(size=1) +
  theme_bw() +
  theme(plot.title = element_blank(), legend.position = "bottom", axis.title=element_blank(),
        legend.key.width = unit(1.5, 'cm'), legend.text = element_text(size=14),
        plot.subtitle = element_blank(), legend.title = element_blank(), plot.tag = element_text(size=16, face = "bold"))  +
  scale_linetype_manual(values = c(2, 3, 1)) +
  geom_area(data = dplyr::filter(temp_simu, components=="component 2"), fill = '#7FC97F',show.legend = F) +
  labs(tag="D")
```

```{r OVL-represention-pdf, include=knitr::is_latex_output(), eval=knitr::is_latex_output(), fig.cap="Illustration of the overlaps between a two-components GMM. Density function of component 1 is given by the red line, its of component 2 by the green line, and total density function $f_\\theta(X)$ is represented in blue. The total overlap is given by the sum of the green and red areas."}
gridExtra::grid.arrange(grobs=list(overlap_plot_pairwise_balanced_1, overlap_plot_pairwise_balanced_2,
                                                   overlap_plot_pairwise_balanced_3, overlap_plot_pairwise_balanced_4),
                        top="", ncol=2, nrow=2)
```

```{r OVL-represention-html, include=knitr::is_html_output(), eval=knitr::is_html_output(), fig.cap="Illustration of the overlaps between a two-components GMM. Density function of component 1 is given by the red line, its of component 2 by the green line, and total density function p<sub>\U03B8</sub>(<i>X</i>) is represented in blue. The total overlap is given by the sum of the green and red areas."}
gridExtra::grid.arrange(grobs=list(overlap_plot_pairwise_balanced_1, overlap_plot_pairwise_balanced_2,
                                                   overlap_plot_pairwise_balanced_3, overlap_plot_pairwise_balanced_4),
                        top="", ncol=2, nrow=2)
```

There are two intersection points, $x_1$ and $x_2$ , with
$\mu_1 < \mu_2$ when solving equation Equation
\@ref(eq:intersection-points):

```{=tex}
\begin{equation}
p_1 \varphi(x, \mu_1, \sigma_1)  = p_2 \varphi(x, \mu_2, \sigma_2)
(\#eq:intersection-points)
\end{equation}
```
in following case: if $\sigma_2 > \sigma_1$, then we must have
$p_1 > \frac{\sigma_1}{\sigma_1 + \sigma_2}$, else if
$\sigma_2 < \sigma_1$, then
$p_1 < \frac{\sigma_1}{\sigma_1 + \sigma_2}$. In that case, they are
given by following formula Equation \@ref(eq:roots):

```{=tex}
\begin{equation}
\begin{split}
(x_1, x_2) = \left(\frac{\sigma_1^2 \mu_2 - \sigma_2^2 \mu_1 \pm \sigma_1 \sigma_2 \sqrt{(\mu_1 - \mu_2)^2 + 2(\sigma_2^2 - \sigma_1^2)\left[ \log (\frac{p1}{p2}) + \log (\frac{\sigma_2}{\sigma_1})\right]}}{\sigma_1^2 - \sigma_2^2}\right)
\end{split}
(\#eq:roots)
\end{equation}
```
Again, sign of term $A$ and order of the roots yield two several cases,
depending whether $\sigma_1$ is greater or not than $\sigma_2$. Both
situations with unbalanced classes are illustrated in panel B and D on
Figure
`r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(fig:OVL-represention-html)', '\\@ref(fig:OVL-represention-pdf)'))`:

-   When $\sigma_1 < \sigma_2$, then $x_2 < x_1$ and
    $p_1 \varphi(x, \mu_1, \sigma_1) < p_2 \varphi(x, \mu_2, \sigma_2)$
    on interval $[x_2, x_1]$. Hence, total overlap is given by Equation
    \@ref(eq:OVL-general-case1):

```{=tex}
\begin{equation}
\begin{split}
\operatorname{OVL} (1, 2) = p_1 \left(\Phi(x_2, \mu_1, \sigma_1) +1 - \Phi(x_1, \mu_1, \sigma_1) \right) + p_2 \left( \Phi(x_1, \mu_2, \sigma_2) -\Phi(x_2, \mu_2, \sigma_2) \right)
\end{split}
(\#eq:OVL-general-case1)
\end{equation}
```
-   When $\sigma_1 > \sigma_2$, then $x_1 < x_2$ and
    $p_1 \varphi(x, \mu_1, \sigma_1) < p_2 \varphi(x, \mu_2, \sigma_2)$
    on interval $[x_1, x_2]$. Hence, total overlap is given by Equation
    \@ref(eq:OVL-general-case2):

```{=tex}
\begin{equation}
\begin{split}
\operatorname{OVL} (1, 2) = p_2 \left(\Phi(x_1, \mu_2, \sigma_2) +1 - \Phi(x_2, \mu_2, \sigma_2) \right) + p_1 \left( \Phi(x_2, \mu_1, \sigma_1) -\Phi(x_1, \mu_1, \sigma_1) \right)
\end{split}
(\#eq:OVL-general-case2)
\end{equation}
```
An interesting result is obtained with the homoscedascity and balanced
classes' assumptions of the *k*-means algorithm. There is only one
intersection point in that case: $x_c= \frac{\mu_1 + \mu_2}{2}$, that is
simply the centre of the segment bounded by the means of the two
components. The overlap is simply then
$\operatorname{OVL} (1, 2) = 2 \Phi (- \frac{|\mu_1 - \mu_2|}{2\sigma})$.



To our knowledge, no closed formula has been determined returning the overlap generalised to more than two components (combinatorial set of inequations to solve), in the unconstrained multivariate setting (cubic equation to solve in bidimensional space). Indeed, even restraining the study to the bivariate setting (the calculation of the OVL then amounts to estimating the zone of intersection between two ellipses), the exact computation of the OVL involves multiple integration and the algebraic resolution of a quartic equation. A first step is provided by [@alberich-carraminana_etal17], stating algebraic conditions for the existence of an intersection region and computing where applicable a closed formula of the OVL between two coplanar ellipses.


Accordingly, only stochastic approximations, relying on randomised algorithms, such as the Monte-Carlo integration with a rejection technique (knowing that the total area under the curve is normalised to one, we randomly simulate observations and the ratio of the number of observations falling in the intersection area is then used as a proxy of the overlap), are available so far [@maitra_melnykov10; @pastore_calcagni19;@nowakowska_etal14].


# Appendix B: Extensions of the EM algorithm to overcome its limitations {.appendix}

Two main alternatives were developed in parallel to the EM algorithm and
are implemented in some of the reviewed packages: the CEM and the SEM
algorithm. However, they do not have its theoretical properties,
especially guarantee of the consistency of the algorithm.

The M-step of the *classification EM* (CEM) algorithm
[@biernacki_etal00] maximises a function where each observation was
assigned to the maximum a posteriori (MAP) estimate Equation
\@ref(eq:posteriori). It generalises the well-known *k*-means algorithm
making no assumption of homoscedascity or equibalanced clusters. Its
main drawback is to not take into account uncertainty of the cluster
assignment, inducing *inconsistency* of the algorithm
[@mclachlan_peel00]. EM\*, referred in @kurban_etal17 and implemented
in the \CRANpkg{DCEM} package, is a faster implementation of the CEM
algorithm, with roughly a twice smaller complexity. To do so, only the
posterior distributions associated to the lower half of the most
uncertainly assigned observations are re-computed in the E-step of the
EM-algorithm. This normally avoids to recompute data that is unlikely to
change of cluster attribution from an iteration to another. However, the
higher speed of this algorithm has not been theoretically proven, as the
gain of running time per iteration of the algorithm may be alleviated by
a greater number of steps to reach the convergence.

The *Stochastic EM* (SEM) replaces the MAP value for $S$ in the E-step
of the CEM algorithm by a random draw (or $N$ of them in the *N-*
variant of the algorithm) of the posterior distribution
$\mathbb{P}_\theta (S|X)$. As this algorithm does not converge to a
unique solution, but rather oscillates around a local maximum, the
estimation is usually performed by averaging the late estimated values
while ignoring the first estimates from the *burn-in phase*. A
theoretical description of these algorithms, with discussion on their
convergence properties, is detailed in @celeux_govaert92. SEM algorithm
has also a relatively faster convergence than EM algorithm but it is
more proned to be trapped in a local maximum or to remove a component.
Increasing the number of draws $N$ may alleviate this issue, but at the
extent of computational performances.

A wide variety of fast algorithms derived from the EM algorithm have
been developed. cwEMM (component-wise EM algorithm), described in
@celeux_etal12, is a variation of the EM algorithm aiming at speeding up
its convergence. The M-step at each iteration is only performed for one
of the components $\theta_j=(p_j, \mu_j, \sigma_j)$, implying that the
parameters of a given component are estimated sequentially rather than
simultaneously. The theory behind relies on a *Gauss-Seidel* scheme and
was first used by the SAGE algorithm. However, the constraints on the
proportions set in Equation (2) are only guaranteed if the
algorithm converges. Additionally, faster convergence is not
theoretically proven for any situation. A list of general acceleration
methods for the EM algorithm, not specific to GMMs, is available on
\CRANpkg{turboEM} [@R-turboEM].


Other EM-inspired algorithms focus on counterbalancing the main
limitations of the EM algorithm. The *Variational Bayesian EM* (VBEM)
algorithm performs a Bayesian estimation of the parameters. Indeed, the
large space of all possible parameter estimates $\Theta$ can be hard to
explore and the usual initialisation methods are uninformative, not
taking into account expert recommendations. VBEM uses these prior
assumptions on the parameters' distribution $\mathbb{P} (\theta)$ to
optimise the posterior distribution $\mathbb{P} (\theta|X)$, based on
Bayes' rule. Direct determination of the Bayesian posterior law of the
parameters is generally an intractable problem, hence Variational Bayes
only maximises an approximation of the true posterior, assuming that the
parameters can be partitioned in independent distributions. This
hypothesis is known as *mean-field approximation* [@murphy12].

The minimum message length (MML) EM algorithm, implemented in the
\pkg{GMKMcharlie} package, is a completely unsupervised algorithm as it
does not require any prior selection of the number of components
[@figueiredo_jain02] , by dealing explicitly with the possibility of
discarding a component during the iteration. To do so, the selection
criteria for the number of components is directly included in the
optimisation procedure. However, its implementation is close from a
Bayesian estimation of the parameters of the model, setting a
non-informative Dirichlet prior distribution on the ratios and the
higher expected performances of the algorithm are not demonstrated on
real use cases [@figueiredo_jain02].

The Expectation/Conditional Maximisation (ECM) [@meng_rubin93] belongs to the super-family of GEM (general EM) algorithms, generally used when the maximisation of the auxiliary function yields a non-closed form to solve. To do so, the ECM algorithm replaces the intractable M-step of the EM algorithm by a number of computationally simpler conditional maximization (CM) steps (instead of inferring all parameters at once, the conditional step retrieves a set of optimal parameters, conditioned by the current value of the others).  ECM is for instance used with GMMs including an additional linear constraint on the means of the components, as provided by the \pkg{mixtools}
package. As documented in Table 2, \pkg{EMMIXmfa} implements an extension of the ECM, termed alternating expectation–conditional maximization (AECM) algorithm [@meng_vandyk97], and which can be used to reduce the computational burden of estimating the parameters of mixtures of factor analysers. The AECM algorithm is an extension of the ECM algorithm that allows the complete data used for estimation to differ on each CM-step (generally, in order to speed the computation, by selecting a subset of the most leveraged observations). GEM algorithms share the same asymptotic theoretical properties of the EM algorithm, especially the local consistency of the estimates returned.

## A small simulation to evaluate the impact of outliers {.appendix}

Classical methods used for the parameters' estimation, especially the
maximum likelihood estimation (MLE), are sensitive to the presence of
outliers. A naive solution consists in assigning null weights to
observations suspected to be outliers, so that they do not contribute
[^5]. Trimming aberrant observations from the distribution is justified
theoretically by the principle of the *spurious outlier model*
[@gallegos_ritter05]. However, this method is quite stringent,
requiring human expertise or the use of general outlier detection tools
not necessarily adapted to GMM estimation.

[^5]: The use of weighted distributions has more general applications.
    It can be used to deal with a component distribution that does not
    fit exactly a Gaussian shape. For instance, to deal with heavy tail
    distributions, more weight can be given to central components and
    less weight to the tails.

Two general approaches for dealing with outliers with a well-defined
theoretical background are the *outliers mixture modelling* and the
*trimming approach*. *Outliers mixture modelling* integrate an
additional component accounting for the outliers in the distribution.
Notably, the \pkg{mclust} [@R-mclust] and \CRANpkg{otrimle} [@R-otrimle]
packages use an improper uniform distribution to model the distribution
of outliers. Unlike \pkg{mclust}, the \pkg{otrimle} package does not
require the user to set in advance the proportion of outliers in the
mixture [@otrimle2016b]. As opposed, in the *trimming approach*,
outliers are first removed before the complete estimation of parameters.
Such methods are implemented in \CRANpkg{tclust} [@R-tclust] and
\CRANpkg{oclust} [@R-oclust] packages.

\CRANpkg{tclust} [@R-tclust] uses a robust constrained clustering
method, where the user has to set an upper threshold to the ratio
between the highest and the lowest variability among all components and
a trimming ratio $\alpha$. It extends the work of
@garcia-escudero_etal08, with released constraints on the Gaussian
distribution. First, the maximal degree of affinity, defined in Equation
\@ref(eq:affinity-degree):

```{=tex}
\begin{equation}
    D(x_i|\theta)=\max_j \left(p_{j} \varphi_{\zeta_j} (x_i) \right)
(\#eq:affinity-degree)
\end{equation}
```
is computed for each observation $x_i$, and corresponds for each point
to the maximum probability to observe it in the distribution, given
parameter $\theta$. Then, $\alpha$ observations the least likely to be
observed are trimmed for the estimation of the parameters. When we reach
convergence of the estimated parameter and there is no change in the
outliers identification from one iteration to another, the iterative
algorithm stops. The use of constraints is an additional feature that
avoids building over-dispersed or unbalanced clusters, the highest
constraint of a ratio of 1 yielding clusters with equal sizes. However,
the identification of an observation as aberrant is highly dependant on
the variability constraint and the determination of these two
hyperparameters is complex and highly dependant on the shape of the
distribution. Additionally, a CEM algorithm is used to retrieve the
parameters and the proportion of outliers, for which the MLE, in
contrast to the EM algorithm, is not asymptotically consistent nor
efficient.

Unlike \pkg{tclust}, \CRANpkg{oclust} [@R-oclust] both retrieves the
proportion of outliers and identifies them. To do so, it compares the
complete log-likelihood of the mixture $\ell(\theta|X)$ with its value
removing one observation $\ell(\theta | X \setminus X_i)$, for all
observations. Observations are iteratively removed, based on the
assumption that the Kullback-Leibler divergence between the original
log-likelihood and the trimmed log-likelihood
$\text{KL}\left(\ell(\theta|X)|| \ell(\theta | X \setminus X_i)\right)$
follows a Beta distribution. At each step, the observation that
maximises the Kullback-Leibler divergence at a statistically significant
threshold is removed. The algorithm stops trimming outliers, when this
measure is not anymore statistically significant. However, the
assumption of a Beta distribution only holds asymptotically and with
non-overlapping clusters.

To integrate the impact of outliers in the estimation, we simulated a
two-components GMM with well-separated and balanced clusters. The
outliers distribution, corresponding to the additional noise component,
was retrieved by randomly selecting *prop.outliers* points out of the
total number of observations and drew their values from an uniform
distribution bounded by an interval five times as big as the 0.05 and
0.95 quantiles of $f_\theta(X)$. All estimates were obtained comparing
the five reviewed initialisation methods, except with \pkg{otrimle}
which has its own hierarchical clustering initialisation method.


```{r outliers, fig.cap="A) Execution times for the nine reviewed packages using hierarchical clustering initialisation, with on the left $2\\%$ of outliers in proportion and on the right, $4\\%$ of outliers.  B) and C) Boxplots of the estimated parameters with $N=200$ repetitions, $n=2000$ observations and respectively $2\\%$ and $4\\%$ of outliers. The red dashed horizontal line corresponds to the true value of the parameters."}
outliers_time_data <- readRDS("./tables/univariate/univariate_outliers_time_computations.rds")
outliers_time_computations <- cowplot::plot_grid(plot_time_computations(outliers_time_data %>% filter(ID=="O1")) +
                                                   labs(tag="A") + theme(plot.tag = element_text(size=18, face = "bold")),
                                                 plot_time_computations(outliers_time_data %>% filter(ID=="O2")) +
                                                   labs(tag="B") + theme(plot.tag = element_text(size=18, face = "bold")),
                                                 ncol = 2, align = 'h', axis="tblr")

outliers_distribution_parameters <- readRDS("./tables/univariate/univariate_outliers_distribution_parameters.rds")
outliers_plot <- gridExtra::arrangeGrob(outliers_time_computations,
                                        plot_boxplots_parameters(outliers_distribution_parameters %>% filter(ID=="O1"),
                                                                 true_theta=list(p=c(0.5, 0.5), mu=c(0, 4), sigma=c(0.3, 0.3)), size_tag=7) +
                                          labs(tag="C")+ theme(plot.tag = element_text(size=18, face = "bold")),
                                        plot_boxplots_parameters(outliers_distribution_parameters %>% filter(ID=="O2"),
                                                                 true_theta=list(p=c(0.5, 0.5), mu=c(0, 4), sigma=c(0.3, 0.3)), size_tag=7) +
                                          labs(tag="D")+ theme(plot.tag = element_text(size=18, face = "bold")),
                                        top="",nrow=3, padding = unit(1, "line"))

ggsave("figs/univariate/outliers.png", outliers_plot, width = 18, height = 24,dpi = 75)
knitr::include_graphics("figs/univariate/outliers.png", dpi = 75)

```

The slowest package is \pkg{otrimle}, most of the time being taken by
the initialisation step where proportion and identification of the
outliers is performed. Running times of the other packages are generally
not impacted by the presence of outliers.

Most of the reviewed packages, except the \pkg{bgmm} package, are not
impacted by the choice of initialisation method. Additionally, the
proportions are rather correctly estimated (related to the choice of an
uniform distribution to model outliers), but the reviewed packages tend
to overestimate the true variability of each component, with the worst
results obtained with \pkg{rebmix} initialisation. \pkg{bgmm} sets apart
from the others by its reduced bias on the means and standard deviations
estimated, a feature left undocumented. However, increasing the number
of outliers (Figure \@ref(fig:outliers), panel C) lead also to biased
estimations for \pkg{bgmm}, while \pkg{otrimle}, a dedicated package, is
still able to correctly estimate the individual parameters of the
components' distributions with a high proportion of outliers. Yet,
analysing the code used to implement the \pkg{bgmm} reveals that there
is no dedicated feature to remove outliers but rather a specific method
used to deal with numerical underflow that artificially increases the
probability of observing outlying distributions ([EM-implementation
differences across reviewed packages]).

# Appendix C: the meta-analysis workflow for the final selection of CRAN and Bioconductor platforms {.appendix}

## General workflow {.appendix}

Table `r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(tab:searched-packages-html)', '\\@ref(tab:searched-packages-pdf)'))`
lists the terms used in the search, the number of packages returned by
the search, the number of packages excluded from review after the
search, and the names of the packages ultimately selected for review.
Indeed, the CRAN and Bioconductor platforms are the two most popular
repositories for R packages, with a constraining review before
publication.

Most packages we excluded from review did not focus on the GMM model,
but on supplying tools for visualising and asserting the quality of a
given clustering. For instance, the search term "cluster" returned many
packages implementing other unsupervised clustering methods, such as
*k*-means, KNN or graph clustering, were specifically dedicated to
specific data, such as single cell analyses. The search term "mixture"
returned either packages dealing only with non-Gaussian components, such
as \CRANpkg{fitmix} with log-normal distributions or were dedicated to chemical mixture designs.

```{r searched-packages-html, layout="l-body-outset", eval = knitr::is_html_output()}
readr::read_delim("./tables/review_packages.csv", delim=";", show_col_types = F) %>%
  mutate(`Packages kept`=stringr::str_replace(`Packages kept`, "varnothing", "\U2205"),
         across(.cols = where(is.character), stringr::str_replace_all, "\\\\n", ", ")) %>%
  rename_with(~ stringr::str_replace_all(.x, "\\\\n", " ")) %>%
  flextable() %>%
  set_caption(caption = "Meta-analysis summary about the selection of packages implementing the estimation of GMMs,
              on CRAN and Bioconductor.", html_escape = F) %>%
  theme_vanilla() %>% align(align = "center", part = "header") %>%   flextable::autofit()
```

```{r searched-packages-pdf, eval = knitr::is_latex_output()}
data <- readr::read_delim("./tables/review_packages.csv", delim=";", show_col_types = F) %>%
  mutate(`Packages kept`=stringr::str_replace(`Packages kept`, "varnothing", "$\\\\varnothing$"),
         across(.cols = where(is.character),kableExtra::linebreak,  linebreaker = "\\\\n"))

data %>% kbl(booktabs=T,
caption = "Meta-analysis summary about the selection of packages implementing the estimation of GMMs,
              on CRAN and Bioconductor.", escape=F, col.names = linebreak(colnames(data), linebreaker = "\\\\n")) %>%
  row_spec(0,bold=T, align = "c", hline_after = T) %>%
  row_spec(1:4, hline_after = T, align = "c") %>%
  column_spec(column = 1:6, latex_valign = "m") %>%
  kable_styling(latex_options=c("hold_position", "scale_down"))
```

At this stage, too many packages for a tractable benchmark remained. We
hence perform stricter selection of them, based on the following
criteria:

-   Some of the packages did not implement the unconstrained GMM (no
    constraint of homoscedascity or equibalanced proportions). Hence,
    \BIOpkg{epigenomix} [@R-epigenomix] , \CRANpkg{EMMIXgene}
    [@R-EMMIXgene] , \CRANpkg{pcensmix} [@R-pcensmix] , \CRANpkg{mixAK}
    [@R-mixAK] (homoscedastic components), \CRANpkg{mixture}
    [@R-mixture] (multi-dimension only), \CRANpkg{AdaptGauss}
    [@R-AdaptGauss] and \CRANpkg{MMDvariance} [@R-MMDvariance] add
    constraints on the number of components, on the standard deviation
    of each component or on mean values of each population, leaving no
    choice to the user to remove such assumptions. \BIOpkg{semisup}
    [@R-semisup] restrains on mixtures with two components, for which a
    part of the observations are labelled. Additionally, it is designed
    for GWAS or differential analyses. Other packages were designed to
    deal with high-dimensional data, projecting the data on a smaller
    subspace using a factor analysis model. Hence, these packages can
    not learn a GMM for an univariate distribution, as we can not
    project on a smaller space than the unidimensional space. This led
    to the exclusion of \pkg{HDclassif},
    \CRANpkg{fabMix} [@R-fabMix] , \pkg{EMMIXmfa} and
    \CRANpkg{pgmm} [@R-pgmm] packages. The \CRANpkg{sensory}
    [@R-sensory] package both imputes missing data and performs factor
    regression on a subspace up to 3 dimensions at most, but requires
    the user to provide its own initial estimates. Alternatively, \CRANpkg{clustvarsel} [@R-clustvarsel] discards the least informative variables, in an attempt to find a locally optimal subset of variables that best discriminate clusters.

-   We assume that our original data is continuous. However, some
    packages are dedicated to deal with discrete data, for instance
    binned size distributions of medical patients. This led to the
    exclusion of \CRANpkg{mixdist} [@R-mixdist].

-   We restrained our review to packages that use the classic EM
    algorithm, using maximum likelihood estimation to retrieve the
    parameters of GMMs. For instance, some packages offer a Bayesian
    estimation of the parameters of the model using MCMC methods, such
    as \CRANpkg{bmixture} @R-bmixture], \CRANpkg{bpgmm} [@R-bpgmm], \CRANpkg{DPP} [@R-DPP]
   , \CRANpkg{dppmix} [@R-dppmix], \CRANpkg{BayesCR} [@R-BayesCR] and
    \BIOpkg{Melissa} [@R-Melissa]. \CRANpkg{polySegratioMM} [@R-polySegratioMM] uses the Bayesian
    framework JAGS's interface in R. Alternatively, other algorithms
    focusing on maximising the likelihood do exist, but rely on
    different statistical methods, such as \CRANpkg{RPMM} [@R-RPMM]
    which implements a recursive algorithm, and \CRANpkg{SAGMM}
    [@R-SAGMM] offering a stochastic approximation.

We then removed the packages in which the MLE estimation of the
unconstrained GMM model was an ancillary task:

-   We removed the packages that focus on learning mixture of Gaussian
    regressions such as \BIOpkg{fmrs} [@R-fmrs] , \CRANpkg{mixreg}
    [@R-mixreg] or \CRANpkg{fmerPack} [@R-fmerPack] , an extension of
    the \CRANpkg{flexmix} package with an additional feature selection
    using the lasso method. \CRANpkg{nlsmsn} [@R-nlsmsn] implements
    regression of skewed Gaussian mixtures, but in unidimensional space
    only. \CRANpkg{RobMixReg} [@R-RobMixReg] performs robust regression
    of Gaussian mixtures using five several methods: CTLERob, a
    component-wise adaptive trimming likelihood estimation; mixbi,
    bi-square estimation; mixL, Laplacian distribution; mixt,
    t-distribution; TLE, trimmed likelihood estimation, and flexmix
    which only performs flexmix regressions with multiple random starts.

-   Some packages were built to deal with highly specific tasks.
    \CRANpkg{RMixtComp} [@R-RMixtComp] and \CRANpkg{clustMD}
    [@R-clustMD] deal with mixed data (continuous + discrete). The
    \CRANpkg{deepgmm} [@R-deepgmm] package learns deep Gaussian mixture
    models, generalising the classical GMM with multiple layers.
    \CRANpkg{IMIX} [@R-IMIX] focuses on clustering multi-omic data that
    is learnt with the \pkg{mclust} package, and \BIOpkg{coseq}
    [@R-coseq] implements RNA-Seq transcriptome clustering using the
    \CRANpkg{Rmixmod} package.

-   Some extend the EM algorithm on Gaussian distributions and overcome
    its main limitations. The \CRANpkg{MGMM} [@R-MGMM] package deal with
    missing data, which is not relevant in unique dimension. The
    \pkg{mixsmsn} package estimates skewed GMMs. \CRANpkg{SMNCensReg}
    [@R-SMNCensReg] fit univariate right, left or interval censored
    data. Some packages offer a robust implementation of the algorithm,
    automatically trimming possible outliers. \pkg{otrimle} models the
    presence of outliers by an extra component following an improper
    uniform distribution, while \pkg{tclust} and \pkg{oclust}
    automatically removes possible outliers before the estimation step
    ([A small simulation to evaluate the impact of outliers]).

-   We also removed packages that were limited in their functionalities
    or complex to install. Indeed, \CRANpkg{ClusterR} [@R-ClusterR]
    (*k*-means), \pkg{rebmix} (REBMIX), \CRANpkg{nor1mix} (univariate
    dimension only, wrong initialisation process), \CRANpkg{MixAll}
    [@R-MixAll] (random and small EM) do not allow to perform the EM
    algorithm with its own initial estimates. The function to provide its own initial estimates for the \pkg{DCEM] package is only internal, and not supposed to be available for the common user. \CRANpkg{pmclust} [@R-pmclust] depends on
    the availability of the OpenMPI framework for its parallelised
    implementation of the EM algorithm.

-   We also removed the \CRANpkg{MixtureInf} [@R-MixtureInf] ,
    \CRANpkg{mixR} [@R-mixR] and \CRANpkg{CAMAN} [@R-CAMAN] packages
    which have not been updated in the last two years or still in beta
    version.

The popularity of the selected packages varies largely, as shown in
Figure
`r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(fig:trend-package-html)', '\\@ref(fig:trend-package-pdf)'))`.
Among them, \pkg{mclust} and \pkg{flexmix} are largely the most popular,
followed by \pkg{mixtools} and increasingly popular \pkg{Rmixmod}
package.

```{r generate-plot-packages-trends, layout = "l-body-outset"}
downloaded_mixture_packages <- cranlogs::cran_downloads(package = c("mclust", "flexmix","mixtools", "Rmixmod",
                                                                    "EMCluster", "bgmm", "DCEM", "GMKMcharlie", "EMMIXmfa", "HDclassif"),
                                                        from = "2022-01-01", to = "2022-04-30")
nb_cols <- length(unique(downloaded_mixture_packages$package)) # store graphical parameters
my_colors <- grDevices::colorRampPalette(RColorBrewer::brewer.pal(12, "Paired"))(nb_cols)


download_trends <- ggplot(downloaded_mixture_packages, aes(x=date, y=count, shape=package, colour=package, linetype=package)) +
  geom_line(size=0.5) +
  theme_bw() +
  geom_point(size=1)+
  scale_y_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),
                labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  labs(title=element_blank(),
       x ="Month", y = "Number of daily downloads") +
  scale_shape_manual(values=seq(1,nb_cols)) +
  theme(legend.title = element_blank(), legend.position = "bottom", legend.text = element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.key.size = unit(1, "cm"), axis.text = element_text(size=12))
```

```{r trend-package-html, fig.cap="Number of daily downloads (logarithmic scale) from the CRAN mirror from the 1<sup>st</sup> of January to the 30<sup>th</sup>, April 2022 for the seven R packages reviewed.", include=knitr::is_html_output(), eval=knitr::is_html_output()}
plotly::ggplotly(download_trends)
```

```{r trend-package-pdf, fig.height = 5, fig.cap="Number of daily downloads (logarithmic scale) from the CRAN mirror from the $1^\\text{st}$ of January to the $30^{\\text{th}}$ April 2022 for the seven R packages reviewed.", include=knitr::is_latex_output(), eval=knitr::is_latex_output()}
download_trends
```



Only the packages dedicated to high-dimensionality, listed in our first bullet point, are relevant to benchmark their performance as a function of the number of dimensions. Indeed, although some packages computing mixtures of regressions do implement features allowing to to handle high-dimensional datasets, such as \pkg{RobMixReg} and \pkg{fmerPack}, they all assume a diagonal covariance structure, and accordingly independent covariates.

The two existing strategies are then limited to projection to a smaller subspace, usually within the theoretical framework of factor analysis or to perform a feature selection strategy.
We quickly discarded \pkg{fabMix}, since it only retrieves the parameters of GMMs within a Bayesian framework, while we focused on strategies retrieving the MLE via the EM algorithm. The core function `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("pgmm::pgmmEM"),   '\\code{pgmmEM}'))` in the \pkg{pgmm} package unfortunately includes a seed for the the algorithm's initialisation that cannot be disabled. Such a feature is generally not recommended for reproducibility, since by defining the seed internally in the function, we were not able to independently generate  new reproducible datasets in our benchmark (instead, it is recommended to set the seed value once and for all at the beginning of the virtual simulation). Additionally, while implementing the same AECM variant of the EM algorithm as \pkg{EMMIXmfa}, as detailed in [Appendix B: Extensions of the EM algorithm to overcome its limitations], its convergence criteria differs from the other benchmarked packages. Indeed, instead of considering a limiting number of iterations along with a prior threshold, either *absolute* or *relative*, it examines only the difference between the current value of the log-likelihood and a corresponding asymptotic estimate, based on the Aitken acceleration [@lindsay95]. In brief, the asymptotic value of the log-likelihood is the limiting sum of a geometric series, whose common ratio, the so-called Aitken acceleration, is the relative fraction of the log-likelihood gain of the current iteration.  Therefore, the use of a different termination criterion precludes any further fair comparison with the other benchmarked packages, as there is no direct equivalence between the two methods.

Finally, \pkg{clustvarsel} is not really tailored for datasets with a large number of dimensions, but rather for datasets with a small number of observations. Indeed, by performing a sequential search in the model space in a  forward-backward process (namely by adding variables to the null model till we recover the full model, with all features), the algorithm requires intensive computational resources (for instance, there are already $2^{10}=1024$ models to be tested in dimension 10). Instead of a greedy strategy, it would have been interesting to implement a in independent and highly parallelizable feature selection process by sampling the model space. To do so, [@chen_chen08; @he_chen16] suggests a stochastic and greedy feature selection strategy, using notably the *eBIC* criteria in order to have an equal chance to draw a model of any dimension^[Indeed, by simply uniformly sampling among the $2^D$ models available, the probability of getting models with $D/2$ features is much higher than drawing models at the boundaries, displaying either few or close to $|D|$ covariates.]. Such a strategy is commonly used in *ensemble learning*.

<!-- ^[Note that with only some workarounds, moving from a regression framework to a more standard clustering task is straightforward, replacing the design matrix storing covariates by a constant term.]. -->






## Practical details for the implementation of our benchmark {.appendix}
\label{subsec:benchmark-details}

First, the number of observations ($n=200$ and $n=500$ respectively in the univariate and bivariate setting) was chosen enough high to both lower the probability of generating a sample without drawing any observation from one of the components in case of highly-unbalanced clusters and decreases the *margin of error* related to the random sampling error. Specifically, the probability of generating at least one simulation among the $N$ generated fo which less than two observations proceed from component $j$ (the minimal number of elements required to estimate both the mean and the variance of the corresponding cluster), with a two-components mixture of $n$ observations, is given by the following formula (Equation \@ref(eq:random-sampling-risk)):

\begin{equation}
1 - \left(1 - (1-p_j)^n - n \times (1-p_j)^{n-1}\times p_j\right)^N
(\#eq:random-sampling-risk)
\end{equation}

Interestingly, the probability of generating a sample among the $N$ repetitions increases exponentially as the level of imbalance increases. For instance, considering $N=100$ repetitions, $n=200$ observations per sample and proportion for the minor component $p_j=0.1$, the probability of generating a degenerate simulation is insignificant: $1.63 \times 10^{-6}$ while the risk considerably increases, keeping the same general simulation parameters and setting minor proportion to $p_j=0.05$, with a probability of $0.04$. We have focused on one of the impacts of high dimensionality, namely that related to the homogenisation and convergence of any distance norm and the increase in sparsity in relation with the number of features added. We deliberately do not consider the case where the number of dimensions exceeds the number of observations (namely, when $D>n$), since in this configuration, the covariance matrix is no longer of full rank and invertible, implying that the corresponding probability distribution does spans completely over a smaller subspace. However, with so few observations, ($n=200$ in scenarios identified as a), we reveal the impact in terms of the quality of the estimation when the number of observations is closed to the number of free parameters required to parametrise the full GMM model (with $k=2$ clusters and $D=10$ dimensions, $k \times \frac{D(D+1)}{2} + kD + 1 = 131$ are needed.).


Unless stated explicitly, we keep the default hyper-parameters and custom global options provided by each package. For instance, the \pkg{flexmix} package has a default option, *minprior*, set by default to 0.05 which removes any component present in the mixture with a ratio below $0.05$. Besides, we only implement the fully unconstrained model in both univariate and multivariate settings, as it is the only parametrisation implemented in all the seven packages and the most popular to perform classic GMM clustering, as no restrictive and difficult-to-test assumptions are required.

Additionally, as stated in [Parameters estimation in a high-dimensional context], the intrinsic dimension $d_j$ for each cluster $j$ is a hyperparameter, which is generally inferred independently from the GMM estimation itself. While a variety of methods from the field of factor analysis, enumerated in [Factor criteria selection](https://en.wikipedia.org/wiki/Factor_analysis#Criteria_for_determining_the_number_of_factors), have been developed to estimate the intrinsic dimension, to our knowledge, only two of them have been implemented in CRAN packages: the *Cattell’s scree-test* [@cattell66] or the dimension selection graph using one of the *penalty metric* discussed in the appendix [Model selection] [@berge_etal12]. However, while \pkg{HDclassif} natively implements a performance criterion method for determining the dimension of the spanning space, performed under the hood by function `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("HDclassif::hdcc()"),   '\\code{mixsmsn::hdcc}'))`, none of the other packages evaluated implemented a dimension selection feature. Instead, we infer it for each of the packages dedicated to high-dimensionality with HDclassif, using using the so-called model "AkjBkQkD", for which the intrinsic dimension is common to all components but the characteristics unique for each component Finally, we use among all supplied parametrisations, the least constrained one. Namely, we used the model "AkjBkQkDk" with HDclassif, in which not only the individual features of the covariance matrix but also the spanning dimension are unique for each cluster, and function `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("EMMIXmfa::mcfa()"),   '\\code{mcfa}'))` of the \pkg{ EMMIXmfa} package, in which the *transition matrix* is common to all components (referred to as the orientation matrix in Appendix [Parameters estimation in a high-dimensional context].



If all the seven reviewed packages accept initial estimates provided by the user, both the input and the output format differ between them, requiring an intensive processing to standardise both the initial estimates input, and the output estimates. Notably, a well-known issue with the mixture models is that they identifiable up to a permutation of the components (alternatively, changing the index of the labels do not change the likelihood of the model). Assigning one component of the mixture to a specific index is generally immaterial, as the main objective is to return the estimates. However, when it comes to compare the estimated parameters with the true estimates, we must associate unequivocally each component to a  specific index. To do so, we set a partial ordering, sorting the components by increasing order of their mean components. Actually, if the ratio or the covariance estimates can be equal for all the components, it is generally not the case for the centroids, as this would result into a degenerate distribution. The consequence and some illustrations of the non-identifiability of the mixture distributions are discussed in section [Identifiability of finite mixture models](https://stats.stackexchange.com/questions/265898/why-is-a-normal-mixture-model-not-identifiable-and-why-does-it-matter), in @dai_mukherjea01 and in Book @robert_casella10.

We detail below some additional functions we implement to both homogenise input and output of the packages and ease the user's task when comparing the performance of these packages:

+ The input observations, mean and covariance matrices have to be transposed compared to the conventional format in packages \pkg{bgmm}, \pkg{EMCluster}, \pkg{GMKMcharlie} and \pkg{Rmixmod}, namely $D \times k$ mean matrix and $D^2 \times k$ covariance array ($D^2$ matrix to store each component variance).

+ To save some storage, the \pkg{EMCluster} package reshapes the covariance matrix, benefiting from its symmetry. Hence, instead of a three-dimensional array, \pkg{EMCluster} expects a compressed $k \times \frac{D(D+1)}{2}$ matrix, each line storing the upper triangular part of the covariance. The memory gain is yet controversial, as decreasing only by a factor two the total space required for the computation. To switch from one format to another, we developed specifically two functions: *trig_mat_to_array()* and *array_to_trig_mat()* in our GitHub package *RGMMBench*, partly inspiring from *vec2sym* function [Handy R functions](https://rdrr.io/github/patr1ckm/patr1ckm/man/vec2sym.html).

+ Instead of the covariance matrix, the \pkg{mclust} package requires the lower triangular matrix resulting from its Cholesky decomposition. One of the main advantages of this input, in addition to save storage space, is that it ensures that the covariance matrix is indeed positive-definite, as the Cholesky factorisation is only defined if this condition is respected [Cholesky decomposition](https://en.wikipedia.org/wiki/Cholesky_decomposition).

+ \pkg{flexmix} starts by the M-step of the EM algorithm instead of the E-step. Hence, it expects the posterior probabilities assigned to each cluster $j$ for each observation $i$, $\eta_i(j)$ (Equation \@ref(eq:posteriori)), instead of the initial estimates. Both approaches are, however, equivalent.


On the contrary, none of the packages we evaluated that were dedicated to high-dimensional datasets allow the user to provide its own estimates. Thus, when any of the benchmarked initialisation methods listed in Table `r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(tab:general-parameter-description-html)', '\\ref{tab:general-parameter-description-pdf}'))`, was internally available in the package, we use it with the same hyperparameters described in main paper, section *Initialisation of the EM algorithm*. If not, we provide instead a vector containing the MAP assignments inferred by the native initialisation method, in a process similar to that used used with hierarchical clustering.



In addition to the plots displaying the bootstrap parameter estimations associated to Scenarios in Tables  \@ref(tab:parameter-configuration-univariate), \@ref(tab:parameter-configuration-bivariate) and \@ref(tab:parameter-configuration-HD), we have computed summary statistics to compare the performances of the reviewed packages:

* The *bias* measures the deviation between the sample mean value of the
estimate and the true parameter:
$\operatorname{Bias}(\hat{\theta})=\mathbb{E} [\hat{\theta}] - \theta$.

* The Mean Squared Error (MSE) summarises both the variability of the
estimator and its bias: $\operatorname{MSE} (\hat{\theta})=\mathbb{E} \left[ (\hat{\theta} - \theta)^2 \right] = \operatorname{var}(\hat{\theta}) + \operatorname{Bias}(\hat{\theta})^2$, where $\operatorname{var}(\hat{\theta})$ is the empiric variance of each
estimator given by the diagonal terms of the empiric covariance matrix.

* We enumerate the number of successes (either the package or the initial method returns an error, or fails in returning a set of parameters enforcing standard constraints of multivariate GMMs, namely the unit simplex constraint over the ratios, positive-definite covariance matrices and in general no missing or infinite value).

* For each scenario, we measured independently the running times taken by the initialisation step and by the estimation of the parameters by the EM
algorithm. To do so, the \CRANpkg{microbenchmark} package
[@R-microbenchmark] was used for its higher accuracy and flexibility for
the computation of the running times in place of `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("System.time"),   '\\code{System.time}'))`.


The main differences across packages as well as performance results obtained across packages in each univariate, bivariate and high-dimensional simulation scenario are thoroughly described in the next section.

# Appendix D: comprehensive report from the univariate and multivariate benchmark {.appendix}

## EM-implementation differences across reviewed packages
\label{sec:em-differences}

Most of the distinct behaviours between the packages result from additional choices external to the EM algorithm itself, aiming at partly overcoming its main limitations (Panel B, Figure \@ref(fig:heatmap-all-correlation-plots-univariate)). We detail below their differences ranked by decreasing order of their leverage effect on the final estimate:

1. Most of the differences between the two classes of packages (Figure (2)) are
related to the either relative or absolute choice for the termination criterion of the EM algorithm. Given an user-defined threshold, the *absolute method* early stops the estimation by comparing the difference between two consecutive log-likelihoods, $|\ell(\hat{\theta}_{q}|X) - \ell(\hat{\theta}_{q-1}|X)|$, while the *relative method* examines the variation rate, $\left\lvert\frac{\ell(\hat{\theta}_{q}|X) - \ell(\hat{\theta}_{q-1}|X)}{\ell(\hat{\theta}_{q-1}|X)}\right\lvert$.
2. Several methods can be used to deal with numerical underflow, mostly happening with highly unlikely observations, distant from any centroid.
    +  The least elaborate feature is from \pkg{Rmixmod}, returning an error when either any of the posterior probabilities or any of the estimated parameters goes below to the precision threshold of the machine (`r .Machine[["double.eps"]] %>% round(digits = 18)` for most OS).

    + If the maximal value of any posterior probability is null, \pkg{bgmm} subtracts the minimal logarithm posterior probability to any log-computed probability. This method avoids numerical
underflow by preventing computation of null ratios but the correctness of the estimates is no longer enforced^[Additionally, \pkg{bgmm} does
not update the estimated variances if any newly computed variance is
below the criterion stop. A remarkable side-effect of these features, as
shown in Figure \@ref(fig:outliers), is that the \pkg{bgmm} package is
less sensitive to the presence of outliers.].

    + The remaining packages handled numeric underflow in a more convincing
manner as they guarantee to return the MLE estimate. The \pkg{flexmix}, \pkg{GMKMcharlie} and \pkg{EMCluster}
packages use the same log-rescaling tip detailed in ([Application of the EM algorithm to GMMs]). The \pkg{mixtools} and
\pkg{mclust} packages use a variant of this trick, taking profit of the factorisation
by the greatest element (Equation \@ref(eq:trick-underflow), Equation 3 p.5 @mixtools2009), but without exploiting the tip of Taylor's development over $\log(1+x)$:

```{=tex}
\begin{equation}
\eta_{i} (j) = \frac{p_{j} \, \varphi_{\zeta_j} (x)}{\sum_{j=1}^k p_{j} \, \varphi_{\zeta_j} (x)}=\frac{\frac{p_{j} \, \varphi_{\zeta_j} (x)}{p_{j_{\min}} \, \varphi_{\zeta_{j_{\min}}} (x)}}{1+ \frac{\sum_{j\neq j_{\min}} p_{j} \, \varphi_{\zeta_j} (x)}{p_{j_{\min}} \, \varphi_{\zeta_{j_{\min}}} (x)}}
    (\#eq:trick-underflow)
\end{equation}
```

In both cases, the computation of the smallest posterior probability, the most proned to be assigned a null value, is avoided, avoiding inconsistent ratios of type $0/0$.

  + The previous two items deal with specific numeric limitations, but do not directly address one  of the main theoretical limitation of the EM algorithm, namely the risk of falling into a suboptimal maximum, plateau or getting trapped on the boundary space (occurs when the proportion of one of the component converges to zero). Some packages specifically handle the case of a vanishing component during the EM optimization: the \pkg{mixtools} package performs random re-initialisations in case one of the computed variance goes below a user-defined threshold (default $10^{-8}$). \pkg{flexmix} and \pkg{GMKMcharlie} deal explicitly with the removal of a component, by updating the corresponding MLE parameters. \pkg{flexmix} removes any component whose associated weight is by default below $0.05$ (such a stringent limitation tends to an underestimation of the true number of components in highly unbalanced mixtures)^[Indeed, at least one of the component was removed in $80\%$ of our estimations in the unbalanced and overlapping case (scenario U9 in \@ref(tab:parameter-configuration-univariate)) and in $20\%$ of the simulations in the unbalanced and well-separated case (scenario U3 in \@ref(tab:parameter-configuration-univariate)).], while GMKMcharlie both implements a lower limit on the proportions of the components and an upper threshold over the ratio of the maximum and minimal eigenvalue resulting from the factorisation of the covariance matrix (Equation \@ref(eq:covariance-decomposition))^[These options are set respectively to 0 and $+\infty$ by default, thus they did not impact our simulations].


We enumerate below some additional features supplied by the packages:

- In addition to log rescaling, \pkg{GMKMcharlie} includes an additional argument, *embedNoise*, to avoid degenerate GMMs by adding a small constant to any diagonal term (by default $10^{-6}$). Besides, instead of controlling whether there was a relative change of the log-likelihood, the EM implementation of \pkg{GMKMcharlie} controls instead that there was no significant relative difference in the estimated parameters in the ten previous optimisations^[In our simulation, the behaviour of the \pkg{GMKMcharlie} did not differ significantly from the remaining packages of the second class. However, the use of an Euclidean distance criterion may be problematic when parameters are not on the same order of magnitude, requiring their prior normalisation]. Finally, since \pkg{GMKMcharlie} has implemented a parallelised version of the algorithm, it ensures using a a time limit that the algorithm indeed terminates (by default, set to one hour).

- \pkg{flexmix} performs an unbiased estimate of the covariance matrix, instead of the corresponding ML covariance estimate (divides by a factor $n-1$ instead of the number of observations $n$). Such a choice does not affect the results in our simulations, but may have a stronger impact when fitting models to a small number of observations.


- Similarly to flexmix, the \pkg{HDclassif} package implements some constraints to preserve numerical stability. The `min.individuals` attribute, like the `minprior` attribute of `flexmix` function, discards any cluster having fewer observations^[by default, set to two, i.e. the minimum number of replications to derive an unbiased estimate of the empirical variance of a sample]. However, unlike \pkg{flexmix}, the algorithm stops instead of reparametrising the mixture problem with a smaller number of components. Coupled with the *Cattell's scree-test*, the `noise.ctrl` attribute is the minimum threshold of a feature's contribution to the overall variance, computed as the corresponding normalised eigenvalue, in order to be included in the mixture of factor analysers. This additional constraint ensures a parsimonious dimension selection process, so that the number of selected intrinsic dimensions cannot be greater than or equal to the order of the discarded eigenvalues.



\newpage

## Supplementary Figures and Tables in the univariate simulation {.appendix}

Table below (\@ref(tab:parameter-configuration-univariate)) lists the complete set of parameters used to simulate the univariate Gaussian mixture distribution in our benchmark:


```{r parameter-configuration-univariate, layout = "l-page"}
univariate_configuration  %>%   mutate(Proportions=purrr::map_chr(true_parameters, ~paste(.x$p, collapse = " / ")),
                                       Means=purrr::map_chr(true_parameters, ~paste(.x$mu, collapse = " / ")),
                                       Correlations=purrr::map_chr(true_parameters, ~paste(.x$sigma, collapse = " / "))) %>%
  select(-c("OVL", "prop_outliers", "true_parameters", "formatted_true_parameters", "old_OVL")) %>%
  rename(Entropy=entropy, OVL=OVL_pairwise) %>%
  kbl(booktabs=T, caption = "The 9 parameter configurations tested to generate the samples of the univariate experiment, with $k=4$ components.",
      escape=F, align = "c") %>%
  kable_styling(latex_options=c("hold_position", "scale_down")) %>%
  row_spec(0,bold=T) %>% row_spec(1:9, hline_after = T) %>%
  kable_styling(bootstrap_options = c("hover"))
```

```{r prepare-legend-univariate}
univariate_html_ouput <- knitr::is_html_output()  # change to TRUE for larger figures

html_caption_balanced_well_separated_univariate <- "Benchmark summary plots of scenario U1 in Table \\@ref(tab:parameter-configuration-univariate) (balanced and well-separated components), organised as such:

- The panel A displays the distribution of the global mixture distribution $f_{\\theta}(X)$
(pink solid line) and of each of its constitutive components scaled by
their respective proportions (dotted lines).

- Running times are displayed in Panel B with the *k*-means initialisation. The number of observations
(x-axis) and the running time (y-axis) is in $\\log(10)$ scale, implying
that any linear relationship between the running time and the number of
observations is represented by a slope of 1. The points represent median
running time. The coloured bands represent the 5th and 95th percentiles
of the running time.

- In panel C are represented the boxplots associated with the distribution of the estimates, with one box per pair of package and initialisation method. The median is displayed with bold
black line, the mean with a yellow cross and the 0.25 and 0.95 quantiles
match the edges of the rectangular band. Solid black lines extending
past the box boundaries represent the $1.5$ IQR, estimates above these
limits considered as outliers and omitted from the plot. Finally, the
true value of the parameter is represented as a dashed red line. The bold black writing in the upper right-hand corner refers to the parameter whose distribution is shown in the corresponding facet. The first, second and third rows are the distributions of the ratios, means and variances of each component, identified by the column index."

pdf_caption_balanced_well_separated_univariate <- "Benchmark summary plots of scenario U1 in Table \\ref{tab:parameter-configuration-univariate} (balanced and well-separated components), organised as such:
The panel A displays the distribution of the global mixture distribution $f_{\\theta}(X)$
(pink solid line) and of each of its constitutive components scaled by
their respective proportions (dotted lines).
Running times are displayed in Panel B with the \\textit{k}-means initialisation. The number of observations
(x-axis) and the running time (y-axis) is in $\\log(10)$ scale, implying
that any linear relationship between the running time and the number of
observations is represented by a slope of 1. The points represent median
running time. The coloured bands represent the 5th and 95th percentiles
of the running time.
In panel C are represented the boxplots associated with the distribution of the estimates, with one box per pair of package and initialisation method. The median is displayed with bold
black line, the mean with a yellow cross and the 0.25 and 0.95 quantiles
match the edges of the rectangular band. Solid black lines extending
past the box boundaries represent the $1.5$ IQR, estimates above these
limits considered as outliers and omitted from the plot. Finally, the
true value of the parameter is represented as a dashed red line. The bold black writing in the upper right-hand corner refers to the parameter whose distribution is shown in the corresponding facet. The first, second and third rows are the distributions of the ratios, means and variances of each component, identified by the column index."


html_caption_unbalanced_well_separated_univariate <- "Benchmark summary plots of scenario U7 in Table \\@ref(tab:parameter-configuration-univariate) (unbalanced and well-separated components), with same layout as in Figure \\@ref(fig:four-component-balanced-separated)."
pdf_caption_unbalanced_well_separated_univariate <- "Benchmark summary plots of scenario U7 in Table \\ref{tab:parameter-configuration-univariate} (unbalanced and well-separated components), with same layout as in Figure \\ref{fig:four-component-balanced-separated}."

html_caption_balanced_overlapping_univariate <- "Benchmark summary plots of scenario U3 in Table \\@ref(tab:parameter-configuration-univariate) (balanced and overlapping components), with same layout as in Figure \\@ref(fig:four-component-balanced-separated)."
pdf_caption_balanced_overlapping_univariate <- "Benchmark summary plots of scenario U3 in Table \\ref{tab:parameter-configuration-univariate} (balanced and overlapping components), with same layout as in Figure \\ref{fig:four-component-balanced-separated}."

html_caption_unbalanced_overlapping_univariate <- "Benchmark summary plots of scenario U9 in Table \\@ref(tab:parameter-configuration-univariate) (unbalanced and overlapping components), with same layout as in Figure \\@ref(fig:four-component-balanced-separated)."
pdf_caption_unbalanced_overlapping_univariate <- "Benchmark summary plots of scenario U9 in Table \\ref{tab:parameter-configuration-univariate} (unbalanced and overlapping components), with same layout as in Figure \\ref{fig:four-component-balanced-separated}."

html_caption_intermediate_scenario_univariate <- "Benchmark summary plots of scenarios U4 and U6 in Table \\@ref(tab:parameter-configuration-univariate) (small unbalance, with additional overlap in scenario U6). Panel A and B display the univariate GMM distributions of respectively scenarios U4 and U6, and Panel C and D the benchmarked distributions of respectively scenarios U4 and U6, built as Panel C of Figure \\@ref(fig:four-component-balanced-separated)."
pdf_caption_intermediate_scenario_univariate <- "Benchmark summary plots of scenarios U4 and U6 in Table \\ref{tab:parameter-configuration-univariate} (small unbalance, with additional overlap in scenario U6). Panel A and B display the univariate GMM distributions of respectively scenarios U4 and U6, and Panel C and D the benchmarked distributions of respectively scenarios U4 and U6, built as Panel C of Figure \\ref{fig:four-component-balanced-separated}."

html_global_heatmap_univariate <- "Correlation heatmaps of the estimated parameters extended to the four initialisation methods benchmarked, using the same configuration described in Figure (2), in the bivariate setting."
pdf_global_heatmap_univariate <- "Correlation heatmaps of the estimated parameters extended to the four initialisation methods benchmarked, using the same configuration described in Figure (2), in the bivariate setting."

html_univariate_init_time <- "Distribution of the running times taken by each initialisation algorithm enumerated in Table \\@ref(tab:general-parameter-description-html), across all scenarios listed in Table \\@ref(tab:parameter-configuration-univariate), sorted by increasing ID number in the lexicographical order."
pdf_univariate_init_time <- "Distribution of the running times taken by each initialisation algorithm enumerated in Table \\ref{tab:general-parameter-description-pdf}, across all scenarios listed in Table \\ref{tab:parameter-configuration-univariate}, sorted by increasing ID number in the lexicographical order."
```



Figure \@ref(fig:four-component-balanced-separated)-Figure
\@ref(fig:four-components-midbalanced) each summarise the benchmarking
results associated with one of the scenarios listed in Table
\@ref(tab:parameter-configuration-univariate).

Summary tables `r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(tab:balanced-well-separated-table-univariate-html)', '\\@ref(tab:balanced-well-separated-table-univariate-pdf)'))`-
`r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(tab:unbalanced-overlapping-table-univariate-html)', '\\@ref(tab:unbalanced-overlapping-table-univariate-pdf)'))`
display the average performance for each package of the benchmark with
each initialisation method. The best performing pair (lowest bias or
MSE) is highlighted in green, and the worst performing in red. The MSE and bias columns were derived by summing respectively the estimated proportions, means and standard deviations associated with the
individual components.


```{r compute-balanced-well-separated-table-univariate-table}
univariate_parameters_local_scores_balanced_separated <- univariate_parameters_local_scores_summary %>%
  filter(ID=="U1") %>% select(-ID)
nrow_summary <- nrow(univariate_parameters_local_scores_balanced_separated)
```

```{r balanced-well-separated-table-univariate-pdf, eval=knitr::is_latex_output()}
univariate_parameters_local_scores_balanced_separated_kable <- univariate_parameters_local_scores_balanced_separated %>%
  kbl(booktabs=T, caption = "MSE and Bias associated to scenario U1, in Table \\ref{tab:parameter-configuration-univariate} (balanced and well-separated components)", escape=F,
      col.names = linebreak(c("Package", "Initialisation\nMethod",
                              "Global \n MSE $p$", "Global\nMSE $\\mu$", "Global\nMSE $\\sigma$",
                              "Global \n Bias $p$", "Global\nBias $\\mu$", "Global\nBias $\\sigma$"))) %>%
  kable_styling(latex_options=c("hold_position", "scale_down"), full_width = T) %>%
  row_spec(0,bold=T, align = "c") %>%
  column_spec(column = 1:8, latex_valign = "m") %>%
  column_spec(1, bold = T) %>% collapse_rows(columns = 1, valign = "middle")

for (metric_col in metric_colnames) {
  # define col vector
  col_vector <- rep("black", nrow_summary); metric_vector <- univariate_parameters_local_scores_balanced_separated %>% pull(metric_col)
  col_vector[which(metric_vector==min(metric_vector))] <- "green" # set minimal values to green
  col_vector[which(metric_vector==max(metric_vector))] <- "red" # set maximal values to red
  # set colours
  univariate_parameters_local_scores_balanced_separated_kable <- univariate_parameters_local_scores_balanced_separated_kable %>%
    column_spec(match(metric_col, names(univariate_parameters_local_scores_balanced_separated)),color=col_vector)
}
univariate_parameters_local_scores_balanced_separated_kable
```

```{r balanced-well-separated-table-univariate-html, eval=knitr::is_html_output(), layout = "l-body-outset"}
# latex + collapse rows + determine min and max

univariate_parameters_local_scores_balanced_separated_flex <- univariate_parameters_local_scores_balanced_separated %>% flextable() %>%
  set_caption(caption = "MSE and Bias associated to scenario U1, in Table \\@ref(tab:parameter-configuration-univariate) (balanced and well-separated components)", html_escape = F)

for (metric_col in metric_colnames) {
  univariate_parameters_local_scores_balanced_separated_flex <- univariate_parameters_local_scores_balanced_separated_flex %>%
    flextable::color(j=metric_col, part = "body",
                     i = as.formula(paste0("~ `", metric_col, "`==min(`", metric_col, "`)")),color = "green")

  univariate_parameters_local_scores_balanced_separated_flex <- univariate_parameters_local_scores_balanced_separated_flex %>%
    flextable::color(j=metric_col, part = "body",
                     i = as.formula(paste0("~ `", metric_col, "`==max(`", metric_col, "`)")),color = "red")
}

univariate_parameters_local_scores_balanced_separated_flex %>%
  flextable::compose(j="global_mse_p", part = "header", value = as_paragraph("Global MSE \U0070")) %>%
  flextable::compose(j="global_mse_mu", part = "header", value = as_paragraph("Global MSE \U03BC")) %>%
  # flextable::compose(j="global_mse_ sigma", part = "header", value = as_paragraph("Global_mse_ ", as_equation("\\sigma"))) %>%
  flextable::compose(j="global_mse_sigma", part = "header", value = as_paragraph("Global MSE \U03C3")) %>%
  flextable::compose(j="global_bias_p", part = "header", value = as_paragraph("Global Bias \U0070")) %>%
  flextable::compose(j="global_bias_mu", part = "header", value = as_paragraph("Global Bias \U03BC")) %>%
  flextable::compose(j="global_bias_sigma", part = "header", value = as_paragraph("Global Bias \U03C3")) %>%
  set_header_labels(package= "Package", `initialisation method`="Initialisation Method") %>%
  theme_vanilla() %>% align(align = "center", part = "header") %>%  # centre align header
  merge_v(j = "Package") %>%   flextable::autofit()
```

\newpage

```{r four-component-balanced-separated, fig.cap=if (univariate_html_ouput) html_caption_balanced_well_separated_univariate else pdf_caption_balanced_well_separated_univariate}

true_theta <- univariate_configuration %>% filter(ID=="U1") %>% pull(true_parameters) %>% magrittr::extract2(1)
upper_part <- cowplot::plot_grid(plot_univariate_normal_density_distribution(true_theta) + labs(tag="A"),
                                 plot_time_computations(univariate_time_computation %>% filter(initialisation_method=="kmeans" & ID == "U1")) +
                                   theme(plot.tag = element_text(size=18, face = "bold")) + labs(tag = "B"), ncol = 2, align = 'h', axis="tblr")

univariate_balanced_separated <- gridExtra::arrangeGrob(upper_part,
                                                             plot_boxplots_parameters(formatted_univariate_distribution_parameters %>%
                                                                                        filter(ID == "U1"), true_theta) +
                                                               theme(plot.tag = element_text(size=18, face = "bold"),
                                                                     axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4)) +
                                                               labs(tag = "C"), top="", nrow = 2, heights = c(1.2, 2), padding = unit(1, "line"))

ggsave("./figs/univariate/univariate_balanced_separated.png", univariate_balanced_separated,
       width = 24, height = 18,dpi = 75)
knitr::include_graphics("./figs/univariate/univariate_balanced_separated.png", dpi = 75)
```


\newpage




```{r compute-unbalanced-well-separated-table-univariate-table}
univariate_parameters_local_scores_unbalanced_separated <- univariate_parameters_local_scores_summary %>%
  filter(ID=="U7") %>% select(-ID)
nrow_summary <- nrow(univariate_parameters_local_scores_unbalanced_separated)
```

```{r unbalanced-well-separated-table-univariate-pdf, eval=knitr::is_latex_output()}
univariate_parameters_local_scores_unbalanced_separated_kable <- univariate_parameters_local_scores_unbalanced_separated %>%
  kbl(booktabs=T, caption = "MSE and Bias associated to scenario U7, in Table \\ref{tab:parameter-configuration-univariate} (unbalanced and well-separated components)", escape=F,
      col.names = linebreak(c("Package", "Initialisation\nMethod",
                              "Global \n MSE $p$", "Global\nMSE $\\mu$", "Global\nMSE $\\sigma$",
                              "Global \n Bias $p$", "Global\nBias $\\mu$", "Global\nBias $\\sigma$"))) %>%
  kable_styling(latex_options=c("hold_position", "scale_down"), full_width = T) %>%
  row_spec(0,bold=T, align = "c") %>%
  column_spec(column = 1:8, latex_valign = "m") %>%
  column_spec(1, bold = T) %>% collapse_rows(columns = 1, valign = "middle")

for (metric_col in metric_colnames) {
  # define col vector
  col_vector <- rep("black", nrow_summary); metric_vector <- univariate_parameters_local_scores_unbalanced_separated %>% pull(metric_col)
  col_vector[which(metric_vector==min(metric_vector))] <- "green" # set minimal values to green
  col_vector[which(metric_vector==max(metric_vector))] <- "red" # set maximal values to red
  # set colours
  univariate_parameters_local_scores_unbalanced_separated_kable <- univariate_parameters_local_scores_unbalanced_separated_kable %>%
    column_spec(match(metric_col, names(univariate_parameters_local_scores_unbalanced_separated)),color=col_vector)
}
univariate_parameters_local_scores_unbalanced_separated_kable
```

```{r unbalanced-well-separated-table-univariate-html, eval=knitr::is_html_output(), layout = "l-body-outset"}
univariate_parameters_local_scores_unbalanced_separated_flex <- univariate_parameters_local_scores_unbalanced_separated %>% flextable() %>%
  set_caption(caption = "MSE and Bias associated to scenario U7, in Table \\@ref(tab:parameter-configuration-univariate) (unbalanced and well-separated components)", html_escape = F)

for (metric_col in metric_colnames) {
  univariate_parameters_local_scores_unbalanced_separated_flex <- univariate_parameters_local_scores_unbalanced_separated_flex %>%
    flextable::color(j=metric_col, part = "body",
                     i = as.formula(paste0("~ `", metric_col, "`==min(`", metric_col, "`)")),color = "green")

  univariate_parameters_local_scores_unbalanced_separated_flex <- univariate_parameters_local_scores_unbalanced_separated_flex %>%
    flextable::color(j=metric_col, part = "body",
                     i = as.formula(paste0("~ `", metric_col, "`==max(`", metric_col, "`)")),color = "red")
}

univariate_parameters_local_scores_unbalanced_separated_flex %>%
  flextable::compose(j="global_mse_p", part = "header", value = as_paragraph("Global MSE \U0070")) %>%
  flextable::compose(j="global_mse_mu", part = "header", value = as_paragraph("Global MSE \U03BC")) %>%
  # flextable::compose(j="global_mse_ sigma", part = "header", value = as_paragraph("Global_mse_ ", as_equation("\\sigma"))) %>%
  flextable::compose(j="global_mse_sigma", part = "header", value = as_paragraph("Global MSE \U03C3")) %>%
  flextable::compose(j="global_bias_p", part = "header", value = as_paragraph("Global Bias \U0070")) %>%
  flextable::compose(j="global_bias_mu", part = "header", value = as_paragraph("Global Bias \U03BC")) %>%
  flextable::compose(j="global_bias_sigma", part = "header", value = as_paragraph("Global Bias \U03C3")) %>%
  set_header_labels(package= "Package", `initialisation method`="Initialisation Method") %>%
  theme_vanilla() %>% align(align = "center", part = "header") %>%  # centre align header
  merge_v(j = "Package") %>%   flextable::autofit()
```

\newpage

```{r four-component-unbalanced-separated, fig.cap=if (univariate_html_ouput) html_caption_unbalanced_well_separated_univariate else pdf_caption_unbalanced_well_separated_univariate}

true_theta <- univariate_configuration %>% filter(ID=="U7") %>% pull(true_parameters) %>% magrittr::extract2(1)
upper_part <- cowplot::plot_grid(plot_univariate_normal_density_distribution(true_theta) + labs(tag="A"),
                                 plot_time_computations(univariate_time_computation %>% filter(initialisation_method=="kmeans" & ID == "U7")) +
                                   theme(plot.tag = element_text(size=18, face = "bold")) + labs(tag = "B"), ncol = 2, align = 'h', axis="tblr")

univariate_unbalanced_separated <- gridExtra::arrangeGrob(upper_part,
                                                             plot_boxplots_parameters(formatted_univariate_distribution_parameters %>%
                                                                                        filter(ID == "U7"), true_theta) +
                                                               theme(plot.tag = element_text(size=18, face = "bold"),
                                                                     axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4)) +
                                                               labs(tag = "C"), top="", nrow = 2, heights = c(1.2, 2), padding = unit(1, "line"))

ggsave("./figs/univariate/univariate_unbalanced_separated.png", univariate_unbalanced_separated,
       width = 24, height = 18,dpi = 75)
knitr::include_graphics("./figs/univariate/univariate_unbalanced_separated.png", dpi = 75)
```









\newpage
```{r compute-balanced-overlapping-table-univariate-table}
univariate_parameters_local_scores_balanced_overlapping <- univariate_parameters_local_scores_summary %>%
  filter(ID=="U3") %>% select(-ID)
nrow_summary <- nrow(univariate_parameters_local_scores_balanced_overlapping)
```

```{r balanced-overlapping-table-univariate-pdf, eval=knitr::is_latex_output()}
univariate_parameters_local_scores_balanced_overlapping_kable <- univariate_parameters_local_scores_balanced_overlapping %>%
  kbl(booktabs=T, caption = "MSE and Bias associated to scenario U3, in Table \\ref{tab:parameter-configuration-univariate} (balanced and overlapping components)", escape=F,
      col.names = linebreak(c("Package", "Initialisation\nMethod",
                              "Global \n MSE $p$", "Global\nMSE $\\mu$", "Global\nMSE $\\sigma$",
                              "Global \n Bias $p$", "Global\nBias $\\mu$", "Global\nBias $\\sigma$"))) %>%
  kable_styling(latex_options=c("hold_position", "scale_down"), full_width = T) %>%
  row_spec(0,bold=T, align = "c") %>%
  column_spec(column = 1:8, latex_valign = "m") %>%
  column_spec(1, bold = T) %>% collapse_rows(columns = 1, valign = "middle")

for (metric_col in metric_colnames) {
  # define col vector
  col_vector <- rep("black", nrow_summary); metric_vector <- univariate_parameters_local_scores_balanced_overlapping %>% pull(metric_col)
  col_vector[which(metric_vector==min(metric_vector))] <- "green" # set minimal values to green
  col_vector[which(metric_vector==max(metric_vector))] <- "red" # set maximal values to red
  # set colours
  univariate_parameters_local_scores_balanced_overlapping_kable <- univariate_parameters_local_scores_balanced_overlapping_kable %>%
    column_spec(match(metric_col, names(univariate_parameters_local_scores_balanced_overlapping)),color=col_vector)
}
univariate_parameters_local_scores_balanced_overlapping_kable
```

```{r balanced-overlapping-table-univariate-html, eval=knitr::is_html_output(), layout = "l-body-outset"}
# latex + collapse rows + determine min and max

univariate_parameters_local_scores_balanced_overlapping_flex <- univariate_parameters_local_scores_balanced_overlapping %>% flextable() %>%
  set_caption(caption = "MSE and Bias associated to scenario U3, in Table \\@ref(tab:parameter-configuration-univariate) (balanced and overlapping components)", html_escape = F)

for (metric_col in metric_colnames) {
  univariate_parameters_local_scores_balanced_overlapping_flex <- univariate_parameters_local_scores_balanced_overlapping_flex %>%
    flextable::color(j=metric_col, part = "body",
                     i = as.formula(paste0("~ `", metric_col, "`==min(`", metric_col, "`)")),color = "green")

  univariate_parameters_local_scores_balanced_overlapping_flex <- univariate_parameters_local_scores_balanced_overlapping_flex %>%
    flextable::color(j=metric_col, part = "body",
                     i = as.formula(paste0("~ `", metric_col, "`==max(`", metric_col, "`)")),color = "red")
}

univariate_parameters_local_scores_balanced_overlapping_flex %>%
  flextable::compose(j="global_mse_p", part = "header", value = as_paragraph("Global MSE \U0070")) %>%
  flextable::compose(j="global_mse_mu", part = "header", value = as_paragraph("Global MSE \U03BC")) %>%
  # flextable::compose(j="global_mse_ sigma", part = "header", value = as_paragraph("Global_mse_ ", as_equation("\\sigma"))) %>%
  flextable::compose(j="global_mse_sigma", part = "header", value = as_paragraph("Global MSE \U03C3")) %>%
  flextable::compose(j="global_bias_p", part = "header", value = as_paragraph("Global Bias \U0070")) %>%
  flextable::compose(j="global_bias_mu", part = "header", value = as_paragraph("Global Bias \U03BC")) %>%
  flextable::compose(j="global_bias_sigma", part = "header", value = as_paragraph("Global Bias \U03C3")) %>%
  set_header_labels(package= "Package", `initialisation method`="Initialisation Method") %>%
  theme_vanilla() %>% align(align = "center", part = "header") %>%  # centre align header
  merge_v(j = "Package") %>%   flextable::autofit()
```

\newpage

```{r four-component-balanced-overlapping, fig.cap=if (univariate_html_ouput) html_caption_balanced_overlapping_univariate else pdf_caption_balanced_overlapping_univariate}

true_theta <- univariate_configuration %>% filter(ID=="U3") %>% pull(true_parameters) %>% magrittr::extract2(1)
upper_part <- cowplot::plot_grid(plot_univariate_normal_density_distribution(true_theta) + labs(tag="A"),
                                 plot_time_computations(univariate_time_computation %>% filter(initialisation_method=="kmeans" & ID == "U3")) +
                                   theme(plot.tag = element_text(size=18, face = "bold")) + labs(tag = "B"), ncol = 2, align = 'h', axis="tblr")

univariate_balanced_overlapping <- gridExtra::arrangeGrob(upper_part,
                                                             plot_boxplots_parameters(formatted_univariate_distribution_parameters %>%
                                                                                        filter(ID == "U3"), true_theta) +
                                                               theme(plot.tag = element_text(size=18, face = "bold"),
                                                                     axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4)) +
                                                               labs(tag = "C"), top="", nrow = 2, heights = c(1.2, 2), padding = unit(1, "line"))

ggsave("./figs/univariate/univariate_balanced_overlapping.png", univariate_balanced_overlapping,
       width = 24, height = 18,dpi = 75)
knitr::include_graphics("./figs/univariate/univariate_balanced_overlapping.png", dpi = 75)
```






\newpage
```{r compute-unbalanced-overlapping-table-univariate-table}
univariate_parameters_local_scores_unbalanced_overlapping <- univariate_parameters_local_scores_summary %>%
  filter(ID=="U9") %>% select(-ID)
nrow_summary <- nrow(univariate_parameters_local_scores_unbalanced_overlapping)
```

```{r unbalanced-overlapping-table-univariate-pdf, eval=knitr::is_latex_output()}
univariate_parameters_local_scores_unbalanced_overlapping_kable <- univariate_parameters_local_scores_unbalanced_overlapping %>%
  kbl(booktabs=T, caption = "MSE and Bias associated to scenario U9, in Table \\ref{tab:parameter-configuration-univariate} (unbalanced and overlapping components)", escape=F,
      col.names = linebreak(c("Package", "Initialisation\nMethod",
                              "Global \n MSE $p$", "Global\nMSE $\\mu$", "Global\nMSE $\\sigma$",
                              "Global \n Bias $p$", "Global\nBias $\\mu$", "Global\nBias $\\sigma$"))) %>%
  kable_styling(latex_options=c("hold_position", "scale_down"), full_width = T) %>%
  row_spec(0,bold=T, align = "c") %>%
  column_spec(column = 1:8, latex_valign = "m") %>%
  column_spec(1, bold = T) %>% collapse_rows(columns = 1, valign = "middle")

for (metric_col in metric_colnames) {
  # define col vector
  col_vector <- rep("black", nrow_summary); metric_vector <- univariate_parameters_local_scores_unbalanced_overlapping %>% pull(metric_col)
  col_vector[which(metric_vector==min(metric_vector))] <- "green" # set minimal values to green
  col_vector[which(metric_vector==max(metric_vector))] <- "red" # set maximal values to red
  # set colours
  univariate_parameters_local_scores_unbalanced_overlapping_kable <- univariate_parameters_local_scores_unbalanced_overlapping_kable %>%
    column_spec(match(metric_col, names(univariate_parameters_local_scores_unbalanced_overlapping)),color=col_vector)
}
univariate_parameters_local_scores_unbalanced_overlapping_kable
```

```{r unbalanced-overlapping-table-univariate-html, eval=knitr::is_html_output(), layout = "l-body-outset"}
# latex + collapse rows + determine min and max

univariate_parameters_local_scores_unbalanced_overlapping_flex <- univariate_parameters_local_scores_unbalanced_overlapping %>% flextable() %>%
  set_caption(caption = "MSE and Bias associated to scenario U9, in Table \\@ref(tab:parameter-configuration-univariate) (unbalanced and overlapping components)", html_escape = F)

for (metric_col in metric_colnames) {
  univariate_parameters_local_scores_unbalanced_overlapping_flex <- univariate_parameters_local_scores_unbalanced_overlapping_flex %>%
    flextable::color(j=metric_col, part = "body",
                     i = as.formula(paste0("~ `", metric_col, "`==min(`", metric_col, "`)")),color = "green")

  univariate_parameters_local_scores_unbalanced_overlapping_flex <- univariate_parameters_local_scores_unbalanced_overlapping_flex %>%
    flextable::color(j=metric_col, part = "body",
                     i = as.formula(paste0("~ `", metric_col, "`==max(`", metric_col, "`)")),color = "red")
}

univariate_parameters_local_scores_unbalanced_overlapping_flex %>%
  flextable::compose(j="global_mse_p", part = "header", value = as_paragraph("Global MSE \U0070")) %>%
  flextable::compose(j="global_mse_mu", part = "header", value = as_paragraph("Global MSE \U03BC")) %>%
  # flextable::compose(j="global_mse_ sigma", part = "header", value = as_paragraph("Global_mse_ ", as_equation("\\sigma"))) %>%
  flextable::compose(j="global_mse_sigma", part = "header", value = as_paragraph("Global MSE \U03C3")) %>%
  flextable::compose(j="global_bias_p", part = "header", value = as_paragraph("Global Bias \U0070")) %>%
  flextable::compose(j="global_bias_mu", part = "header", value = as_paragraph("Global Bias \U03BC")) %>%
  flextable::compose(j="global_bias_sigma", part = "header", value = as_paragraph("Global Bias \U03C3")) %>%
  set_header_labels(package= "Package", `initialisation method`="Initialisation Method") %>%
  theme_vanilla() %>% align(align = "center", part = "header") %>%  # centre align header
  merge_v(j = "Package") %>%   flextable::autofit()
```

\newpage

```{r four-component-unbalanced-overlapping, fig.cap=if (univariate_html_ouput) html_caption_unbalanced_overlapping_univariate else pdf_caption_unbalanced_overlapping_univariate}

true_theta <- univariate_configuration %>% filter(ID=="U9") %>% pull(true_parameters) %>% magrittr::extract2(1)
upper_part <- cowplot::plot_grid(plot_univariate_normal_density_distribution(true_theta) + labs(tag="A"),
                                 plot_time_computations(univariate_time_computation %>% filter(initialisation_method=="kmeans" & ID == "U9")) +
                                   theme(plot.tag = element_text(size=18, face = "bold")) + labs(tag = "B"), ncol = 2, align = 'h', axis="tblr")

univariate_unbalanced_overlapping <- gridExtra::arrangeGrob(upper_part,
                                                             plot_boxplots_parameters(formatted_univariate_distribution_parameters %>%
                                                                                        filter(ID == "U9"), true_theta) +
                                                               theme(plot.tag = element_text(size=18, face = "bold"),
                                                                     axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4)) +
                                                               labs(tag = "C"), top="", nrow = 2, heights = c(1.2, 2), padding = unit(1, "line"))

ggsave("./figs/univariate/univariate_unbalanced_overlapping.png", univariate_unbalanced_overlapping,
       width = 24, height = 18,dpi = 75)
knitr::include_graphics("./figs/univariate/univariate_unbalanced_overlapping.png", dpi = 75)
```




\newpage
```{r four-components-midbalanced, fig.cap=if (univariate_html_ouput) html_caption_intermediate_scenario_univariate else pdf_caption_intermediate_scenario_univariate}
true_theta_scenario_U4 <- univariate_configuration %>% filter(ID=="U4") %>% pull(true_parameters) %>% magrittr::extract2(1)
true_theta_scenario_U6 <- univariate_configuration %>% filter(ID=="U6") %>% pull(true_parameters) %>% magrittr::extract2(1)

four_components_midbalanced_separated <- gridExtra::arrangeGrob(plot_univariate_normal_density_distribution(true_theta_scenario_U4) + labs(tag="A") +
                                                                  theme(plot.tag = element_text(size=18, face = "bold")),
                                                                plot_univariate_normal_density_distribution(true_theta_scenario_U6) + labs(tag="B") +
                                                                  theme(plot.tag = element_text(size=18, face = "bold")),
                                                                plot_boxplots_parameters(formatted_univariate_distribution_parameters %>% filter(ID=="U4"), true_theta_scenario_U6) +
                                                                  theme(plot.tag = element_text(size=18, face = "bold"),
                                                                        axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4)) + labs(tag = "C"),
                                                                plot_boxplots_parameters(formatted_univariate_distribution_parameters %>% filter(ID=="U6"), true_theta_scenario_U6) +
                                                                  theme(plot.tag = element_text(size=18, face = "bold"),
                                                                        axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4)) + labs(tag = "D"),
                                                                top="", layout_matrix = matrix(c(1, 2, 3, 3, 4, 4), byrow = TRUE, nrow = 3), heights = c(2, 2.5, 2.5), padding = unit(1, "line"))

ggsave("figs/univariate/univariate_midbalanced.png", four_components_midbalanced_separated,
       width = 22, height = 26,dpi = 75)
knitr::include_graphics("figs/univariate/univariate_midbalanced.png", dpi = 75)
```




\newpage
```{r heatmap-all-correlation-plots-univariate, fig.cap= if (univariate_html_ouput) html_global_heatmap_univariate else pdf_global_heatmap_univariate, layout = "l-body-outset"}
discriminating_data <- univariate_distribution_parameters %>% dplyr::filter(ID=="U9") %>%
  dplyr::bind_rows(readRDS("./tables/univariate/univariate_small_EM.rds"))

univariate_correlation_heatmaps <- plot_correlation_Heatmap(discriminating_data)

univariate_correlation_heatmaps <- gridExtra::arrangeGrob(grobs=univariate_correlation_heatmaps %>% purrr::map(~grid::grid.grabExpr(ComplexHeatmap::draw(.x))),
                                                          top="", ncol = 2)
ggsave("./figs/univariate/heatmap_univariate.png", univariate_correlation_heatmaps, width = 10.5, height = 13.5,dpi = 75)

knitr::include_graphics("./figs/univariate/heatmap_univariate.png", dpi = 75)
```

\newpage
```{r univariate-initialisation-time-computations, fig.cap= if (univariate_html_ouput) html_univariate_init_time else pdf_univariate_init_time}

initialisation_time_data <- readRDS("./tables/univariate/univariate_initialisation_time_computation.rds")
inialisation_plot <- plot_initialisation_time_computations(initialisation_time_data, size=4, x = 2.6)
ggsave("./figs/univariate_initialisation_time_computations.png", inialisation_plot, dpi = 75)

knitr::include_graphics("./figs/univariate_initialisation_time_computations.png", dpi = 75)
```


The panels indexed by the B letter, from Figure
\@ref(fig:four-component-balanced-separated) to Figure
\@ref(fig:four-components-midbalanced), display the 0.05, 0.5 and 0.95 quantiles of the distribution of the operating times
taken for parameter estimation, for the scenarios listed in Table \@ref(tab:parameter-configuration-univariate).


First, we note that the execution time grows asymptotically
linearly with the number of observations, confirming empirically the
expected linear complexity of the EM algorithm. The most important
factor playing on the differences observed is related to the complexity
of the distribution, and especially the degree of overlap between the
components:

* On the one hand hand, when components are well-separated (scenarios 1 and 3 in Table \@ref(tab:parameter-configuration-univariate)), the estimation of
the parameters is simple, leading to a reduced number of iterations
required to reach the convergence and shorter running times.

* On the other hand, the time taken by the slowest package for the estimation of the parameters increases by a hundred factor with the most
complex scenario (see scenario U9, \@ref(tab:parameter-configuration-univariate), illustrated in Figure
\@ref(fig:four-component-unbalanced-overlapping)), compared to the simplest scenario (see U1, \@ref(tab:parameter-configuration-univariate), shown in Figure \@ref(fig:four-component-balanced-separated)). Indeed, the average running time for a complete run of the EM algorithm increases from 0.215 seconds to 10.8 seconds.


To better understand the running times' differences observed between the
packages for a given scenario, we perform a three-way anova, taking into
account the choice of initialisation method, the programming language
and the class of packages^[To compare
whether differences between mean running times or estimation
performances differ across packages, we used the between-subjects Anova test `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("rstatix::anova_test()"),   '\\code{rstatix::anova\\_test()}'))` to generate
the *p*-values and `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("rstatix::partial_eta_squared()"),   '\\code{rstatix::partial\\_eta\\_squared()}'))` to compute the corresponding effect sizes.]:

* With well-separated components (Scenarios U1
and U7 in Table \@ref(tab:parameter-configuration-univariate)), the class
of packages (namely the choice of the convergence criterion) has a negligible impact compared to the choice of initialisation algorithm or the programming language. The effect sizes associated to the programming language and  the initialisation method are respectively $1.688 \times 10^{-2}$ (*p*-value of $3 \times 10^{-60}$) and $13 \times 10^{-5}$ (*p*-value of
$3 \times 10^{-60}$), while the choice of the termination criteria did
not significantly impact the execution time, with an effect size of
$8.119 \times 10^{-4}$( *p*-value of $0.35$). Faster running times with
packages natively encoded in Fortran or C compared to those encoded in R
only were expected, as R is a high-level programming language known to
be slower. Indeed, the \pkg{flexmix} package is the slowest, preceded by
our baseline R implementation. Additionally,
\pkg{mclust}, followed by \pkg{mixtools}, \pkg{Rmixmod} and \pkg{bgmm}
are the fastest.

* On the other hand, with overlapping components (Scenarios U3 and U9 in Table \@ref(tab:parameter-configuration-univariate)), the package class and the programming language have a statistically significant impact on the average running times (the effect sizes associated to the
choice of the termination criteria and the programming language are
respectively $0.111$ (numerical null *p*-value) and $0.0852$ (*p*-value of
$8 \times 10^{-307}$)) while the initialisation method has no substantial impact (effect size of
$2.967 \times 10^{-4}$ and *p*-value of $0.32$). In the context of highly overlapping mixture, the fastest ones are \pkg{mclust} and \pkg{GMKMcharlie}, benefiting from both using
relative ratios and a fast programming language, while our
baseline implementation `emnmix`, preceded by \pkg{Rmixmod} and \pkg{mixtools}, are on average a hundred times slower.

\newpage


## Supplementary Figures and Tables in the bivariate simulation {.appendix}


Table below (\@ref(tab:parameter-configuration-bivariate)) lists the complete set of parameters used to simulate the multivariate Gaussian mixture distribution in our benchmark:



```{r parameter-configuration-bivariate, layout = "l-page"}
bivariate_configuration  %>% mutate(Proportions=purrr::map_chr(true_parameters, ~paste(.x$p, collapse = " / ")),
                                       Means=purrr::map_chr(true_parameters, ~paste0("(", paste0(.x$mu[,1], collapse = ","), ");(", paste0(.x$mu[,2], collapse = ","), ")")),
                                       Correlations=purrr::map_chr(true_parameters, ~paste(c(.x$sigma[1, 2, 1], .x$sigma[1, 2, 2]), collapse = " / "))) %>%
  select(-c("OVL", "nobservations", "prop_outliers", "true_parameters", "formatted_true_parameters", "true_parameters_factor")) %>%
  rename(Entropy=entropy, OVL=OVL_pairwise) %>%
  kbl(booktabs=T, caption = "The 20 parameter configurations tested to generate the samples of the bivariate experiment.",
      escape=F, align = "c") %>%
  kable_styling(latex_options=c("hold_position", "scale_down")) %>%
  row_spec(0,bold=T) %>% row_spec(1:20, hline_after = T) %>%
  kable_styling(bootstrap_options = c("hover"))
```

```{r prepare-legend-bivariate}
html_output <- knitr::is_html_output()  # change to TRUE for larger figures

html_caption_unbalanced_negative_correlated_bivariate <- paste("Results of scenario B11 in Table \\@ref(tab:parameter-configuration-bivariate) (unbalanced, overlapping and negative correlated components), organised as such:

 - The panel A displays the bivariate contour maps associated to the two-components multivariate Gaussian distribution corresponding to the parametrisation described by the scenario, warmer colours corresponding to regions of higher densities. The two centroids, whose coordinates are given by the mean components' elements, are represented with distinct shaped and coloured point estimates.

  - In both Panels A and B, the ellipsoids correspond to the $95\\%$ confidence region associated to each component's distribution. To generate them, we largely inspired from the",
downlit::autolink('mixtools::ellipse()'), "and website [How to draw ellipses](https://cookierobotics.com/007/). To generate them, we retain for each individual parameter its mean (similar results with the median) over the $N=100$ sampling experiments, restrained to the random initialisation method.

- The running times are displayed in Panel C with the *k*-means initialisation. The number of observations
(x-axis) and the running time (y-axis) is in $\\log(10)$ scale. The points represent median
running time. The coloured bands represent the $5^{\\text{th}}$ and $95^{\\text{th}}$ percentiles
of the running time.

- The distributions of the Hellinger distances (a closed form is only available for the Gaussian multivariate distribution, not the mixture) are computed for each component, each initialisation method and each package with respect to the true Gaussian distribution expected for each component. The more dissimilar are the distributions, the higher is the Hellinger distance, knowing it is normalised between 0 and 1^[More on the Hellinger distance as a quantifier of the similarity between two probability distributions in [Hellinger distance](https://en.wikipedia.org/wiki/Hellinger_distance)]. We represent them using boxplot representations in Panel D.

- In panel E we represent the boxplots associated with the distribution of the estimates, with one box per pair of package and initialisation method, using the same convenstions detailed in [Supplementary Figures and Tables in the univariate simulation]. As the correlation is a symmetric operator, we only represent the distribution of the lower part of the lower matrix. Each column is associated to the parameters of a component. First row represents the distribution of the estimated ratios, second and third respectively the distributions of the mean vector on the x-axis and on the y-axis, third and four the distributions of the individual variances of each feature and finally the fifth row shows the distribution of the correlation between dimension 1 and 2.")


pdf_caption_unbalanced_negative_correlated_bivariate <- "Results of scenario B11 in Table \\ref{tab:parameter-configuration-bivariate} (unbalanced, overlapping and negative correlated components),  organised as such:
The panel A displays the bivariate contour maps associated to the two-components multivariate Gaussian distribution corresponding to the parametrisation described by the scenario, warmer colours corresponding to regions of higher densities. The two centroids, whose coordinates are given by the mean components' elements, are represented with distinct shaped and coloured point estimates.
In both Panels A and B, the ellipsoids correspond to the $95\\%$ confidence region associated to each component's distribution. To generate them, we largely inspired from the \\code{mixtools::ellipse()} and website \\href{https://cookierobotics.com/007/}{How to draw ellipses}. To generate them, we retain for each individual parameter its mean (similar results with the median) over the $N=100$ sampling experiments, restrained to the random initialisation method.
The running times are displayed in Panel C with the \\textit{k}-means initialisation. The number of observations (x-axis) and the running time (y-axis) is in $\\log(10)$ scale. The points represent median running time. The coloured bands represent the $5^{\\text{th}}$ and $95^{\\text{th}}$ percentiles of the running time.
The distributions of the Hellinger distances (a closed form is only available for the Gaussian multivariate distribution, not the mixture) are computed for each component, each initialisation method and each package with respect to the true Gaussian distribution expected for each component. The more dissimilar are the distributions, the higher is the Hellinger distance, knowing it is normalised between 0 and 1. We represent them using boxplot representations in Panel D.
In panel E we represent the boxplots associated with the distribution of the estimates, with one box per pair of package and initialisation method. As the correlation is a symmetric operator, we only represent the distribution of the lower part of the lower matrix. Each column is associated to the parameters of a component. First row represents the distribution of the estimated ratios, second and third respectively the distributions of the mean vector on the x-axis and on the y-axis, third and four the distributions of the individual variances of each feature and finally the fifth row shows the distribution of the correlation between dimension 1 and 2."



html_caption_unbalanced_opposite_correlated_bivariate <- "Results of scenario B12 in Table \\@ref(tab:parameter-configuration-bivariate)  (unbalanced, overlapping and opposite correlated components), with the same layout as Figure \\@ref(fig:multivariate-overlapping-unbalanced-negative-correlated)."
pdf_caption_unbalanced_opposite_correlated_bivariate <- "Results of scenario B12 in Table \\ref{tab:parameter-configuration-bivariate} (unbalanced, overlapping and opposite correlated components), with the same layout as Figure \\ref{fig:multivariate-overlapping-unbalanced-negative-correlated}."

html_caption_unbalanced_positive_correlated_bivariate <- "Results of scenario B14 in Table \\@ref(tab:parameter-configuration-bivariate) (unbalanced, overlapping and positive correlated components), with the same layout as Figure \\@ref(fig:multivariate-overlapping-unbalanced-negative-correlated)."
pdf_caption_unbalanced_positive_correlated_bivariate <- "Results of scenario B14 in Table \\ref{tab:parameter-configuration-bivariate} (unbalanced, overlapping and positive correlated components), with the same layout as Figure \\ref{fig:multivariate-overlapping-unbalanced-negative-correlated}."

html_caption_unbalanced_uncorrelated_bivariate <- "Results of scenario B15 in Table \\@ref(tab:parameter-configuration-bivariate) (unbalanced, overlapping and uncorrelated components), with the same layout as Figure \\@ref(fig:multivariate-overlapping-unbalanced-negative-correlated)."
pdf_caption_unbalanced_uncorrelated_bivariate <- "Results of scenario B15 in Table \\ref{tab:parameter-configuration-bivariate} (unbalanced, overlapping and uncorrelated components), with the same layout as Figure \\ref{fig:multivariate-overlapping-unbalanced-negative-correlated}."



html_balanced_overlapping_general_bivariate <- "Benchmark summary plots of scenarios B1, B2 and B5 in Table \\@ref(tab:parameter-configuration-bivariate) featuring balanced and overlapping clusters. Summary plots of B1, B2 and B5 are represented in this order on each row, with the left column displaying the 95% confidence ellipsoidal regions associated to the mean estimated parameters across each package and the right column the distribution of the Hellinger distances."
pdf_balanced_overlapping_general_bivariate <- "Benchmark summary plots of respectively scenarios B1, B2 and B5 in Table \\ref{tab:parameter-configuration-bivariate} featuring balanced and overlapping clusters. Summary plots of B1, B2 and B5 are represented in this order on each row, with the left column displaying the $95\\%$ confidence ellipsoidal regions associated to the mean estimated parameters across each package and the right column the distribution of the Hellinger distances."

html_balanced_well_separated_general_bivariate <- "Benchmark summary plots of respectively scenarios B6, B7 and B10 in Table \\@ref(tab:parameter-configuration-bivariate) featuring balanced and well-separated clusters, with the same layout as Figure \\@ref(fig:general-balanced-overlapping-bivariate)."
pdf_balanced_well_separated_general_bivariate <- "Benchmark summary plots of respectively scenarios B6, B7 and B10 in Table \\ref{tab:parameter-configuration-bivariate} featuring balanced and well-separated clusters, with the same layout as Figure \\ref{fig:general-balanced-overlapping-bivariate}."

html_unbalanced_well_separated_general_bivariate <- "Benchmark summary plots of respectively scenarios B16, B17 and B20 in Table \\@ref(tab:parameter-configuration-bivariate) featuring unbalanced and well-separated clusters, with the same layout as Figure \\@ref(fig:general-balanced-overlapping-bivariate).."
pdf_unbalanced_well_separated_general_bivariate <- "Benchmark summary plots of respectively scenarios B16, B17 and B20 in Table \\ref{tab:parameter-configuration-bivariate} featuring unbalanced and well-separated clusters, with the same layout as Figure \\ref{fig:general-balanced-overlapping-bivariate}."

html_global_heatmap_bivariate <- "Correlation heatmaps of the estimated parameters in the bivariate setting extended to the four initialisation methods benchmarked, with the most discriminating scenario B11, using the same process described in Figure (2)."
pdf_global_heatmap_bivariate <- "Correlation heatmaps of the estimated parameters in the bivariate setting extended to the four initialisation methods benchmarked, with the most discriminating scenario B11, using the same process described in Figure (2)."
```

Figures \@ref(fig:multivariate-overlapping-unbalanced-negative-correlated)-
\@ref(fig:multivariate-overlapping-unbalanced-uncorrelated) are associated to scenarios B11 - B15 of Table \@ref(tab:parameter-configuration-bivariate). Summary tables
`r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(tab:multivariate-overlapping-unbalanced-negative-correlated-html)', '\\@ref(tab:multivariate-overlapping-unbalanced-negative-correlated-pdf)'))`-`r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(tab:multivariate-overlapping-unbalanced-uncorrelated-html)', '\\@ref(tab:multivariate-overlapping-unbalanced-uncorrelated-pdf)'))`
show the average performance for each combination of a benchmarked package and initialisation method, with the same conventions as discussed in [Supplementary Figures and Tables in the univariate simulation].


First, we can directly observe that the OVL increases as the individual variance of each component, the proximity of the centroids of the clusters and the level of imbalance is increased. We demonstrate this statement formally in section [An analytic formula of the overlap for univariate Gaussian mixtures]. Nonetheless, the influence of the correlation between the $x$ and the $y$-axis (the off-diagonal term of the covariance matrix) is not immediate, notably the assumption of independent features does not automatically entail a lower OVL or simpler estimation.


From our experiments, we deduce that the highest OVL is obtained when the main axis of the two respective components aligns with the line joining the two centroids. For instance, in our scenario, the lowest OVL is obtained when the correlation term is positive for both clusters (scenario 14, Table \@ref(tab:parameter-configuration-bivariate) and isodensity plot in panel A, Figure \@ref(fig:multivariate-overlapping-unbalanced-positive-correlated)), whereas the highest OVL is obtained with a negative correlation (scenario 11, Table \@ref(tab:parameter-configuration-bivariate)and isodensity plot in panel A, Figure \@ref(fig:multivariate-overlapping-unbalanced-negative-correlated)). Recall that the slope joining the two centroids of the two components in all our simulated distributions is indeed negative.



```{r compute-bivariate-overlapping-unbalanced-negative-correlated}
specific_bivariate_parameters_local_scores <- bivariate_parameters_local_scores_summary %>%
  dplyr::filter(ID=="B11") %>% select(-ID)
nrow_summary <- nrow(specific_bivariate_parameters_local_scores)
```

```{r multivariate-overlapping-unbalanced-negative-correlated-pdf, eval=knitr::is_latex_output()}
specific_bivariate_parameters_local_scores_kable <- specific_bivariate_parameters_local_scores %>%
  kbl(booktabs=T, caption = "MSE and Bias associated to scenario B11, in Table \\ref{tab:parameter-configuration-bivariate} (unbalanced, overlapping and negative correlated components).", escape=F,
      col.names = linebreak(c("Package", "Initialisation\nMethod",
                              "Global \n MSE $p$", "Global\nMSE $\\mu$", "Global\nMSE $\\sigma$",
                              "Global \n Bias $p$", "Global\nBias $\\mu$", "Global\nBias $\\sigma$"))) %>%
  kable_styling(latex_options=c("hold_position", "scale_down"), full_width = T) %>%
  row_spec(0,bold=T, align = "c") %>%
  column_spec(column = 1:8, latex_valign = "m") %>%
  column_spec(1, bold = T) %>% collapse_rows(columns = 1, valign = "middle")


for (metric_col in metric_colnames) {
  # define col vector
  col_vector <- rep("black", nrow_summary); metric_vector <- specific_bivariate_parameters_local_scores %>% pull(metric_col)
  col_vector[which(metric_vector==min(metric_vector))] <- "green" # set minimal values to green
  col_vector[which(metric_vector==max(metric_vector))] <- "red" # set maximal values to red
  # set colours
  specific_bivariate_parameters_local_scores_kable <- specific_bivariate_parameters_local_scores_kable %>%
    column_spec(match(metric_col, names(specific_bivariate_parameters_local_scores)),color=col_vector)
}
specific_bivariate_parameters_local_scores_kable
```

```{r multivariate-overlapping-unbalanced-negative-correlated-html, eval=knitr::is_html_output(), layout = "l-body-outset"}
specific_bivariate_parameters_local_scores_flex <- specific_bivariate_parameters_local_scores %>% flextable() %>%
  set_caption(caption = "MSE and Bias associated to scenario B11, in Table \\@ref(tab:parameter-configuration-bivariate) (unbalanced, overlapping and negative correlated components.)", html_escape = F)

for (metric_col in metric_colnames) {
  specific_bivariate_parameters_local_scores_flex <- specific_bivariate_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body",
                     i = as.formula(paste0("~ `", metric_col, "`==min(`", metric_col, "`)")),color = "green")

  specific_bivariate_parameters_local_scores_flex <- specific_bivariate_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body",
                     i = as.formula(paste0("~ `", metric_col, "`==max(`", metric_col, "`)")),color = "red")
}

specific_bivariate_parameters_local_scores_flex %>%
  flextable::compose(j="global_mse_p", part = "header", value = as_paragraph("Global MSE \U0070")) %>%
  flextable::compose(j="global_mse_mu", part = "header", value = as_paragraph("Global MSE \U03BC")) %>%
  # flextable::compose(j="global_mse_ sigma", part = "header", value = as_paragraph("Global_mse_ ", as_equation("\\sigma"))) %>%
  flextable::compose(j="global_mse_sigma", part = "header", value = as_paragraph("Global MSE \U03C3")) %>%
  flextable::compose(j="global_bias_p", part = "header", value = as_paragraph("Global Bias \U0070")) %>%
  flextable::compose(j="global_bias_mu", part = "header", value = as_paragraph("Global Bias \U03BC")) %>%
  flextable::compose(j="global_bias_sigma", part = "header", value = as_paragraph("Global Bias \U03C3")) %>%
  set_header_labels(package= "Package", `initialisation method`="Initialisation Method") %>%
  theme_vanilla() %>% align(align = "center", part = "header") %>%  # centre align header
  merge_v(j = "Package") %>%   flextable::autofit()

```

\newpage

```{r multivariate-overlapping-unbalanced-negative-correlated, fig.cap=if (html_output) html_caption_unbalanced_negative_correlated_bivariate else pdf_caption_unbalanced_negative_correlated_bivariate, out.width="60%", layout = "l-body-outset"}
true_theta <- bivariate_configuration %>% filter(ID=="B11") %>% pull(true_parameters) %>% magrittr::extract2(1)

# generate plots
bivariate_density <- plot_bivariate_normal_density_distribution(true_theta=true_theta, n = 2000)
computation_time <- plot_time_computations(bivariate_time_computation %>% filter(ID == "B11"))
ellipse <- plot_ellipses_bivariate(bivariate_parameters %>% filter(ID == "B11"), true_theta)
hellinger <- plot_Hellinger(formatted_bivariate_parameters %>% filter(ID == "B11"), true_theta) +
  theme(axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4))
boxplot_parameter_bivariate <- plot_boxplots_parameters(formatted_bivariate_parameters %>% filter(ID == "B11"), true_theta) + theme(axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4))


# concatenate plots together
upper_part <- cowplot::plot_grid(bivariate_density + labs(tag="A") + theme(plot.tag = element_text(size=24, face = "bold")),
                                 ellipse + labs(tag="B") + theme(plot.tag = element_text(size=24, face = "bold"), legend.title = element_blank()),
                                 computation_time + labs(tag="C") + theme(plot.tag = element_text(size=24, face = "bold")),
                                 hellinger + labs(tag = "D") + theme(plot.tag = element_text(size=24, face = "bold")) ,
                                 ncol = 2, align = 'h', axis="tblr", nrow=2)
global_figure <- gridExtra::arrangeGrob(upper_part,
                                        boxplot_parameter_bivariate +
                                          theme(plot.tag = element_text(size=24, face = "bold")) + labs(tag = "E"),
                                        top="", nrow = 2, heights = c(1.5, 1), padding = unit(1, "line"))
ggsave("figs/bivariate/multivariate_unbalanced_overlapping_negatively_correlated.png",
       global_figure, width = 20, height = 24, dpi = 75)

knitr::include_graphics("figs/bivariate/multivariate_unbalanced_overlapping_negatively_correlated.png", dpi = 75)
```








\newpage
```{r compute-bivariate-overlapping-unbalanced-opposite-correlated}
specific_bivariate_parameters_local_scores <- bivariate_parameters_local_scores_summary %>%
  dplyr::filter(ID=="B12") %>% select(-ID)
nrow_summary <- nrow(specific_bivariate_parameters_local_scores)
```

```{r multivariate-overlapping-unbalanced-opposite-correlated-pdf, eval=knitr::is_latex_output()}
specific_bivariate_parameters_local_scores_kable <- specific_bivariate_parameters_local_scores %>%
  kbl(booktabs=T, caption = "MSE and Bias associated to scenario B12, in Table \\ref{tab:parameter-configuration-bivariate} (unbalanced, overlapping and opposite correlated components).", escape=F,
      col.names = linebreak(c("Package", "Initialisation\nMethod",
                              "Global \n MSE $p$", "Global\nMSE $\\mu$", "Global\nMSE $\\sigma$",
                              "Global \n Bias $p$", "Global\nBias $\\mu$", "Global\nBias $\\sigma$"))) %>%
  kable_styling(latex_options=c("hold_position", "scale_down"), full_width = T) %>%
  row_spec(0,bold=T, align = "c") %>%
  column_spec(column = 1:8, latex_valign = "m") %>%
  column_spec(1, bold = T) %>% collapse_rows(columns = 1, valign = "middle")


for (metric_col in metric_colnames) {
  # define col vector
  col_vector <- rep("black", nrow_summary); metric_vector <- specific_bivariate_parameters_local_scores %>% pull(metric_col)
  col_vector[which(metric_vector==min(metric_vector))] <- "green" # set minimal values to green
  col_vector[which(metric_vector==max(metric_vector))] <- "red" # set maximal values to red
  # set colours
  specific_bivariate_parameters_local_scores_kable <- specific_bivariate_parameters_local_scores_kable %>%
    column_spec(match(metric_col, names(specific_bivariate_parameters_local_scores)),color=col_vector)
}
specific_bivariate_parameters_local_scores_kable
```

```{r multivariate-overlapping-unbalanced-opposite-correlated-html, eval=knitr::is_html_output(), layout = "l-body-outset"}
specific_bivariate_parameters_local_scores_flex <- specific_bivariate_parameters_local_scores %>% flextable() %>%
  set_caption(caption = "MSE and Bias associated to scenario B12, in Table \\@ref(tab:parameter-configuration-bivariate)(unbalanced, overlapping and opposite correlated components.)", html_escape = F)

for (metric_col in metric_colnames) {
  specific_bivariate_parameters_local_scores_flex <- specific_bivariate_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body",
                     i = as.formula(paste0("~ `", metric_col, "`==min(`", metric_col, "`)")),color = "green")

  specific_bivariate_parameters_local_scores_flex <- specific_bivariate_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body",
                     i = as.formula(paste0("~ `", metric_col, "`==max(`", metric_col, "`)")),color = "red")
}

specific_bivariate_parameters_local_scores_flex %>%
  flextable::compose(j="global_mse_p", part = "header", value = as_paragraph("Global MSE \U0070")) %>%
  flextable::compose(j="global_mse_mu", part = "header", value = as_paragraph("Global MSE \U03BC")) %>%
  # flextable::compose(j="global_mse_ sigma", part = "header", value = as_paragraph("Global_mse_ ", as_equation("\\sigma"))) %>%
  flextable::compose(j="global_mse_sigma", part = "header", value = as_paragraph("Global MSE \U03C3")) %>%
  flextable::compose(j="global_bias_p", part = "header", value = as_paragraph("Global Bias \U0070")) %>%
  flextable::compose(j="global_bias_mu", part = "header", value = as_paragraph("Global Bias \U03BC")) %>%
  flextable::compose(j="global_bias_sigma", part = "header", value = as_paragraph("Global Bias \U03C3")) %>%
  set_header_labels(package= "Package", `initialisation method`="Initialisation Method") %>%
  theme_vanilla() %>% align(align = "center", part = "header") %>%  # centre align header
  merge_v(j = "Package") %>%   flextable::autofit()

```

\newpage

```{r multivariate-overlapping-unbalanced-opposite-correlated, fig.cap=if (html_output) html_caption_unbalanced_opposite_correlated_bivariate else pdf_caption_unbalanced_opposite_correlated_bivariate, layout = "l-body-outset"}
true_theta <- bivariate_configuration %>% filter(ID=="B12") %>% pull(true_parameters) %>% magrittr::extract2(1)

# generate plots
bivariate_density <- plot_bivariate_normal_density_distribution(true_theta=true_theta, n = 2000)
computation_time <- plot_time_computations(bivariate_time_computation %>% filter(ID == "B13"))
ellipse <- plot_ellipses_bivariate(bivariate_parameters %>% filter(ID == "B13"), true_theta)
hellinger <- plot_Hellinger(formatted_bivariate_parameters%>% filter(ID == "B13"), true_theta) +
  theme(axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4))
boxplot_parameter_bivariate <- plot_boxplots_parameters(formatted_bivariate_parameters %>% filter(ID == "B13"), true_theta) +
  theme(axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4))


# concatenate plots together
upper_part <- cowplot::plot_grid(bivariate_density + labs(tag="A") + theme(plot.tag = element_text(size=24, face = "bold")),
                                 ellipse + labs(tag="B") + theme(plot.tag = element_text(size=24, face = "bold"), legend.title = element_blank()),
                                 computation_time + labs(tag="C") + theme(plot.tag = element_text(size=24, face = "bold")),
                                 hellinger + labs(tag = "D") + theme(plot.tag = element_text(size=24, face = "bold")) ,
                                 ncol = 2, align = 'h', axis="tblr", nrow=2)
global_figure <- gridExtra::arrangeGrob(upper_part,
                                        boxplot_parameter_bivariate +
                                          theme(plot.tag = element_text(size=24, face = "bold")) + labs(tag = "E"),
                                        top="", nrow = 2, heights = c(1.5, 1), padding = unit(1, "line"))
ggsave("figs/bivariate/multivariate_unbalanced_overlapping_opposite_correlated.png",
       global_figure, width = 20, height = 24, dpi = 75)

knitr::include_graphics("figs/bivariate/multivariate_unbalanced_overlapping_opposite_correlated.png", dpi = 75)
```


\newpage





```{r compute-bivariate-overlapping-unbalanced-positive-correlated}
specific_bivariate_parameters_local_scores <- bivariate_parameters_local_scores_summary %>%
  dplyr::filter(ID=="B14") %>% select(-ID)
nrow_summary <- nrow(specific_bivariate_parameters_local_scores)
```

```{r multivariate-overlapping-unbalanced-positive-correlated-pdf, eval=knitr::is_latex_output()}
specific_bivariate_parameters_local_scores_kable <- specific_bivariate_parameters_local_scores %>%
  kbl(booktabs=T, caption = "MSE and Bias associated to scenario B14, in Table \\ref{tab:parameter-configuration-bivariate}  (unbalanced, overlapping and positive correlated components).", escape=F,
      col.names = linebreak(c("Package", "Initialisation\nMethod",
                              "Global \n MSE $p$", "Global\nMSE $\\mu$", "Global\nMSE $\\sigma$",
                              "Global \n Bias $p$", "Global\nBias $\\mu$", "Global\nBias $\\sigma$"))) %>%
  kable_styling(latex_options=c("hold_position", "scale_down"), full_width = T) %>%
  row_spec(0,bold=T, align = "c") %>%
  column_spec(column = 1:8, latex_valign = "m") %>%
  column_spec(1, bold = T) %>% collapse_rows(columns = 1, valign = "middle")


for (metric_col in metric_colnames) {
  # define col vector
  col_vector <- rep("black", nrow_summary); metric_vector <- specific_bivariate_parameters_local_scores %>% pull(metric_col)
  col_vector[which(metric_vector==min(metric_vector))] <- "green" # set minimal values to green
  col_vector[which(metric_vector==max(metric_vector))] <- "red" # set maximal values to red
  # set colours
  specific_bivariate_parameters_local_scores_kable <- specific_bivariate_parameters_local_scores_kable %>%
    column_spec(match(metric_col, names(specific_bivariate_parameters_local_scores)),color=col_vector)
}
specific_bivariate_parameters_local_scores_kable
```

```{r multivariate-overlapping-unbalanced-positive-correlated-html, eval=knitr::is_html_output(), layout = "l-body-outset"}
specific_bivariate_parameters_local_scores_flex <- specific_bivariate_parameters_local_scores %>% flextable() %>%
  set_caption(caption = "MSE and Bias associated to scenario B14, in Table \\@ref(tab:parameter-configuration-bivariate) (unbalanced, overlapping and positive correlated components.)", html_escape = F)

for (metric_col in metric_colnames) {
  specific_bivariate_parameters_local_scores_flex <- specific_bivariate_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body",
                     i = as.formula(paste0("~ `", metric_col, "`==min(`", metric_col, "`)")),color = "green")

  specific_bivariate_parameters_local_scores_flex <- specific_bivariate_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body",
                     i = as.formula(paste0("~ `", metric_col, "`==max(`", metric_col, "`)")),color = "red")
}

specific_bivariate_parameters_local_scores_flex %>%
  flextable::compose(j="global_mse_p", part = "header", value = as_paragraph("Global MSE \U0070")) %>%
  flextable::compose(j="global_mse_mu", part = "header", value = as_paragraph("Global MSE \U03BC")) %>%
  # flextable::compose(j="global_mse_ sigma", part = "header", value = as_paragraph("Global_mse_ ", as_equation("\\sigma"))) %>%
  flextable::compose(j="global_mse_sigma", part = "header", value = as_paragraph("Global MSE \U03C3")) %>%
  flextable::compose(j="global_bias_p", part = "header", value = as_paragraph("Global Bias \U0070")) %>%
  flextable::compose(j="global_bias_mu", part = "header", value = as_paragraph("Global Bias \U03BC")) %>%
  flextable::compose(j="global_bias_sigma", part = "header", value = as_paragraph("Global Bias \U03C3")) %>%
  set_header_labels(package= "Package", `initialisation method`="Initialisation Method") %>%
  theme_vanilla() %>% align(align = "center", part = "header") %>%  # centre align header
  merge_v(j = "Package") %>%   flextable::autofit()

```

\newpage

```{r multivariate-overlapping-unbalanced-positive-correlated, fig.cap=if (html_output) html_caption_unbalanced_positive_correlated_bivariate else pdf_caption_unbalanced_positive_correlated_bivariate, layout = "l-body-outset"}
true_theta <- bivariate_configuration %>% filter(ID=="B14") %>% pull(true_parameters) %>% magrittr::extract2(1)

# generate plots
bivariate_density <- plot_bivariate_normal_density_distribution(true_theta=true_theta, n = 2000)
computation_time <- plot_time_computations(bivariate_time_computation %>% filter(ID == "B14"))
ellipse <- plot_ellipses_bivariate(bivariate_parameters %>% filter(ID == "B14"), true_theta)
hellinger <- plot_Hellinger(formatted_bivariate_parameters %>% filter(ID == "B14"), true_theta) +
  theme(axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4))
boxplot_parameter_bivariate <- plot_boxplots_parameters(formatted_bivariate_parameters %>% filter(ID == "B14"), true_theta) +
  theme(axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4))


# concatenate plots together
upper_part <- cowplot::plot_grid(bivariate_density + labs(tag="A") + theme(plot.tag = element_text(size=24, face = "bold")),
                                 ellipse + labs(tag="B") + theme(plot.tag = element_text(size=24, face = "bold"), legend.title = element_blank()),
                                 computation_time + labs(tag="C") + theme(plot.tag = element_text(size=24, face = "bold")),
                                 hellinger + labs(tag = "D") + theme(plot.tag = element_text(size=24, face = "bold")) ,
                                 ncol = 2, align = 'h', axis="tblr", nrow=2)
global_figure <- gridExtra::arrangeGrob(upper_part,
                                        boxplot_parameter_bivariate +
                                          theme(plot.tag = element_text(size=24, face = "bold")) + labs(tag = "E"),
                                        top="", nrow = 2, heights = c(1.5, 1), padding = unit(1, "line"))
ggsave("figs/bivariate/multivariate_unbalanced_overlapping_positive_correlated.png",
       global_figure, width = 20, height = 24, dpi = 75)

knitr::include_graphics("figs/bivariate/multivariate_unbalanced_overlapping_positive_correlated.png", dpi = 75)
```





\newpage

```{r compute-bivariate-overlapping-unbalanced-uncorrelated}
specific_bivariate_parameters_local_scores <- bivariate_parameters_local_scores_summary %>%
  dplyr::filter(ID=="B15") %>% select(-ID)
nrow_summary <- nrow(specific_bivariate_parameters_local_scores)
```

```{r multivariate-overlapping-unbalanced-uncorrelated-pdf, eval=knitr::is_latex_output()}
specific_bivariate_parameters_local_scores_kable <- specific_bivariate_parameters_local_scores %>%
  kbl(booktabs=T, caption = "MSE and Bias associated to scenario B15, in Table \\ref{tab:parameter-configuration-bivariate}   (unbalanced, overlapping and uncorrelated components).", escape=F,
      col.names = linebreak(c("Package", "Initialisation\nMethod",
                              "Global \n MSE $p$", "Global\nMSE $\\mu$", "Global\nMSE $\\sigma$",
                              "Global \n Bias $p$", "Global\nBias $\\mu$", "Global\nBias $\\sigma$"))) %>%
  kable_styling(latex_options=c("hold_position", "scale_down"), full_width = T) %>%
  row_spec(0,bold=T, align = "c") %>%
  column_spec(column = 1:8, latex_valign = "m") %>%
  column_spec(1, bold = T) %>% collapse_rows(columns = 1, valign = "middle")


for (metric_col in metric_colnames) {
  # define col vector
  col_vector <- rep("black", nrow_summary); metric_vector <- specific_bivariate_parameters_local_scores %>% pull(metric_col)
  col_vector[which(metric_vector==min(metric_vector))] <- "green" # set minimal values to green
  col_vector[which(metric_vector==max(metric_vector))] <- "red" # set maximal values to red
  # set colours
  specific_bivariate_parameters_local_scores_kable <- specific_bivariate_parameters_local_scores_kable %>%
    column_spec(match(metric_col, names(specific_bivariate_parameters_local_scores)),color=col_vector)
}
specific_bivariate_parameters_local_scores_kable
```

```{r multivariate-overlapping-unbalanced-uncorrelated-html, eval=knitr::is_html_output(), layout = "l-body-outset"}
specific_bivariate_parameters_local_scores_flex <- specific_bivariate_parameters_local_scores %>% flextable() %>%
  set_caption(caption = "MSE and Bias associated to scenario B15, in Table \\@ref(tab:parameter-configuration-bivariate) (unbalanced, overlapping and uncorrelated components.)", html_escape = F)

for (metric_col in metric_colnames) {
  specific_bivariate_parameters_local_scores_flex <- specific_bivariate_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body",
                     i = as.formula(paste0("~ `", metric_col, "`==min(`", metric_col, "`)")),color = "green")

  specific_bivariate_parameters_local_scores_flex <- specific_bivariate_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body",
                     i = as.formula(paste0("~ `", metric_col, "`==max(`", metric_col, "`)")),color = "red")
}

specific_bivariate_parameters_local_scores_flex %>%
  flextable::compose(j="global_mse_p", part = "header", value = as_paragraph("Global MSE \U0070")) %>%
  flextable::compose(j="global_mse_mu", part = "header", value = as_paragraph("Global MSE \U03BC")) %>%
  # flextable::compose(j="global_mse_ sigma", part = "header", value = as_paragraph("Global_mse_ ", as_equation("\\sigma"))) %>%
  flextable::compose(j="global_mse_sigma", part = "header", value = as_paragraph("Global MSE \U03C3")) %>%
  flextable::compose(j="global_bias_p", part = "header", value = as_paragraph("Global Bias \U0070")) %>%
  flextable::compose(j="global_bias_mu", part = "header", value = as_paragraph("Global Bias \U03BC")) %>%
  flextable::compose(j="global_bias_sigma", part = "header", value = as_paragraph("Global Bias \U03C3")) %>%
  set_header_labels(package= "Package", `initialisation method`="Initialisation Method") %>%
  theme_vanilla() %>% align(align = "center", part = "header") %>%  # centre align header
  merge_v(j = "Package") %>%   flextable::autofit()

```

\newpage

```{r multivariate-overlapping-unbalanced-uncorrelated, fig.cap=if (html_output) html_caption_unbalanced_uncorrelated_bivariate else pdf_caption_unbalanced_uncorrelated_bivariate, layout = "l-body-outset"}
true_theta <- bivariate_configuration %>% filter(ID=="B15") %>% pull(true_parameters) %>% magrittr::extract2(1)

# generate plots
bivariate_density <- plot_bivariate_normal_density_distribution(true_theta=true_theta, n = 2000)
computation_time <- plot_time_computations(bivariate_time_computation %>% filter(ID == "B15"))
ellipse <- plot_ellipses_bivariate(bivariate_parameters %>% filter(ID == "B15"), true_theta)
hellinger <- plot_Hellinger(formatted_bivariate_parameters %>% filter(ID == "B15"), true_theta) +
  theme(axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4))
boxplot_parameter_bivariate <- plot_boxplots_parameters(formatted_bivariate_parameters %>% filter(ID == "B15"), true_theta) +
  theme(axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4))


# concatenate plots together
upper_part <- cowplot::plot_grid(bivariate_density + labs(tag="A") + theme(plot.tag = element_text(size=24, face = "bold")),
                                 ellipse + labs(tag="B") + theme(plot.tag = element_text(size=24, face = "bold"), legend.title = element_blank()),
                                 computation_time + labs(tag="C") + theme(plot.tag = element_text(size=24, face = "bold")),
                                 hellinger + labs(tag = "D") + theme(plot.tag = element_text(size=24, face = "bold")) ,
                                 ncol = 2, align = 'h', axis="tblr", nrow=2)
global_figure <- gridExtra::arrangeGrob(upper_part,
                                        boxplot_parameter_bivariate +
                                          theme(plot.tag = element_text(size=24, face = "bold")) + labs(tag = "E"),
                                        top="", nrow = 2, heights = c(1.5, 1), padding = unit(1, "line"))
ggsave("figs/bivariate/multivariate_unbalanced_uncorrelated.png",
       global_figure, width = 20, height = 24, dpi = 75)

knitr::include_graphics("figs/bivariate/multivariate_unbalanced_uncorrelated.png", dpi = 75)
```



In contrast to the univariate setting ([Supplementary Figures and Tables in the univariate simulation]), the fastest packages are \pkg{bgmm}, \pkg{EMCluster}, \pkg{flexmix}, and \pkg{Rmixmod}, and the slowest ones \pkg{mclust}, \pkg{GMKMcharlie} and \pkg{mixtools}, independently from the difficulty of the simulation.


Finally, Figures \@ref(fig:general-balanced-overlapping-bivariate), \@ref(fig:general-balanced-well-separated-bivariate) and \@ref(fig:general-unbalanced-well-separated-bivariate) represent in a synthetic way less interesting scenarios benchmarked with to the left, the contour maps and to the right the corresponding Hellinger boxplots, with one scenario being illustrated per row.

```{r general-balanced-overlapping-bivariate, fig.cap=if (html_output) html_balanced_overlapping_general_bivariate else pdf_balanced_overlapping_general_bivariate, layout = "l-body-outset"}
true_theta_v1 <- bivariate_configuration %>% filter(ID=="B1") %>% pull(true_parameters) %>% magrittr::extract2(1)
ellipse_v1 <- plot_ellipses_bivariate(bivariate_parameters %>% filter(ID=="B1"), true_theta_v1) +
  theme(legend.title = element_blank())
hellinger_v1 <- plot_Hellinger(formatted_bivariate_parameters %>% filter(ID=="B1"), true_theta_v1)

true_theta_v2 <- bivariate_configuration %>% filter(ID=="B2") %>% pull(true_parameters) %>% magrittr::extract2(1)
ellipse_v2 <- plot_ellipses_bivariate(bivariate_parameters %>% filter(ID=="B2"), true_theta_v2) +
  theme(legend.title = element_blank())
hellinger_v2 <- plot_Hellinger(formatted_bivariate_parameters %>% filter(ID=="B2"), true_theta_v2)

true_theta_v3 <- bivariate_configuration %>% filter(ID=="B5") %>% pull(true_parameters) %>% magrittr::extract2(1)
ellipse_v3 <- plot_ellipses_bivariate(bivariate_parameters %>% filter(ID=="B5"), true_theta_v3) +
  theme(legend.title = element_blank())
hellinger_v3 <- plot_Hellinger(formatted_bivariate_parameters %>% filter(ID=="B5"), true_theta_v3)

# concatenate all plots
global_figure <- gridExtra::arrangeGrob(grobs=list(ellipse_v1 + labs(tag="A") + theme(plot.tag = element_text(size=24, face = "bold")),
                                    hellinger_v1 + labs(tag = "B") + theme(plot.tag = element_text(size=24, face = "bold")),
                                    ellipse_v2 + labs(tag="C") + theme(plot.tag = element_text(size=24, face = "bold")),
                                    hellinger_v2 + labs(tag = "D") + theme(plot.tag = element_text(size=24, face = "bold")),
                                    ellipse_v3 + labs(tag="E") + theme(plot.tag = element_text(size=24, face = "bold")),
                                    hellinger_v3 + labs(tag = "F") + theme(plot.tag = element_text(size=24, face = "bold"))),
                                    ncol = 2, nrow=3, widths = c(1.5, 1))
ggsave("figs/bivariate/multivariate_balanced_overlapping.png",
       global_figure, width = 20, height = 24,dpi = 75)

knitr::include_graphics("figs/bivariate/multivariate_balanced_overlapping.png", dpi = 75)
```

\newpage

```{r general-balanced-well-separated-bivariate, fig.cap=if (html_output) html_balanced_well_separated_general_bivariate else pdf_balanced_well_separated_general_bivariate, layout = "l-body-outset"}
true_theta_v1 <- bivariate_configuration %>% filter(ID=="B6") %>% pull(true_parameters) %>% magrittr::extract2(1)
ellipse_v1 <- plot_ellipses_bivariate(bivariate_parameters %>% filter(ID=="B6"), true_theta_v1) +
  theme(legend.title = element_blank())
hellinger_v1 <- plot_Hellinger(formatted_bivariate_parameters %>% filter(ID=="B6"), true_theta_v1)

true_theta_v2 <- bivariate_configuration %>% filter(ID=="B7") %>% pull(true_parameters) %>% magrittr::extract2(1)
ellipse_v2 <- plot_ellipses_bivariate(bivariate_parameters %>% filter(ID=="B7"), true_theta_v2) +
  theme(legend.title = element_blank())
hellinger_v2 <- plot_Hellinger(formatted_bivariate_parameters %>% filter(ID=="B7"), true_theta_v2)

true_theta_v3 <- bivariate_configuration %>% filter(ID=="B10") %>% pull(true_parameters) %>% magrittr::extract2(1)
ellipse_v3 <- plot_ellipses_bivariate(bivariate_parameters %>% filter(ID=="B10"), true_theta_v3) +
  theme(legend.title = element_blank())
hellinger_v3 <- plot_Hellinger(formatted_bivariate_parameters %>% filter(ID=="B10"), true_theta_v3)

# concatenate all plots
global_figure <- gridExtra::arrangeGrob(grobs=list(ellipse_v1 + labs(tag="A") + theme(plot.tag = element_text(size=24, face = "bold")),
                                    hellinger_v1 + labs(tag = "B") + theme(plot.tag = element_text(size=24, face = "bold")),
                                    ellipse_v2 + labs(tag="C") + theme(plot.tag = element_text(size=24, face = "bold")),
                                    hellinger_v2 + labs(tag = "D") + theme(plot.tag = element_text(size=24, face = "bold")),
                                    ellipse_v3 + labs(tag="E") + theme(plot.tag = element_text(size=24, face = "bold")),
                                    hellinger_v3 + labs(tag = "F") + theme(plot.tag = element_text(size=24, face = "bold"))),
                                    ncol = 2, nrow=3, widths = c(1.5, 1))
ggsave("figs/bivariate/multivariate_balanced_well_separated.png",
       global_figure, width = 20, height = 24,dpi = 75)

knitr::include_graphics("figs/bivariate/multivariate_balanced_well_separated.png", dpi = 75)
```

\newpage


```{r general-unbalanced-well-separated-bivariate, fig.cap=if (html_output) html_unbalanced_well_separated_general_bivariate else pdf_unbalanced_well_separated_general_bivariate, layout = "l-body-outset"}
true_theta_v1 <- bivariate_configuration %>% filter(ID=="B16") %>% pull(true_parameters) %>% magrittr::extract2(1)
ellipse_v1 <- plot_ellipses_bivariate(bivariate_parameters %>% filter(ID=="B16"), true_theta_v1) +
  theme(legend.title = element_blank())
hellinger_v1 <- plot_Hellinger(formatted_bivariate_parameters %>% filter(ID=="B16"), true_theta_v1)

true_theta_v2 <- bivariate_configuration %>% filter(ID=="B17") %>% pull(true_parameters) %>% magrittr::extract2(1)
ellipse_v2 <- plot_ellipses_bivariate(bivariate_parameters %>% filter(ID=="B17"), true_theta_v2) +
  theme(legend.title = element_blank())
hellinger_v2 <- plot_Hellinger(formatted_bivariate_parameters %>% filter(ID=="B17"), true_theta_v2)

true_theta_v3 <- bivariate_configuration %>% filter(ID=="B20") %>% pull(true_parameters) %>% magrittr::extract2(1)
ellipse_v3 <- plot_ellipses_bivariate(bivariate_parameters %>% filter(ID=="B20"), true_theta_v3) +
  theme(legend.title = element_blank())
hellinger_v3 <- plot_Hellinger(formatted_bivariate_parameters %>% filter(ID=="B20"), true_theta_v3)

# concatenate all plots
global_figure <- gridExtra::arrangeGrob(grobs=list(ellipse_v1 + labs(tag="A") + theme(plot.tag = element_text(size=24, face = "bold")),
                                    hellinger_v1 + labs(tag = "B") + theme(plot.tag = element_text(size=24, face = "bold")),
                                    ellipse_v2 + labs(tag="C") + theme(plot.tag = element_text(size=24, face = "bold")),
                                    hellinger_v2 + labs(tag = "D") + theme(plot.tag = element_text(size=24, face = "bold")),
                                    ellipse_v3 + labs(tag="E") + theme(plot.tag = element_text(size=24, face = "bold")),
                                    hellinger_v3 + labs(tag = "F") + theme(plot.tag = element_text(size=24, face = "bold"))),
                                    ncol = 2, nrow=3, widths = c(1.5, 1))
ggsave("figs/bivariate/multivariate_unbalanced_well_separated.png",
       global_figure, width = 20, height = 24,dpi = 75)

knitr::include_graphics("figs/bivariate/multivariate_unbalanced_well_separated.png", dpi = 75)
```

```{r heatmap-all-correlation-plots-bivariate, fig.cap= if (html_output) html_global_heatmap_bivariate else pdf_global_heatmap_bivariate, layout = "l-body-outset"}
bivariate_parameters <- readRDS("./tables/bivariate/bivariate_distributions.rds")%>%
  dplyr::group_by(ID, package, initialisation_method) %>%
  dplyr::mutate(N.bootstrap=dplyr::row_number()) %>% dplyr::ungroup()
multivariate_correlation_heatmaps <- plot_correlation_Heatmap(bivariate_parameters %>% filter(ID=="B11"))
multivariate_correlation_heatmaps <- gridExtra::arrangeGrob(grobs=multivariate_correlation_heatmaps %>%
                                                              purrr::map(~grid::grid.grabExpr(ComplexHeatmap::draw(.x))),
                                                            top="", ncol = 2)
ggsave("./figs/bivariate/heatmap_bivariate.png", multivariate_correlation_heatmaps, width = 10.5, height = 13.5,dpi = 75)
knitr::include_graphics("./figs/bivariate/heatmap_bivariate.png", dpi = 75)
```






## Supplementary Figures and Tables in the HD simulation {.appendix}

Table below (\@ref(tab:parameter-configuration-HD)) lists the complete set of parameters used to simulate Gaussian distributions in the high dimensional benchmark:

```{r parameter-configuration-HD, layout = "l-page"}
HD_configuration_formatted <- HD_configuration  %>%
  rename(Entropy=entropy, Spherical=is_diagonal, `Number of observations`=nobservations) %>%
  mutate(Proportions=purrr::map_chr(true_parameters, ~paste(round(.x$p, digits = 2), collapse = " / ")),
         Spherical=dplyr::if_else(Spherical, "\\greentick",   "\\redcross"), ID=paste0("HD", ID)) %>%
  select(-c("true_parameters", "Entropy")) %>%
  dplyr::relocate(Spherical, .after = Proportions)


HD_configuration_formatted %>%
  kbl(booktabs=T, caption = "The 16 parameter configurations tested to generate the samples in a high dimensional context. The first digit of each ID index refers
      to an unique parameter configuration (identified by its level of overlap, entropy and topological structure, either circular or ellipsoidal,
      of the covariance matrix, while the lowercase letter depicts the number of observations, a) with $n=200$ and b) with $n=2000$.",
      escape=F, align = "c",
      col.names = linebreak(c("ID", "OVL", "Number of \nobservations", "Proportions", "Spherical"))) %>%
  kable_styling(latex_options=c("hold_position", "scale_down")) %>%
  row_spec(0,bold=T) %>% row_spec(1:16, hline_after = T) %>%
  kable_styling(bootstrap_options = c("hover"))
```

\newpage


```{r prepare-legend-HD}
html_output <- knitr::is_html_output()  # change to TRUE for larger figures

html_caption_HD_separated_unbalanced_ellipsoidal <- paste("Results of scenario HD4a) in Table \\@ref(tab:parameter-configuration-HD) (unbalanced, overlapping and negative correlated components), organised as such:

 - The panel A displays the bivariate factorial projection of a random sample drawn from the 10-dimensional multivariate Gaussian distribution parametrised by Table \\@ref(tab:parameter-configuration-HD). Each component is associated to a specific color, a centroid whose coordinates are given by the mean components' elements in the bivariate projected space and a $95\\%$ confidence ellipse. Arrows represent the correlation circle of the dimensional variables.
 Both panels were displayed respectively using functions",  downlit::autolink('factoextra::fviz_eig()') , "and", downlit::autolink('factoextra::fviz_pca_biplot()'), "while the underlying computations proceed from the principal component analysis performed by", downlit::autolink('ade4::dudi.pca()'), "package preceded by standard scaling of the sampling dataset.

- The panel B pictures the *parallel distribution plots* from a random sampling of $n=100$ observations, generated using the",
downlit::autolink('GGally::ggparcoord()'), "function, and representing the coordinates of each simulated data point in 10 dimensions.

- The running times are displayed in Panel C with the *k*-means initialisation. The number of observations (x-axis) and the running time (y-axis) is in $\\log(10)$ scale.

- The distributions of the Hellinger distances are computed for each component in Panel D, each initialisation method and each package with respect to the true Gaussian distribution expected for each component.

- In panel E we represent the boxplots associated with the distribution of some of the estimates. Since it was impractical to represent all of the $k + kD + k\\frac{D \\times (D+1)}{2}$ with $k=2$ and $D=10$ parameters, we only represent the first component's mean, two first components' variances and their covariance term.")


pdf_caption_HD_separated_unbalanced_ellipsoidal <- "Results of scenario HD4a) in Table \\ref{tab:parameter-configuration-HD} (unbalanced, overlapping and negative correlated components), organised as such:
The panel A displays the bivariate factorial projection of a random sample drawn from the 10-dimensional multivariate Gaussian distribution parametrised by Table \\ref{tab:parameter-configuration-HD}. Each component is associated to a specific color, a centroid whose coordinates are given by the mean components' elements in the bivariate projected space and a $95\\%$ confidence ellipse. Arrows represent the correlation circle of the dimensional variables.
 Both panels were displayed respectively using functions \\code{factoextra::fviz\\_eig} and \\code{factoextra::fviz\\_pca\\_biplot} while the underlying computations proceed from the principal component analysis performed by \\code{ade4::dudi.pca} preceded by standard scaling of the sampling dataset.
The panel B pictures the \\textit{parallel distribution plots} from a random sampling of $n=100$ observations, generated using \\code{GGally::ggparcoord}, and representing the coordinates of each simulated data point in 10 dimensions.
The running times are displayed in Panel C with the \\textit{k}-means initialisation. The number of observations (x-axis) and the running time (y-axis) is in $\\log(10)$ scale.
The distributions of the Hellinger distances are computed for each component in Panel D, each initialisation method and each package with respect to the true Gaussian distribution expected for each component.
In panel E we represent the boxplots associated with the distribution of some of the estimates. Since it was impractical to represent all of the $k + kD + k\\frac{D \\times (D+1)}{2}$ with $k=2$ and $D=10$ parameters, we only represent the first component's mean, two first components' variances and their covariance term."



html_caption_HD_overlapping_balanced_ellipsoidal<- "Results of scenario HD7a) in Table \\@ref(tab:parameter-configuration-HD)  (balanced and overlapping components, with full covariance structure), with the same layout as Figure \\@ref(fig:HD-separated-unbalanced-ellipsoidal-plot)."
pdf_caption_HD_overlapping_balanced_ellipsoidal <- "Results of scenario HD7a) in Table \\ref{tab:parameter-configuration-HD} (balanced and overlapping components, with full covariance structure), with the same layout as Figure \\ref{fig:HD-separated-unbalanced-ellipsoidal-plot}."


html_caption_HD_overlapping_spherical<- "We gathered on the same plot two multivariate benchmark scenarios, in which we consider a strictly spherical structure of the covariance matrix:

- We represent in Panel A and B, respectively the bivariate projection and parallel distribution plot, associated to scenario HD5a) in Table \\@ref(tab:parameter-configuration-HD)  (balanced and overlapping components, with spherical covariance structure).

- In Panel C, we display the boxplots associated to scenario HD5a), computing them similarly as in Panel E of Figure \\@ref(fig:HD-separated-unbalanced-ellipsoidal-plot).

- In Panel D, we display the boxplots associated to scenario HD6a) (unbalanced and overlapping components, with spherical covariance structure)."

pdf_caption_HD_overlapping_spherical <- "We gathered on the same plot two multivariate benchmark scenarios, in which we consider a strictly spherical structure of the covariance matrix:
We represent in Panel A and B, respectively the bivariate projection and parallel distribution plot, associated to scenario HD5a) in Table \\ref{tab:parameter-configuration-HD}  (balanced and overlapping components, with spherical covariance structure).
In Panel C, we display the boxplots associated to scenario HD5a), computing them similarly as in Panel E of Figure \\ref{fig:HD-separated-unbalanced-ellipsoidal-plot}.
In Panel D, we display the boxplots associated to scenario HD6a) (unbalanced and overlapping components, with spherical covariance structure)."



html_caption_HD_impact_num_observations <- "Overview of scenarios HD1 a) and b) and HD8 a) and b) in Table \\@ref(tab:parameter-configuration-HD) comparing the performance of the algorithms in respectively the easiest and most complex scenario. The left-hand column shows box plots of the estimated parameters from simulations with n=200 observations on the left and n=2000 observations on the right."
pdf_caption_HD_impact_num_observations <- "Overview of scenarios HD1 a) and b) and HD8 a) and b) in Table \\ref{tab:parameter-configuration-HD}comparing the performance of the algorithms in respectively the easiest and most complex scenario. The left-hand column shows box plots of the estimated parameters from simulations with $n=200$ observations on the left and $n=2000$ observations on the right."



html_global_heatmap_HD <- "Correlation heatmaps of the estimated parameters in the high dimensional (HD) setting extended to the three initialisation methods benchmarked (respectively hc, *k*-means and rebmix) in the most discriminating scenario HD8a), using the same process described in Figure (2)."
pdf_global_heatmap_HD <- "Correlation heatmaps of the estimated parameters in the high dimensional (HD) setting extended to the three initialisation methods benchmarked (respectively hc, \\textit{k}-means and rebmix) in the most discriminating scenario HD8a), using the same process described in Figure (2)."
```

```{r compute-HD-separated-unbalanced-ellipsoidal}
specific_HD_parameters_local_scores <- HD_parameters_local_scores_summary %>%
  dplyr::filter(ID=="4a") %>% select(-ID)
nrow_summary <- nrow(specific_HD_parameters_local_scores)
```

```{r HD-separated-unbalanced-ellipsoidal-pdf, eval=knitr::is_latex_output()}
specific_HD_parameters_local_scores_kable <- specific_HD_parameters_local_scores %>%
  kbl(booktabs=T, caption = "MSE and Bias associated to scenario HD4a, in Table
      \\ref{tab:parameter-configuration-HD} (unbalanced, separated and ellipsoidal components).", escape=F,
      col.names = linebreak(c("Package", "Initialisation\nMethod",
                              "Global \n MSE $p$", "Global\nMSE $\\mu$", "Global\nMSE $\\sigma$",
                              "Global \n Bias $p$", "Global\nBias $\\mu$", "Global\nBias $\\sigma$", "$\\%$ Success"))) %>%
  kable_styling(latex_options=c("hold_position", "scale_down"), full_width = T) %>%
  row_spec(0,bold=T, align = "c") %>%
  column_spec(column = 1:9, latex_valign = "m") %>%
  column_spec(1, bold = T) %>% collapse_rows(columns = 1, valign = "middle", latex_hline = "major")

for (metric_col in c(metric_colnames, "Success")) {
  averaged_metric <- specific_HD_parameters_local_scores %>% pull(metric_col)
  min_index <- which(averaged_metric == min(averaged_metric))
  max_index <- which(averaged_metric == max(averaged_metric))
  if (metric_col =="Success")  seqinr::swap(min_index, max_index)

  col_vector <- rep("black", nrow(specific_HD_parameters_local_scores))
  col_vector[min_index] <- "green"; col_vector[max_index] <- "red"
  specific_HD_parameters_local_scores_kable <- specific_HD_parameters_local_scores_kable %>%
        column_spec(match(metric_col, names(specific_HD_parameters_local_scores)),color=col_vector)
}

specific_HD_parameters_local_scores_kable %>%
  row_spec(12,  hline_after = T)
  # column_spec(9, color="white",
  #             background = spec_color( specific_HD_parameters_local_scores$Success, alpha=0.8, end = 0.7, option="A")) %>%

```

```{r HD-separated-unbalanced-ellipsoidal-html, eval=knitr::is_html_output(), layout = "l-body-outset"}
specific_HD_parameters_local_scores_flex <- specific_HD_parameters_local_scores %>% flextable() %>%
  set_caption(caption = "MSE and Bias associated to scenario HD4a, in Table
      \\@ref(tab:parameter-configuration-HD) (unbalanced, separated and ellipsoidal components).", html_escape = F)

for (metric_col in c(metric_colnames, "Success")) {
  averaged_metric <- purrr::map_dbl(specific_HD_parameters_local_scores %>% pull(metric_col) %>%
                                      stringr::str_split(pattern = " // "),
                                    ~mean(as.numeric(.x)))
  min_index <- which(averaged_metric == min(averaged_metric))
  max_index <- which(averaged_metric == max(averaged_metric))
  if (metric_col =="Success")  seqinr::swap(min_index, max_index)

  specific_HD_parameters_local_scores_flex <- specific_HD_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body", i = min_index,color = "green")
  specific_HD_parameters_local_scores_flex <- specific_HD_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body", i = max_index,color = "red")
}

specific_HD_parameters_local_scores_flex %>%
  flextable::compose(j="global_mse_p", part = "header", value = as_paragraph("Global MSE \U0070")) %>%
  flextable::compose(j="global_mse_mu", part = "header", value = as_paragraph("Global MSE \U03BC")) %>%
  # flextable::compose(j="global_mse_ sigma", part = "header", value = as_paragraph("Global_mse_ ", as_equation("\\sigma"))) %>%
  flextable::compose(j="global_mse_sigma", part = "header", value = as_paragraph("Global MSE \U03C3")) %>%
  flextable::compose(j="global_bias_p", part = "header", value = as_paragraph("Global Bias \U0070")) %>%
  flextable::compose(j="global_bias_mu", part = "header", value = as_paragraph("Global Bias \U03BC")) %>%
  flextable::compose(j="global_bias_sigma", part = "header", value = as_paragraph("Global Bias \U03C3")) %>%
  set_header_labels(package= "Package", `initialisation method`="Initialisation Method") %>%
  theme_vanilla() %>% align(align = "center", part = "header") %>%  # centre align header
  merge_v(j = "Package") %>%  flextable::hline(i = 12, border = officer::fp_border(width = 2)) %>%
  flextable::autofit()

```

\newpage


```{r HD-separated-unbalanced-ellipsoidal-plot, fig.cap=if (html_output) html_caption_HD_separated_unbalanced_ellipsoidal else pdf_caption_HD_separated_unbalanced_ellipsoidal, out.width="60%", layout = "l-body-outset"}
true_theta <- HD_configuration %>% filter(ID=="4a") %>% pull(true_parameters) %>% magrittr::extract2(1)

# generate plots
HD_density <- plot_HD_density_distribution(true_theta=true_theta, n = 200, k = 2, ade_plot = F)
computation_time <- plot_time_computations(HD_time_computation %>% filter(ID == "4"))
hellinger <- plot_Hellinger(formatted_HD_distributions %>% filter(ID == "4a"), true_theta) +
  theme(axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4))
boxplot_parameter_HD <- plot_boxplots_parameters(formatted_HD_distributions %>% filter(ID == "4a"),
                                                 num_col = 2, true_theta = true_theta,
                                                 match_symbol = "^p[[:digit:]]+$|mu_var1_|sd_var1_var1|sd_var2_var1|sd_var2_var2") +
  theme(axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4))


# concatenate plots together
upper_part <- cowplot::plot_grid(computation_time + labs(tag="C") + theme(plot.tag = element_text(size=24, face = "bold")),
                                 hellinger + labs(tag = "D") + theme(plot.tag = element_text(size=24, face = "bold")) ,
                                 ncol = 2, align = 'hv', axis="tblr")
global_figure <- gridExtra::arrangeGrob(HD_density, upper_part,
                                        boxplot_parameter_HD +
                                          theme(plot.tag = element_text(size=24, face = "bold")) + labs(tag = "E"),
                                        top="", nrow = 3, heights = c(2, 1.5, 2), padding = unit(1, "line"))
ggsave("figs/HD/HD-separated-unbalanced-ellipsoidal.png",global_figure, width = 24, height = 32, dpi = 75)

knitr::include_graphics("figs/HD/HD-separated-unbalanced-ellipsoidal.png", dpi = 75)
```

\newpage


```{r compute-HD-overlapping-balanced-ellipsoidal}
specific_HD_parameters_local_scores <- HD_parameters_local_scores_summary %>%
  dplyr::filter(ID=="7a") %>% select(-ID)
nrow_summary <- nrow(specific_HD_parameters_local_scores)
```


```{r HD-overlapping-balanced-ellipsoidal-pdf, eval=knitr::is_latex_output()}
specific_HD_parameters_local_scores_kable <- specific_HD_parameters_local_scores %>%
  kbl(booktabs=T, caption = "MSE and Bias associated to scenario HD7a, in Table
      \\ref{tab:parameter-configuration-HD} (balanced, overlapping and ellipsoidal components).", escape=F,
      col.names = linebreak(c("Package", "Initialisation\nMethod",
                              "Global \n MSE $p$", "Global\nMSE $\\mu$", "Global\nMSE $\\sigma$",
                              "Global \n Bias $p$", "Global\nBias $\\mu$", "Global\nBias $\\sigma$", "$\\%$ Success"))) %>%
  kable_styling(latex_options=c("hold_position", "scale_down"), full_width = T) %>%
  row_spec(0,bold=T, align = "c") %>%
  column_spec(column = 1:9, latex_valign = "m") %>%
  column_spec(1, bold = T) %>% collapse_rows(columns = 1, valign = "middle", latex_hline = "major")

for (metric_col in c(metric_colnames, "Success")) {
  averaged_metric <- specific_HD_parameters_local_scores %>% pull(metric_col)
  min_index <- which(averaged_metric == min(averaged_metric))
  max_index <- which(averaged_metric == max(averaged_metric))
  if (metric_col =="Success")  seqinr::swap(min_index, max_index)

  col_vector <- rep("black", nrow(specific_HD_parameters_local_scores))
  col_vector[min_index] <- "green"; col_vector[max_index] <- "red"
  specific_HD_parameters_local_scores_kable <- specific_HD_parameters_local_scores_kable %>%
        column_spec(match(metric_col, names(specific_HD_parameters_local_scores)),color=col_vector)
}

specific_HD_parameters_local_scores_kable %>%
  # column_spec(9, color="white",
  #             background = spec_color( specific_HD_parameters_local_scores$Success, alpha=0.8, end = 0.7, option="A")) %>%
  row_spec(12,  hline_after = T)
```

```{r HD-overlapping-balanced-ellipsoidal-html, eval=knitr::is_html_output(), layout = "l-body-outset"}
specific_HD_parameters_local_scores_flex <- specific_HD_parameters_local_scores %>% flextable() %>%
  set_caption(caption = "MSE and Bias associated to scenario HD7a, in Table
      \\@ref(tab:parameter-configuration-HD) (balanced, overlapping and ellipsoidal components).", html_escape = F)

for (metric_col in c(metric_colnames, "Success")) {
  averaged_metric <- purrr::map_dbl(specific_HD_parameters_local_scores %>% pull(metric_col) %>%
                                      stringr::str_split(pattern = " // "),
                                    ~mean(as.numeric(.x)))
  min_index <- which(averaged_metric == min(averaged_metric))
  max_index <- which(averaged_metric == max(averaged_metric))
  if (metric_col =="Success")  seqinr::swap(min_index, max_index)

  specific_HD_parameters_local_scores_flex <- specific_HD_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body", i = min_index,color = "green")
  specific_HD_parameters_local_scores_flex <- specific_HD_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body", i = max_index,color = "red")
}


specific_HD_parameters_local_scores_flex %>%
  flextable::compose(j="global_mse_p", part = "header", value = as_paragraph("Global MSE \U0070")) %>%
  flextable::compose(j="global_mse_mu", part = "header", value = as_paragraph("Global MSE \U03BC")) %>%
  # flextable::compose(j="global_mse_ sigma", part = "header", value = as_paragraph("Global_mse_ ", as_equation("\\sigma"))) %>%
  flextable::compose(j="global_mse_sigma", part = "header", value = as_paragraph("Global MSE \U03C3")) %>%
  flextable::compose(j="global_bias_p", part = "header", value = as_paragraph("Global Bias \U0070")) %>%
  flextable::compose(j="global_bias_mu", part = "header", value = as_paragraph("Global Bias \U03BC")) %>%
  flextable::compose(j="global_bias_sigma", part = "header", value = as_paragraph("Global Bias \U03C3")) %>%
  set_header_labels(package= "Package", `initialisation method`="Initialisation Method") %>%
  theme_vanilla() %>% align(align = "center", part = "header") %>%  # centre align header
  merge_v(j = "Package") %>%
  flextable::hline(i = 12, border = officer::fp_border(width = 2)) %>%
  flextable::autofit()

```

\newpage

```{r HD-overlapping-balanced-ellipsoidal-plot, fig.cap=if (html_output) html_caption_HD_overlapping_balanced_ellipsoidal else pdf_caption_HD_overlapping_balanced_ellipsoidal, out.width="80%", layout = "l-body-outset"}
true_theta <- HD_configuration %>% filter(ID=="7a") %>% pull(true_parameters) %>% magrittr::extract2(1)

# generate plots
HD_density <- plot_HD_density_distribution(true_theta=true_theta, n = 200, k = 2, ade_plot = F)
computation_time <- plot_time_computations(HD_time_computation %>% filter(ID == "4"))
hellinger <- plot_Hellinger(formatted_HD_distributions %>% filter(ID == "7a"), true_theta) +
  theme(axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4))
boxplot_parameter_HD <- plot_boxplots_parameters(formatted_HD_distributions %>% filter(ID == "7a"),
                                                 num_col = 2, true_theta = true_theta,
                                                 match_symbol = "^p[[:digit:]]+$|mu_var1_|sd_var1_var1|sd_var2_var1|sd_var2_var2") +
  theme(axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4))


# concatenate plots together
upper_part <- cowplot::plot_grid(computation_time + labs(tag="C") + theme(plot.tag = element_text(size=24, face = "bold")),
                                 hellinger + labs(tag = "D") + theme(plot.tag = element_text(size=24, face = "bold")) ,
                                 ncol = 2, align = 'hv', axis="tblr")
global_figure <- gridExtra::arrangeGrob(HD_density, upper_part,
                                        boxplot_parameter_HD +
                                          theme(plot.tag = element_text(size=24, face = "bold")) + labs(tag = "E"),
                                        top="", nrow = 3, heights = c(2, 1.5, 2), padding = unit(1, "line"))
ggsave("figs/HD/HD-overlapping-balanced-ellipsoidal.png",global_figure, width = 24, height = 32, dpi = 75)

knitr::include_graphics("figs/HD/HD-overlapping-balanced-ellipsoidal.png", dpi = 75)
```

\newpage
```{r compute-HD-overlapping-spherical}
HD_parameters_local_scores_balanced <- HD_parameters_local_scores_summary %>%
  dplyr::filter(ID=="5a") %>% select(-ID)
HD_parameters_local_scores_unbalanced <- HD_parameters_local_scores_summary %>%
  dplyr::filter(ID=="6a") %>% select(-ID)

specific_HD_parameters_local_scores <- HD_parameters_local_scores_balanced  %>%
  dplyr::mutate(global_mse_p=paste(global_mse_p,HD_parameters_local_scores_unbalanced$global_mse_p, sep=" // "),
                global_mse_mu=paste(global_mse_mu,HD_parameters_local_scores_unbalanced$global_mse_mu, sep=" // "),
                global_mse_sigma=paste(global_mse_sigma,HD_parameters_local_scores_unbalanced$global_mse_sigma, sep=" // "),
                global_bias_p=paste(global_bias_p,HD_parameters_local_scores_unbalanced$global_bias_p, sep=" // "),
                global_bias_mu=paste(global_bias_mu,HD_parameters_local_scores_unbalanced$global_bias_mu, sep=" // "),
                global_bias_sigma=paste(global_bias_sigma,HD_parameters_local_scores_unbalanced$global_bias_sigma, sep=" // "),
                Success=paste(Success,HD_parameters_local_scores_unbalanced$Success, sep=" // "))
```

```{r HD-overlapping-spherical-pdf, eval=knitr::is_latex_output()}
specific_HD_parameters_local_scores_kable <- specific_HD_parameters_local_scores %>%
  kbl(booktabs=T, caption = "MSE and Bias associated to scenarios HD5a) and HD6a), in Table
\\ref{tab:parameter-configuration-HD} (overlapping and spherical-distributed components). We delimite each scenario by doubled backslashes with respectively balanced and unbalanced clusters.", escape=F,
      col.names = linebreak(c("Package", "Initialisation\nMethod",
                              "Global \n MSE $p$", "Global\nMSE $\\mu$", "Global\nMSE $\\sigma$",
                              "Global \n Bias $p$", "Global\nBias $\\mu$", "Global\nBias $\\sigma$", "$\\%$ Success"))) %>%
  kable_styling(latex_options=c("hold_position", "scale_down"), full_width = T) %>%
  row_spec(0,bold=T, align = "c") %>%
  column_spec(column = 1:9, latex_valign = "m") %>%
  column_spec(1, bold = T) %>% collapse_rows(columns = 1, valign = "middle", latex_hline = "major")

for (metric_col in c(metric_colnames, "Success")) {
  averaged_metric <- purrr::map_dbl(specific_HD_parameters_local_scores %>% pull(metric_col) %>%
                                      stringr::str_split(pattern = " // "),
                                    ~mean(as.numeric(.x)))
  min_index <- which(averaged_metric == min(averaged_metric))
  max_index <- which(averaged_metric == max(averaged_metric))
  if (metric_col =="Success")  seqinr::swap(min_index, max_index)

  col_vector <- rep("black", nrow(specific_HD_parameters_local_scores))
  col_vector[min_index] <- "green"; col_vector[max_index] <- "red"
  specific_HD_parameters_local_scores_kable <- specific_HD_parameters_local_scores_kable %>%
        column_spec(match(metric_col, names(specific_HD_parameters_local_scores)),color=col_vector)
}

specific_HD_parameters_local_scores_kable %>%
  row_spec(12,  hline_after = T)
```

```{r HD-overlapping-spherical-html, eval=knitr::is_html_output(), layout = "l-body-outset"}
specific_HD_parameters_local_scores_flex <- specific_HD_parameters_local_scores %>% flextable() %>%
  set_caption(caption = "MSE and Bias associated to scenarios HD5a) and HD6a), in Table
\\@ref(tab:parameter-configuration-HD) (overlapping and spherical-distributed components). We delimite each scenario by doubled backslashes with respectively balanced and unbalanced clusters.", html_escape = F)

for (metric_col in c(metric_colnames, "Success")) {
  averaged_metric <- purrr::map_dbl(specific_HD_parameters_local_scores %>% pull(metric_col) %>%
                                      stringr::str_split(pattern = " // "),
                                    ~mean(as.numeric(.x)))
  min_index <- which(averaged_metric == min(averaged_metric))
  max_index <- which(averaged_metric == max(averaged_metric))
  if (metric_col =="Success")  seqinr::swap(min_index, max_index)

  specific_HD_parameters_local_scores_flex <- specific_HD_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body", i = min_index,color = "green")
  specific_HD_parameters_local_scores_flex <- specific_HD_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body", i = max_index,color = "red")
}


specific_HD_parameters_local_scores_flex %>%
  flextable::compose(j="global_mse_p", part = "header", value = as_paragraph("Global MSE \U0070")) %>%
  flextable::compose(j="global_mse_mu", part = "header", value = as_paragraph("Global MSE \U03BC")) %>%
  flextable::compose(j="global_mse_sigma", part = "header", value = as_paragraph("Global MSE \U03C3")) %>%
  flextable::compose(j="global_bias_p", part = "header", value = as_paragraph("Global Bias \U0070")) %>%
  flextable::compose(j="global_bias_mu", part = "header", value = as_paragraph("Global Bias \U03BC")) %>%
  flextable::compose(j="global_bias_sigma", part = "header", value = as_paragraph("Global Bias \U03C3")) %>%
  set_header_labels(package= "Package", `initialisation method`="Initialisation Method") %>%
  theme_vanilla() %>% align(align = "center", part = "header") %>%  # centre align header
  merge_v(j = "Package") %>%  flextable::hline(i = 12, border = officer::fp_border(width = 2)) %>%
  flextable::autofit()

```





```{r compute-HD-overlapping-spherical-offterms}
# modify the computation to set apart diagonal and off-diagonal terms
diagonal_terms <- stringr::str_subset(colnames(HD_parameters_local_scores),
                                     "^sd_(var[[:digit:]]+)_\\1_comp[[:digit:]]+$")
off_terms <- setdiff(stringr::str_subset(colnames(HD_parameters_local_scores), "^sd"), diagonal_terms)

HD_parameters_offterms_spherical <- HD_parameters_local_scores %>% dplyr::filter(scores=="mse") %>%
  rowwise() %>% summarise(ID, initialisation_method, package,
                          global_mse_diagonal=sum(dplyr::c_across(dplyr::all_of(diagonal_terms))),
                          global_mse_off=sum(dplyr::c_across(dplyr::all_of(off_terms)))) %>%
  dplyr::inner_join(HD_parameters_local_scores %>% dplyr::filter(scores=="bias") %>%
                      rowwise() %>% summarise(ID, initialisation_method, package,
                                              global_bias_diagonal=sum(abs(dplyr::c_across(dplyr::all_of(diagonal_terms)))),
                                              global_bias_off=sum(abs(dplyr::c_across(dplyr::all_of(off_terms))))),
                    by = join_by(ID, initialisation_method, package)) %>%
  dplyr::filter(if_all(everything(), ~ !is.na(.))) %>% # remove columns with NULL values
  dplyr::arrange(ID, package, initialisation_method) %>%
  dplyr::relocate(initialisation_method, .after = package) %>%
  filter(initialisation_method!="rebmix") %>%
  dplyr::rename(Package=package, `Initialisation Method`=initialisation_method) %>%
  mutate(across(where(is.numeric), ~format(.x, digits=2, nsmall=0) %>% as.numeric))


# compute it precisely for the the scenario we are interested in
HD_parameters_local_scores_balanced <- HD_parameters_offterms_spherical %>%
  dplyr::filter(ID=="5a") %>% select(-ID)
HD_parameters_local_scores_unbalanced <- HD_parameters_offterms_spherical %>%
  dplyr::filter(ID=="6a") %>% select(-ID)

spherical_HD_parameters_local_scores <- HD_parameters_local_scores_balanced  %>%
  dplyr::mutate(global_mse_diagonal=paste(global_mse_diagonal,
                                       HD_parameters_local_scores_unbalanced$global_mse_diagonal, sep=" // "),
                global_mse_off=paste(global_mse_off,
                                       HD_parameters_local_scores_unbalanced$global_mse_off, sep=" // "),
                global_bias_diagonal=paste(global_bias_diagonal,
                                           HD_parameters_local_scores_unbalanced$global_bias_diagonal, sep=" // "),
                global_bias_off=paste(global_bias_off,
                                           HD_parameters_local_scores_unbalanced$global_bias_off, sep=" // "))
```

```{r HD-overlapping-spherical-pdf-offterms, eval=knitr::is_latex_output()}
spherical_HD_parameters_local_scores_kable <- spherical_HD_parameters_local_scores %>%
  kbl(booktabs=T, caption = "Minimal example setting apart MSE and Bias whether it proceeds from
  diagonal or offset terms of the covariance matrix, for scenarios HD5a) and HD6a), in Table
\\ref{tab:parameter-configuration-HD} (overlapping and spherical-distributed components).
      We delimite each scenario by doubled backslashes with respectively balanced and unbalanced clusters.", escape=F,
      col.names = linebreak(c("Package", "Initialisation\nMethod",
                              "Global\nMSE $\\text{diag} (\\boldsymbol{\\Sigma})$", "Global\nMSE $\\text{upper.tri} (\\boldsymbol{\\Sigma})$",
                              "Global\nBias $\\text{diag} (\\boldsymbol{\\Sigma})$", "Global\nBias $\\text{upper.tri} (\\boldsymbol{\\Sigma})$"))) %>%
  kable_styling(latex_options=c("hold_position", "scale_down"), full_width = T) %>%
  row_spec(0,bold=T, align = "c") %>%
  column_spec(column = 1:6, latex_valign = "m") %>%
  column_spec(1, bold = T) %>% collapse_rows(columns = 1, valign = "middle", latex_hline = "major")

for (metric_col in colnames(spherical_HD_parameters_local_scores)[-c(1,2)]) {
  averaged_metric <- purrr::map_dbl(spherical_HD_parameters_local_scores %>% pull(metric_col) %>%
                                      stringr::str_split(pattern = " // "),
                                    ~mean(as.numeric(.x)))
  min_index <- which(averaged_metric == min(averaged_metric))
  max_index <- which(averaged_metric == max(averaged_metric))

  col_vector <- rep("black", nrow(spherical_HD_parameters_local_scores))
  col_vector[min_index] <- "green"; col_vector[max_index] <- "red"
    spherical_HD_parameters_local_scores_kable <- spherical_HD_parameters_local_scores_kable %>%
      column_spec(match(metric_col, names(spherical_HD_parameters_local_scores)),color=col_vector)
}

spherical_HD_parameters_local_scores_kable %>%
  row_spec(8,  hline_after = T)
```

```{r HD-overlapping-spherical-html-offterms, eval=knitr::is_html_output(), layout = "l-body-outset"}
spherical_HD_parameters_local_scores_flex <- spherical_HD_parameters_local_scores %>% flextable() %>%
  set_caption(caption = "Minimal example setting apart MSE and Bias whether it proceeds from
  diagonal or offset terms of the covariance matrix, for scenarios HD5a) and HD6a), in Table
      \\@ref(tab:parameter-configuration-HD) (overlapping and spherical-distributed components).
      We delimite by doubled backslashes for each entry of the summary metrics table.", html_escape = F)

for (metric_col in colnames(spherical_HD_parameters_local_scores)[-c(1,2)]) {
  averaged_metric <- purrr::map_dbl(spherical_HD_parameters_local_scores %>% pull(metric_col) %>%
                                      stringr::str_split(pattern = " // "),
                                    ~mean(as.numeric(.x)))
  min_index <- which(averaged_metric == min(averaged_metric))
  max_index <- which(averaged_metric == max(averaged_metric))

  spherical_HD_parameters_local_scores_flex <- spherical_HD_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body", i = min_index,color = "green")
  spherical_HD_parameters_local_scores_flex <- spherical_HD_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body", i = max_index,color = "red")
}


spherical_HD_parameters_local_scores_flex %>%
  flextable::compose(j="global_mse_diagonal", part = "header", value = as_paragraph("Global MSE Diag(\U03A3)")) %>%
  flextable::compose(j="global_mse_off", part = "header", value = as_paragraph("Global MSE Upper(\U03A3)")) %>%
  flextable::compose(j="global_bias_diagonal", part = "header", value = as_paragraph("Global bias Diag(\U03A3)")) %>%
  flextable::compose(j="global_bias_off", part = "header", value = as_paragraph("Global bias Upper(\U03A3)")) %>%
  set_header_labels(package= "Package", `initialisation method`="Initialisation Method") %>%
  theme_vanilla() %>% align(align = "center", part = "header") %>%  # centre align header
  merge_v(j = "Package") %>%  flextable::hline(i = 8, border = officer::fp_border(width = 2)) %>%
  flextable::autofit()
```



\newpage


```{r HD-overlapping-spherical-plot, fig.cap=if (html_output) html_caption_HD_overlapping_spherical else pdf_caption_HD_overlapping_spherical, out.width="80%", layout = "l-body-outset"}
true_theta_balanced <- HD_configuration %>% filter(ID=="5a") %>%
  pull(true_parameters) %>% magrittr::extract2(1) %>% RGMMBench:::enforce_identifiability()
true_theta_unbalanced <- HD_configuration %>% filter(ID=="6a") %>% pull(true_parameters) %>%
  magrittr::extract2(1)  %>% RGMMBench:::enforce_identifiability()

# generate plots
HD_density <- plot_HD_density_distribution(true_theta=true_theta_balanced, n = 200, k = 2, ade_plot = F)
boxplot_parameter_HD_balanced <- plot_boxplots_parameters(formatted_HD_distributions %>% filter(ID == "5a"),
                                                 num_col = 2, true_theta = true_theta_balanced,
                                                 match_symbol = "^p[[:digit:]]+$|mu_var1_|sd_var1_var1|sd_var2_var1|sd_var2_var2") +
  theme(axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4))
boxplot_parameter_HD_unbalanced <- plot_boxplots_parameters(formatted_HD_distributions %>% filter(ID == "6a"),
                                                 num_col = 2, true_theta = true_theta_unbalanced,
                                                 match_symbol = "^p[[:digit:]]+$|mu_var1_|sd_var1_var1|sd_var2_var1|sd_var2_var2") +
  theme(axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4))


# concatenate plots together
global_figure <- cowplot::plot_grid(HD_density, boxplot_parameter_HD_balanced +
                                      theme(plot.tag = element_text(size=24, face = "bold")) + labs(tag = "C"),
                                    boxplot_parameter_HD_unbalanced +
                                      theme(plot.tag = element_text(size=24, face = "bold")) + labs(tag = "D"),
                                    nrow = 3, align = 'hv', axis="tblr")
ggsave("figs/HD/HD-overlapping-spherical.png",global_figure, width = 24, height = 32, dpi = 75)

knitr::include_graphics("figs/HD/HD-overlapping-spherical.png", dpi = 75)
```

\newpage
```{r compute-HD-impact-num-observations-tab1}
HD_parameters_local_scores_smalln <- HD_parameters_local_scores_summary %>%
  dplyr::filter(ID=="1a") %>% select(-ID)
HD_parameters_local_scores_highn <- HD_parameters_local_scores_summary %>%
  dplyr::filter(ID=="1b") %>% select(-ID)

specific_HD_parameters_local_scores <- HD_parameters_local_scores_smalln  %>%
  dplyr::mutate(global_mse_p=paste(global_mse_p,HD_parameters_local_scores_highn$global_mse_p, sep=" // "),
                global_mse_mu=paste(global_mse_mu,HD_parameters_local_scores_highn$global_mse_mu, sep=" // "),
                global_mse_sigma=paste(global_mse_sigma,HD_parameters_local_scores_highn$global_mse_sigma, sep=" // "),
                global_bias_p=paste(global_bias_p,HD_parameters_local_scores_highn$global_bias_p, sep=" // "),
                global_bias_mu=paste(global_bias_mu,HD_parameters_local_scores_highn$global_bias_mu, sep=" // "),
                global_bias_sigma=paste(global_bias_sigma,HD_parameters_local_scores_highn$global_bias_sigma, sep=" // "),
                Success=paste(Success,HD_parameters_local_scores_highn$Success, sep=" // "))
```

```{r HD-impact-num-observations-pdf-tab1, eval=knitr::is_latex_output()}

specific_HD_parameters_local_scores_kable <- specific_HD_parameters_local_scores %>%
  kbl(booktabs=T, caption = "MSE and Bias associated to scenarios HD1a) and HD1b), in Table
      \\ref{tab:parameter-configuration-HD} (well-separated and spherical-distributed components).
      We delimite by doubled backslashes for each entry of the summary metrics table respectively the scores with $n=200$ and $n=2000$ observations.", escape=F,
      col.names = linebreak(c("Package", "Initialisation\nMethod",
                              "Global \n MSE $p$", "Global\nMSE $\\mu$", "Global\nMSE $\\sigma$",
                              "Global \n Bias $p$", "Global\nBias $\\mu$", "Global\nBias $\\sigma$", "$\\%$ Success"))) %>%
  kable_styling(latex_options=c("hold_position", "scale_down"), full_width = T) %>%
  row_spec(0,bold=T, align = "c") %>%
  column_spec(column = 1:9, latex_valign = "m") %>%
  column_spec(1, bold = T) %>% collapse_rows(columns = 1, valign = "middle", latex_hline = "major")

for (metric_col in c(metric_colnames, "Success")) {
  averaged_metric <- purrr::map_dbl(specific_HD_parameters_local_scores %>% pull(metric_col) %>%
                                      stringr::str_split(pattern = " // "),
                                    ~mean(as.numeric(.x)))
  min_index <- which(averaged_metric == min(averaged_metric))
  max_index <- which(averaged_metric == max(averaged_metric))
  if (metric_col =="Success")  seqinr::swap(min_index, max_index)

  col_vector <- rep("black", nrow(specific_HD_parameters_local_scores))
  col_vector[min_index] <- "green"; col_vector[max_index] <- "red"
  specific_HD_parameters_local_scores_kable <- specific_HD_parameters_local_scores_kable %>%
        column_spec(match(metric_col, names(specific_HD_parameters_local_scores)),color=col_vector)
}

specific_HD_parameters_local_scores_kable %>%
  row_spec(12,  hline_after = T)
```

```{r HD-impact-num-observations-html-tab1, eval=knitr::is_html_output(), layout = "l-body-outset"}
specific_HD_parameters_local_scores_flex <- specific_HD_parameters_local_scores %>% flextable() %>%
  set_caption(caption = "MSE and Bias associated to scenarios HD1a) and HD1b), in Table
      \\@ref(tab:parameter-configuration-HD) (well-separated and spherical-distributed components).
      We delimite by doubled backslashes for each entry of the summary metrics table respectively the scores with n=200 and n=2000 observations.", html_escape = F)

for (metric_col in c(metric_colnames, "Success")) {
  averaged_metric <- purrr::map_dbl(specific_HD_parameters_local_scores %>% pull(metric_col) %>%
                                      stringr::str_split(pattern = " // "),
                                    ~mean(as.numeric(.x)))
  min_index <- which(averaged_metric == min(averaged_metric))
  max_index <- which(averaged_metric == max(averaged_metric))
  if (metric_col =="Success")  seqinr::swap(min_index, max_index)

  specific_HD_parameters_local_scores_flex <- specific_HD_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body", i = min_index,color = "green")
  specific_HD_parameters_local_scores_flex <- specific_HD_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body", i = max_index,color = "red")
}


specific_HD_parameters_local_scores_flex %>%
  flextable::compose(j="global_mse_p", part = "header", value = as_paragraph("Global MSE \U0070")) %>%
  flextable::compose(j="global_mse_mu", part = "header", value = as_paragraph("Global MSE \U03BC")) %>%
  flextable::compose(j="global_mse_sigma", part = "header", value = as_paragraph("Global MSE \U03C3")) %>%
  flextable::compose(j="global_bias_p", part = "header", value = as_paragraph("Global Bias \U0070")) %>%
  flextable::compose(j="global_bias_mu", part = "header", value = as_paragraph("Global Bias \U03BC")) %>%
  flextable::compose(j="global_bias_sigma", part = "header", value = as_paragraph("Global Bias \U03C3")) %>%
  set_header_labels(package= "Package", `initialisation method`="Initialisation Method") %>%
  theme_vanilla() %>% align(align = "center", part = "header") %>%  # centre align header
  merge_v(j = "Package") %>%  flextable::hline(i = 12, border = officer::fp_border(width = 2)) %>%
  flextable::autofit()
```

```{r compute-HD-impact-num-observations-tab2}
HD_parameters_local_scores_smalln <- HD_parameters_local_scores_summary %>%
  dplyr::filter(ID=="8a") %>% select(-ID) %>% filter(Package!="EMMIXmfa")
HD_parameters_local_scores_highn <- HD_parameters_local_scores_summary %>%
  dplyr::filter(ID=="8b") %>% select(-ID)

specific_HD_parameters_local_scores <- HD_parameters_local_scores_smalln  %>%
  dplyr::mutate(global_mse_p=paste(global_mse_p,HD_parameters_local_scores_highn$global_mse_p, sep=" // "),
                global_mse_mu=paste(global_mse_mu,HD_parameters_local_scores_highn$global_mse_mu, sep=" // "),
                global_mse_sigma=paste(global_mse_sigma,HD_parameters_local_scores_highn$global_mse_sigma, sep=" // "),
                global_bias_p=paste(global_bias_p,HD_parameters_local_scores_highn$global_bias_p, sep=" // "),
                global_bias_mu=paste(global_bias_mu,HD_parameters_local_scores_highn$global_bias_mu, sep=" // "),
                global_bias_sigma=paste(global_bias_sigma,HD_parameters_local_scores_highn$global_bias_sigma, sep=" // "),
                Success=paste(Success,HD_parameters_local_scores_highn$Success, sep=" // "))
```

```{r HD-impact-num-observations-pdf-tab2, eval=knitr::is_latex_output()}
specific_HD_parameters_local_scores_kable <- specific_HD_parameters_local_scores %>%
  kbl(booktabs=T, caption = "MSE and Bias associated to scenarios HD8a) and HD8b), in Table
      \\ref{tab:parameter-configuration-HD} (overlapping components with full covariance structure).
      We delimite by doubled backslashes for each entry of the summary metrics table respectively the scores with $n=200$ and
              $n=2000$ observations.", escape=F,
      col.names = linebreak(c("Package", "Initialisation\nMethod",
                              "Global \n MSE $p$", "Global\nMSE $\\mu$", "Global\nMSE $\\sigma$",
                              "Global \n Bias $p$", "Global\nBias $\\mu$", "Global\nBias $\\sigma$", "$\\%$ Success"))) %>%
  kable_styling(latex_options=c("hold_position", "scale_down"), full_width = T) %>%
  row_spec(0,bold=T, align = "c") %>%
  column_spec(column = 1:9, latex_valign = "m") %>%
  column_spec(1, bold = T) %>% collapse_rows(columns = 1, valign = "middle", latex_hline = "major")

for (metric_col in c(metric_colnames, "Success")) {
  averaged_metric <- purrr::map_dbl(specific_HD_parameters_local_scores %>% pull(metric_col) %>%
                                      stringr::str_split(pattern = " // "),
                                    ~mean(as.numeric(.x)))
  min_index <- which(averaged_metric == min(averaged_metric))
  max_index <- which(averaged_metric == max(averaged_metric))
  if (metric_col =="Success")  seqinr::swap(min_index, max_index)

  col_vector <- rep("black", nrow(specific_HD_parameters_local_scores))
  col_vector[min_index] <- "green"; col_vector[max_index] <- "red"
  specific_HD_parameters_local_scores_kable <- specific_HD_parameters_local_scores_kable %>%
        column_spec(match(metric_col, names(specific_HD_parameters_local_scores)),color=col_vector)
}

specific_HD_parameters_local_scores_kable %>%
  row_spec(12,  hline_after = T)


```

```{r HD-impact-num-observations-html-tab2, eval=knitr::is_html_output(), layout = "l-body-outset"}
specific_HD_parameters_local_scores_flex <- specific_HD_parameters_local_scores %>% flextable() %>%
  set_caption(caption = "MSE and Bias associated to scenarios HD8a) and HD8b), in Table
      \\@ref(tab:parameter-configuration-HD) (overlapping and spherical-distributed components).
      We delimite by doubled backslashes for each entry of the summary metrics table respectively the scores with n=200 and
              n=2000 observations.", html_escape = F)

for (metric_col in c(metric_colnames, "Success")) {
  averaged_metric <- purrr::map_dbl(specific_HD_parameters_local_scores %>% pull(metric_col) %>%
                                      stringr::str_split(pattern = " // "),
                                    ~mean(as.numeric(.x)))
  min_index <- which(averaged_metric == min(averaged_metric))
  max_index <- which(averaged_metric == max(averaged_metric))
  if (metric_col =="Success")  seqinr::swap(min_index, max_index)

  specific_HD_parameters_local_scores_flex <- specific_HD_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body", i = min_index,color = "green")
  specific_HD_parameters_local_scores_flex <- specific_HD_parameters_local_scores_flex %>%
    flextable::color(j=metric_col, part = "body", i = max_index,color = "red")
}


specific_HD_parameters_local_scores_flex %>%
  flextable::compose(j="global_mse_p", part = "header", value = as_paragraph("Global MSE \U0070")) %>%
  flextable::compose(j="global_mse_mu", part = "header", value = as_paragraph("Global MSE \U03BC")) %>%
  flextable::compose(j="global_mse_sigma", part = "header", value = as_paragraph("Global MSE \U03C3")) %>%
  flextable::compose(j="global_bias_p", part = "header", value = as_paragraph("Global Bias \U0070")) %>%
  flextable::compose(j="global_bias_mu", part = "header", value = as_paragraph("Global Bias \U03BC")) %>%
  flextable::compose(j="global_bias_sigma", part = "header", value = as_paragraph("Global Bias \U03C3")) %>%
  set_header_labels(package= "Package", `initialisation method`="Initialisation Method") %>%
  theme_vanilla() %>% align(align = "center", part = "header") %>%  # centre align header
  merge_v(j = "Package") %>%  flextable::hline(i = 12, border = officer::fp_border(width = 2)) %>%
  flextable::autofit()
```
\newpage

```{r HD-impact-num-observations, fig.cap=if (html_output) html_caption_HD_impact_num_observations else pdf_caption_HD_impact_num_observations, out.width="80%"}
hard_theta_small_n <- HD_configuration %>% filter(ID=="8a") %>% pull(true_parameters) %>% magrittr::extract2(1)
hard_theta_high_n <- HD_configuration %>% filter(ID=="8b") %>% pull(true_parameters) %>% magrittr::extract2(1)
boxplot_parameter_hard_small <- plot_boxplots_parameters(formatted_HD_distributions %>% filter(ID == "8a"),
                                                 num_col = 2, true_theta = hard_theta_small_n,
                                                 match_symbol = "^p[[:digit:]]+$|mu_var1_|sd_var1_var1|sd_var2_var1|sd_var2_var2") +
  theme(axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4))
boxplot_parameter_hard_high <- plot_boxplots_parameters(formatted_HD_distributions %>% filter(ID == "8b"),
                                                 num_col = 2, true_theta = hard_theta_high_n,
                                                 match_symbol = "^p[[:digit:]]+$|mu_var1_|sd_var1_var1|sd_var2_var1|sd_var2_var2") +
  theme(axis.text.x = ggtext::element_markdown(angle = 90, size = 12, hjust=0.5, vjust = 0.4))

# concatenate plots together
global_figure <- cowplot::plot_grid(                              boxplot_parameter_hard_small +
                                      theme(plot.tag = element_text(size=24, face = "bold")) + labs(tag = "A"),
                                    boxplot_parameter_hard_high +
                                      theme(plot.tag = element_text(size=24, face = "bold")) + labs(tag = "B"),
                                    nrow = 2, ncol=2, align = 'hv', axis="tblr")

ggsave("figs/HD/HD-impact-nobservations.png",global_figure, width = 30, height = 40, dpi = 75)

knitr::include_graphics("figs/HD/HD-impact-nobservations.png", dpi = 75)
```


```{r heatmap-all-correlation-plots-HD, fig.cap= if (html_output) html_global_heatmap_HD else pdf_global_heatmap_HD, out.width="80%", layout = "l-body-outset"}
multivariate_correlation_heatmaps <- plot_correlation_Heatmap(HD_distributions %>% filter(ID=="7a"), row_km = 3)
multivariate_correlation_heatmaps <- gridExtra::arrangeGrob(grobs=multivariate_correlation_heatmaps %>%
                                                              purrr::map(~grid::grid.grabExpr(ComplexHeatmap::draw(.x))),
                                                            top="", ncol = 3, nrow = 1)
ggsave("./figs/HD/heatmap_global_HD.png", multivariate_correlation_heatmaps, width = 18, height = 6,dpi = 75)
knitr::include_graphics("./figs/HD/heatmap_global_HD.png", dpi = 75)
```

\newpage 
# Supplementary bibliography{.appendix}

<div id="refs"></div>
